{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fa877c14",
   "metadata": {},
   "source": [
    "# TF-co-occurences for WP2 - Data\n",
    "### Outline of this notebook:\n",
    "    1. Constants, Path and Interface Definitions \n",
    "    2. Market Basket analysis with tf-comb, for all cluster/celltypes of a tissue\n",
    "    3. Differential analysis with all market basket analysis (CombObj ´s) of step 2 for the clusters/celltypes of a single tissue. (One DiffCombObj for each tissue)\n",
    "    4. Analysis for biological questions \n",
    "    \n",
    "The aim is to find transcription-factor-co-occurences for cluster/celltypes of human tissues with the help of the python-library tf-comb. For the data of WP2. The data basis comes from the \"cell atlas of chromatin accessibility across 25 adult human tissues\"(https://doi.org/10.1101/2021.02.17.431699) \n",
    "\n",
    "**Biological question, that we want to answer with this notebook:**\n",
    "\n",
    "1. Find Transcriptionfactor-co-occurences, which only occure in one (or more) \"cluster/celltypes of a tissue\".\n",
    "    Maybe we can identify a cluster through this co-occurences.\n",
    "\n",
    "**How to use this notbook:**\n",
    "1. Please adapted the paths in Constants, Path and interface defintions for your approach\n",
    "2. Please make sure you have installed the kernel as it is described in the ReadMe\n",
    "3. Check if the WP2 Data structure is correctly provided (ReadMe).\n",
    "    - Has all tissues as a folder in it: ../OUPTPUT_FOLDER/\n",
    "    - You find the data (open cromatin regions per cluster .bed-files) e.g ../OUPTPUT_FOLDER/<tissue>/WP6/*cluster_x.bed\n",
    "4. Execute each notebook window from top to bottom one after another."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa1aab90",
   "metadata": {},
   "source": [
    "## 1. Constants, Path and Interface Definitions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "487586df",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tfcomb import CombObj, DiffCombObj, utils\n",
    "import os\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "'''\n",
    "Constants for this script.\n",
    "\n",
    "This window contains all paths and constants, which are later used for this juypter notbook.\n",
    "\n",
    "Please adapt paths or constants, if use other files. \n",
    "For example adapted the genome path, if you use another genome. \n",
    "'''\n",
    "\n",
    "# Path to genome fasta file. Is used for the market basket analysis of tfcomb.\n",
    "genome_path=\"/mnt/workspace_stud/allstud/homo_sapiens.104.mainChr.fa\"\n",
    "\n",
    "# Path to the jaspar file (contains transcription factors (TF) binding profiles\n",
    "# as position frequency matrices (PFMs)). Is used for the market basket analysis of tfcomb\n",
    "main_jaspar_file=\"../testdaten/JASPAR2020_CORE_vertebrates.meme\" \n",
    "\n",
    "# Path where results of this notebook will be written to (eg. TF_COMB objects, .pkl).\n",
    "result_path=\"./results/wp2/\"\n",
    "\n",
    "# Paths in the result folder:\n",
    "# Path to folder, where the resulting market basket analysis for a cluster/celltype is put \n",
    "main_analysis_path=f\"{result_path}main/\"\n",
    "\n",
    "# Path to folder, where the differential analysis for a tissue is put \n",
    "differential_analysis_path=f\"{result_path}diff_analysis/\"\n",
    "# Path to folder, where answers of our results are put to\n",
    "answers_path=f\"{result_path}answers/\"\n",
    "\n",
    "# Path to folder (Interface folder) of wp2, where the clusters of each tissue can be found.\n",
    "path_to_tissues=\"/mnt/workspace_stud/stud3/WP2_OUTPUT/FINISHED/\"\n",
    "\n",
    "# Tag for WP6 data in WP2 Interface\n",
    "cluster_folder_tag=\"WP6/\"\n",
    "\n",
    "# Path to Folder with celltype annotation tables of wp2, not longer needed\n",
    "#celltype_annotation_path = \"/mnt/workspace_stud/stud4/celltype_assignment_tables/\"\n",
    "\n",
    "# The following lines, initally check if all file/paths are available. \n",
    "# If a result folder does not exist it is created automatically\n",
    "if not os.path.exists(result_path):\n",
    "     pathlib.Path(result_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(main_analysis_path):\n",
    "     pathlib.Path(main_analysis_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(differential_analysis_path):\n",
    "     pathlib.Path(differential_analysis_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(answers_path):\n",
    "     pathlib.Path(answers_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(genome_path):\n",
    "    print(f\"ERROR: path {genome_path} does not exist\")\n",
    "\n",
    "if not os.path.exists(main_jaspar_file):\n",
    "    print(f\"ERROR: path {main_jaspar_file} does not exist\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11adea88",
   "metadata": {},
   "source": [
    "### Helper functions for reading-in folders/files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "92e608ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_folder_names_in_folder(rel_folder_path:str):\n",
    "    ''' \n",
    "        Read in the names of the folders in a folder.(rel_folder_path)\n",
    "        ---\n",
    "        Parameters:\n",
    "        \n",
    "        rel_folder_path: String\n",
    "            relative Path to the folder location that is read in.\n",
    "        ---\n",
    "        Return: a List of Strings (folder names)\n",
    "    '''\n",
    "\n",
    "    dirlist = [ item for item in os.listdir(rel_folder_path) if os.path.isdir(os.path.join(rel_folder_path, item))]\n",
    "    folder_names = []\n",
    "    for folder in dirlist:\n",
    "        folder_names.append(folder)\n",
    "    return folder_names\n",
    "\n",
    "def read_in_file_names_of_folder(rel_path:str):\n",
    "    ''' \n",
    "        Read in the file names in a folder (rel_path).\n",
    "        ---\n",
    "        Parameters:\n",
    "        \n",
    "        rel_path: String\n",
    "            relative Path to the folder location.\n",
    "        ---\n",
    "        Return: a List of Strings (file names)\n",
    "    '''\n",
    "    return [f for f in os.listdir(rel_path) if os.path.isfile(os.path.join(rel_path, f))]\n",
    "\n",
    "\n",
    "#cluster_file_names = read_in_file_names_of_folder(rel_path=path_to_clusters)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0dba2e",
   "metadata": {},
   "source": [
    "## 2. Market basket analysis with tf-comb\n",
    "We do a market basket analysis with tfcomb for each cluster/celltype, which has been culstered by WP2 and comes from the raw single-cell ATAC-Data of the cell atlas project. As a result, we get the transcription-factor-co-occurences for each cluster. The Trancriptionsfactor motif´s come as a position-frequency-matrix from https://jaspar.genereg.net/search?q=&collection=CORE&tax_group=vertebrates. The corresponding genome, which is used is **homo_sapiens.104.mainChr.fa** .\n",
    "\n",
    "Approach:\n",
    "1. Read-in tissue folders of WP2\n",
    "2. For each tissue: read-in single .bed (Content= open-chromatin regions) files for each cluster/celltype\n",
    "3. Do market basket analysis for each cluster/celltype\n",
    "4. Result .pkl CombObj files can be found under **/{result_path}/{main_analysis_path}/{tissue_name}/{cluster_name}.pkl **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f22f857",
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_market_basket_analyses_for_cell_cluster(mb_file_name: str, cell_cluster_path:str, tissue:str):\n",
    "    '''\n",
    "        Does market basket analysis with tfcomb. Saves the tfcomb-Object as .pkl file to main_analysis_path.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        mb_file_name: string\n",
    "            Name for the result file of the market basket analyses. \n",
    "            e.g \"<tissue>_<cluster_number>_<celltype>\".\n",
    "        \n",
    "        cell_cluster_path: string\n",
    "            Path to the .bed-File with genome regions to check for tf-co-occurences\n",
    "         \n",
    "        tissue: string\n",
    "            Tissue name, origin of the cluster.\n",
    "    '''    \n",
    "    # Save path initalization, if folder for tissue does not exists, new folder is created.\n",
    "    save_path = f'{main_analysis_path}{tissue}/'\n",
    "    if not os.path.exists(save_path):\n",
    "         pathlib.Path(save_path).mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # TF-comb market basket analysis\n",
    "    comb = CombObj()\n",
    "    comb.TFBS_from_motifs(regions= cell_cluster_path,\n",
    "                   motifs=main_jaspar_file,\n",
    "                   genome=genome_path,\n",
    "                   threads=4)\n",
    "    \n",
    "    print(f'Start market basket analyses for cell-cluster/type: {mb_file_name}')\n",
    "    comb.market_basket(threads=10)\n",
    "    \n",
    "    # if rules are empty nothing is saved \n",
    "    if len(comb.rules) <= 0:\n",
    "        print(f'Could not find TF-cooccurences for cell-cluster/type: {mb_file_name}')\n",
    "        return\n",
    "    print(f'Finished market basket analyses for cell-cluster/type: {mb_file_name}')\n",
    "    print(f'Found rules: {len(comb.rules)}')\n",
    "    \n",
    "    # save tf-comb obj to .pkl\n",
    "    comb.to_pickle(f'{save_path}{mb_file_name}.pkl')\n",
    "    print(f'Saved: {save_path}{mb_file_name}.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0dceb84",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQYCP_lower_leg_skin', 'ACCPU_upper_lobe_of_left_lung', 'A8CPH_esophagus_muscularis_mucosa', 'A8WNZ_right_lobe_of_liver', 'ACCQ1_colon_transverse', 'AZPYJ_esophagus_squamous_epithelium', 'CHLWW_tibial_artery', 'IOBHK_tibial_artery']\n"
     ]
    }
   ],
   "source": [
    "# Load tissue folder names for wp2 data\n",
    "tissues=get_folder_names_in_folder(rel_folder_path=path_to_tissues)\n",
    "print(tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f2f4a27",
   "metadata": {},
   "source": [
    "### Functions for adding the celltype annotations of wp2 to the corresponding clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92e9bb1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Melanocytes'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_celltype_annotation_table(celltype_annotation_path:str) -> pd.DataFrame:\n",
    "    '''\n",
    "    Read in the the given annotation file(celltype_annotation_path) with pandas read_table function\n",
    "    ---\n",
    "    Parameters:\n",
    "    celltype_annotation_path : string\n",
    "        Is the path to a file which has the celltype annotations for the cluster numbers.\n",
    "        e.g. file content:\n",
    "        6\tStromal cells\n",
    "        3\tAdipocyte progenitor cells\n",
    "        4\tStromal cells\n",
    "        7\tFibroblasts\n",
    "    ---\n",
    "    Returns a celltype annotation table as a pandas Dataframe.\n",
    "    '''\n",
    "    df = pd.DataFrame()\n",
    "    df = pd.read_table(celltype_annotation_path, header=None, index_col=None)\n",
    "    return df\n",
    "\n",
    "def get_celltype_for_cluster(cluster_number: int, celltype_annotation_table: pd.DataFrame):\n",
    "    '''\n",
    "    Find celltype annotation for cluster number in the corresponding celltype_annotation_table\n",
    "    ---\n",
    "    Parameters:\n",
    "    cluster_number: integer\n",
    "        number of the cluster\n",
    "    \n",
    "    celltype_annotation_table: pd.DataFrame\n",
    "        table with the celltype annotations, e.g. Dataframe\n",
    "         0\t1 \n",
    "       1 6\tStromal cells\n",
    "       2 3\tAdipocyte progenitor cells\n",
    "       3 4\tStromal cells\n",
    "       4 7\tFibroblasts\n",
    "    ---\n",
    "    Returns the celltype for a cluster_number. (if no celltype kann be found \"na\" is returned) \n",
    "    '''\n",
    "    # initalize celltype name with na (not available), if no celltype can be found na is returned\n",
    "    celltype_name = \"na\"\n",
    "    try:\n",
    "        # Try to find the cell-type annotation in the table, if no annotation can be found key error is raised\n",
    "        celltype_name = celltype_annotation_table[celltype_annotation_table[0] == int(cluster_number)].iat[0,1]\n",
    "        # replace whitespaces in celltype names, e.g. \"Stromal cell\" -> \"Stromal_cell\"\n",
    "        celltype_name = celltype_name.replace(\" \", \"_\")\n",
    "    except (KeyError, IndexError) as e:\n",
    "        print(f\"Could not find celltype annotation for cluster: {cluster_number}, set name to na.\")\n",
    "        celltype_name = \"na\"\n",
    "   \n",
    "    return celltype_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "392e4bd1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def make_mb_for_clusters(path_to_clusters:str, tissue:str):\n",
    "    '''\n",
    "        Wrapper function, that does the market basket analysis for all clusters/celltypes in a tissue.\n",
    "        Also annotates the cluster with a celltype.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        path_to_clusters: string\n",
    "            Path to the .bed files(cluster) of a tissue.\n",
    "         \n",
    "        tissue: string\n",
    "            Tissue name from where the cluster corresponds to.\n",
    "        --- \n",
    "        Catch Exceptions with a message, if any error occures in the market basket analyses a message is printed\n",
    "        and the programm continues with the next tissue.\n",
    "    '''\n",
    "    # Read in the .bed files for each cluster of the specific tissue\n",
    "    cluster_file_names = read_in_file_names_of_folder(rel_path=path_to_clusters)\n",
    "    print(cluster_file_names)\n",
    "    \n",
    "    # Get the celltype annotation table. To add celltype for a cluster. e.g. cluster6 = fibroblast\n",
    "    celltype_table = get_celltype_annotation_table(\n",
    "        celltype_annotation_path=f'{path_to_tissues}{tissue}/annotation/annotation.txt') # pandas dataframe is returned\n",
    "    \n",
    "    # Do a market basket analysis for each cluster of a tissue\n",
    "    for file_name in cluster_file_names:\n",
    "        # e.g JF1O6_body_of_pancreas.10_peaks.bed -> [JF1O6_body_of_pancreas.10_peaks] = clustername = JF1O6_body_of_pancreas.10_peaks\n",
    "        cluster_name = file_name.split('.bed')[0]\n",
    "        print(cluster_name)\n",
    "        #e.g JF1O6_body_of_pancreas.10_peaks -> [JF1O6_body_of_pancreas, 10_peaks], cluster_number_with_tag = 10_peaks\n",
    "        cluster_number_with_tag = cluster_name.split('.')[1]\n",
    "        # e.g 10_peaks -> [10 , peaks], cluster_number = 10\n",
    "        cluster_number = int(cluster_number_with_tag.split('_')[0])\n",
    "        \n",
    "        celltype_name = get_celltype_for_cluster(cluster_number=cluster_number,\n",
    "                                                 celltype_annotation_table=celltype_table)\n",
    "        \n",
    "        try:\n",
    "            print(cluster_name)\n",
    "            print(cluster_number)\n",
    "            print(celltype_name)\n",
    "            print(file_name)\n",
    "            # Prepare names and paths\n",
    "            cluster_path=f\"{path_to_clusters}{file_name}\"\n",
    "            # e.g. JF1O6_body_of_pancreas_c3_fibroblast\n",
    "            mb_file_name = f\"{tissue}_c{str(cluster_number)}_{celltype_name}\"\n",
    "            \n",
    "            do_market_basket_analyses_for_cell_cluster(mb_file_name=mb_file_name, cell_cluster_path=cluster_path, tissue=tissue)\n",
    "        except Exception:\n",
    "            print(f\"ERROR: Market basket for cluster:{cluster_name} in tissue {tissue}, did not work\")\n",
    "            continue\n",
    "            \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e14e7350",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQYCP_lower_leg_skin.4_peaks.bed', 'IQYCP_lower_leg_skin.5_peaks.bed', 'IQYCP_lower_leg_skin.1_peaks.bed', 'IQYCP_lower_leg_skin.2_peaks.bed', 'IQYCP_lower_leg_skin.3_peaks.bed', 'IQYCP_lower_leg_skin.6_peaks.bed', 'IQYCP_lower_leg_skin.10_peaks.bed', 'IQYCP_lower_leg_skin.9_peaks.bed', 'IQYCP_lower_leg_skin.7_peaks.bed']\n",
      "IQYCP_lower_leg_skin.4_peaks\n",
      "IQYCP_lower_leg_skin.4_peaks\n",
      "4\n",
      "Melanocytes\n",
      "IQYCP_lower_leg_skin.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 95866 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c4_Melanocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c4_Melanocytes\n",
      "Found rules: 143345\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c4_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin.5_peaks\n",
      "Could not find celltype annotation for cluster: 5, set name to na.\n",
      "IQYCP_lower_leg_skin.5_peaks\n",
      "5\n",
      "na\n",
      "IQYCP_lower_leg_skin.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 24762 TFBS (736 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c5_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 26%\n",
      "INFO: Progress: 34%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c5_na\n",
      "Found rules: 52221\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c5_na.pkl\n",
      "IQYCP_lower_leg_skin.1_peaks\n",
      "IQYCP_lower_leg_skin.1_peaks\n",
      "1\n",
      "Melanocytes\n",
      "IQYCP_lower_leg_skin.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 21%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 112819 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c1_Melanocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c1_Melanocytes\n",
      "Found rules: 159804\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c1_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin.2_peaks\n",
      "IQYCP_lower_leg_skin.2_peaks\n",
      "2\n",
      "Adipocytes\n",
      "IQYCP_lower_leg_skin.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 233451 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 16%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 54%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "Found rules: 266532\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c2_Adipocytes.pkl\n",
      "IQYCP_lower_leg_skin.3_peaks\n",
      "IQYCP_lower_leg_skin.3_peaks\n",
      "3\n",
      "Sebocytes\n",
      "IQYCP_lower_leg_skin.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 245785 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "Found rules: 278318\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c3_Sebocytes.pkl\n",
      "IQYCP_lower_leg_skin.6_peaks\n",
      "IQYCP_lower_leg_skin.6_peaks\n",
      "6\n",
      "Melanocytes\n",
      "IQYCP_lower_leg_skin.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 379961 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 34%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "Found rules: 348686\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c6_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin.10_peaks\n",
      "Could not find celltype annotation for cluster: 10, set name to na.\n",
      "IQYCP_lower_leg_skin.10_peaks\n",
      "10\n",
      "na\n",
      "IQYCP_lower_leg_skin.10_peaks.bed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 9%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 29%\n",
      "INFO: Progress: 37%\n",
      "INFO: Progress: 46%\n",
      "INFO: Progress: 57%\n",
      "INFO: Progress: 65%\n",
      "INFO: Progress: 73%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 49029 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c10_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 34%\n",
      "INFO: Progress: 46%\n",
      "INFO: Progress: 58%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c10_na\n",
      "Found rules: 85148\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c10_na.pkl\n",
      "IQYCP_lower_leg_skin.9_peaks\n",
      "Could not find celltype annotation for cluster: 9, set name to na.\n",
      "IQYCP_lower_leg_skin.9_peaks\n",
      "9\n",
      "na\n",
      "IQYCP_lower_leg_skin.9_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 38433 TFBS (744 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c9_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 18%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 46%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c9_na\n",
      "Found rules: 72819\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c9_na.pkl\n",
      "IQYCP_lower_leg_skin.7_peaks\n",
      "IQYCP_lower_leg_skin.7_peaks\n",
      "7\n",
      "Melanocytes\n",
      "IQYCP_lower_leg_skin.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 290210 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 34%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "Found rules: 302189\n",
      "Saved: ./results/wp2/main/IQYCP_lower_leg_skin/IQYCP_lower_leg_skin_c7_Melanocytes.pkl\n",
      "['ACCPU_upper_lobe_of_left_lung.3_peaks.bed', 'ACCPU_upper_lobe_of_left_lung.1_peaks.bed', 'ACCPU_upper_lobe_of_left_lung.2_peaks.bed']\n",
      "ACCPU_upper_lobe_of_left_lung.3_peaks\n",
      "Could not find celltype annotation for cluster: 3, set name to na.\n",
      "ACCPU_upper_lobe_of_left_lung.3_peaks\n",
      "3\n",
      "na\n",
      "ACCPU_upper_lobe_of_left_lung.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 49%\n",
      "INFO: Progress: 59%\n",
      "INFO: Progress: 69%\n",
      "INFO: Progress: 79%\n",
      "INFO: Progress: 89%\n",
      "INFO: Progress: 99%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 4956 TFBS (586 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c3_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 56%\n",
      "INFO: Progress: 78%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c3_na\n",
      "Found rules: 17591\n",
      "Saved: ./results/wp2/main/ACCPU_upper_lobe_of_left_lung/ACCPU_upper_lobe_of_left_lung_c3_na.pkl\n",
      "ACCPU_upper_lobe_of_left_lung.1_peaks\n",
      "ACCPU_upper_lobe_of_left_lung.1_peaks\n",
      "1\n",
      "Clara_cells\n",
      "ACCPU_upper_lobe_of_left_lung.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 9%\n",
      "INFO: Progress: 19%\n",
      "INFO: Progress: 28%\n",
      "INFO: Progress: 38%\n",
      "INFO: Progress: 47%\n",
      "INFO: Progress: 56%\n",
      "INFO: Progress: 66%\n",
      "INFO: Progress: 75%\n",
      "INFO: Progress: 85%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 4961 TFBS (558 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c1_Clara_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 88%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c1_Clara_cells\n",
      "Found rules: 16269\n",
      "Saved: ./results/wp2/main/ACCPU_upper_lobe_of_left_lung/ACCPU_upper_lobe_of_left_lung_c1_Clara_cells.pkl\n",
      "ACCPU_upper_lobe_of_left_lung.2_peaks\n",
      "ACCPU_upper_lobe_of_left_lung.2_peaks\n",
      "2\n",
      "Adipocytes\n",
      "ACCPU_upper_lobe_of_left_lung.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 18%\n",
      "INFO: Progress: 28%\n",
      "INFO: Progress: 36%\n",
      "INFO: Progress: 45%\n",
      "INFO: Progress: 55%\n",
      "INFO: Progress: 65%\n",
      "INFO: Progress: 73%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 41527 TFBS (740 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 26%\n",
      "INFO: Progress: 38%\n",
      "INFO: Progress: 48%\n",
      "INFO: Progress: 58%\n",
      "INFO: Progress: 68%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
      "Found rules: 93333\n",
      "Saved: ./results/wp2/main/ACCPU_upper_lobe_of_left_lung/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes.pkl\n",
      "['A8CPH_esophagus_muscularis_mucosa.4_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.6_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.2_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.7_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.1_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.3_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.9_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.5_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.8_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.10_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.13_peaks.bed', 'A8CPH_esophagus_muscularis_mucosa.11_peaks.bed']\n",
      "A8CPH_esophagus_muscularis_mucosa.4_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.4_peaks\n",
      "4\n",
      "Enterocytes\n",
      "A8CPH_esophagus_muscularis_mucosa.4_peaks.bed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 21%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 323521 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes\n",
      "Found rules: 299792\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.6_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.6_peaks\n",
      "6\n",
      "Enterocytes\n",
      "A8CPH_esophagus_muscularis_mucosa.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 171460 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 16%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes\n",
      "Found rules: 207348\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.2_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.2_peaks\n",
      "2\n",
      "Enterocytes\n",
      "A8CPH_esophagus_muscularis_mucosa.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 226661 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes\n",
      "Found rules: 264399\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.7_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.7_peaks\n",
      "7\n",
      "Fibroblasts\n",
      "A8CPH_esophagus_muscularis_mucosa.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 470983 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 54%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "Found rules: 381326\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.1_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.1_peaks\n",
      "1\n",
      "Fibroblasts\n",
      "A8CPH_esophagus_muscularis_mucosa.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 785174 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "Found rules: 452808\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.3_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.3_peaks\n",
      "3\n",
      "Gastric_chief_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 594980 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "Found rules: 408631\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.9_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.9_peaks\n",
      "9\n",
      "Adipocyte_progenitor_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.9_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 264518 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 36%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "Found rules: 281976\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.5_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.5_peaks\n",
      "5\n",
      "Stromal_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 963536 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "Found rules: 473140\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.8_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.8_peaks\n",
      "8\n",
      "Gastric_chief_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.8_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 723181 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "Found rules: 442181\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.10_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.10_peaks\n",
      "10\n",
      "Stromal_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.10_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 614779 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "Found rules: 412716\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.13_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.13_peaks\n",
      "13\n",
      "Gastric_chief_cells\n",
      "A8CPH_esophagus_muscularis_mucosa.13_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1205042 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "Found rules: 488805\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa.11_peaks\n",
      "A8CPH_esophagus_muscularis_mucosa.11_peaks\n",
      "11\n",
      "Fibroblasts\n",
      "A8CPH_esophagus_muscularis_mucosa.11_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 620862 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "Found rules: 416934\n",
      "Saved: ./results/wp2/main/A8CPH_esophagus_muscularis_mucosa/A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts.pkl\n",
      "['A8WNZ_right_lobe_of_liver.5_peaks.bed', 'A8WNZ_right_lobe_of_liver.2_peaks.bed', 'A8WNZ_right_lobe_of_liver.1_peaks.bed', 'A8WNZ_right_lobe_of_liver.3_peaks.bed', 'A8WNZ_right_lobe_of_liver.6_peaks.bed', 'A8WNZ_right_lobe_of_liver.8_peaks.bed', 'A8WNZ_right_lobe_of_liver.4_peaks.bed', 'A8WNZ_right_lobe_of_liver.7_peaks.bed']\n",
      "A8WNZ_right_lobe_of_liver.5_peaks\n",
      "A8WNZ_right_lobe_of_liver.5_peaks\n",
      "5\n",
      "Cholangiocytes\n",
      "A8WNZ_right_lobe_of_liver.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 80634 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 18%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes\n",
      "Found rules: 146621\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c5_Cholangiocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver.2_peaks\n",
      "A8WNZ_right_lobe_of_liver.2_peaks\n",
      "2\n",
      "Kupffer_cells\n",
      "A8WNZ_right_lobe_of_liver.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 51%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 142003 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells\n",
      "Found rules: 208662\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c2_Kupffer_cells.pkl\n",
      "A8WNZ_right_lobe_of_liver.1_peaks\n",
      "A8WNZ_right_lobe_of_liver.1_peaks\n",
      "1\n",
      "Adipocytes\n",
      "A8WNZ_right_lobe_of_liver.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 217734 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c1_Adipocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 54%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c1_Adipocytes\n",
      "Found rules: 266490\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c1_Adipocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver.3_peaks\n",
      "A8WNZ_right_lobe_of_liver.3_peaks\n",
      "3\n",
      "Hepatocytes\n",
      "A8WNZ_right_lobe_of_liver.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 405156 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "Found rules: 380088\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c3_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver.6_peaks\n",
      "A8WNZ_right_lobe_of_liver.6_peaks\n",
      "6\n",
      "Hepatocytes\n",
      "A8WNZ_right_lobe_of_liver.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 436490 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "Found rules: 387020\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c6_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver.8_peaks\n",
      "A8WNZ_right_lobe_of_liver.8_peaks\n",
      "8\n",
      "Adipocyte_progenitor_cells\n",
      "A8WNZ_right_lobe_of_liver.8_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 355158 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 76%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "Found rules: 354942\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells.pkl\n",
      "A8WNZ_right_lobe_of_liver.4_peaks\n",
      "A8WNZ_right_lobe_of_liver.4_peaks\n",
      "4\n",
      "Hepatocytes\n",
      "A8WNZ_right_lobe_of_liver.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 925899 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "Found rules: 469808\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c4_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver.7_peaks\n",
      "A8WNZ_right_lobe_of_liver.7_peaks\n",
      "7\n",
      "Hepatocytes\n",
      "A8WNZ_right_lobe_of_liver.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 633451 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "Found rules: 434062\n",
      "Saved: ./results/wp2/main/A8WNZ_right_lobe_of_liver/A8WNZ_right_lobe_of_liver_c7_Hepatocytes.pkl\n",
      "['ACCQ1_colon_transverse.5_peaks.bed', 'ACCQ1_colon_transverse.3_peaks.bed', 'ACCQ1_colon_transverse.2_peaks.bed', 'ACCQ1_colon_transverse.6_peaks.bed', 'ACCQ1_colon_transverse.1_peaks.bed', 'ACCQ1_colon_transverse.4_peaks.bed', 'ACCQ1_colon_transverse.7_peaks.bed']\n",
      "ACCQ1_colon_transverse.5_peaks\n",
      "ACCQ1_colon_transverse.5_peaks\n",
      "5\n",
      "Microfold_cells\n",
      "ACCQ1_colon_transverse.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 42964 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c5_Microfold_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 28%\n",
      "INFO: Progress: 38%\n",
      "INFO: Progress: 46%\n",
      "INFO: Progress: 56%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 76%\n",
      "INFO: Progress: 84%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c5_Microfold_cells\n",
      "Found rules: 88393\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c5_Microfold_cells.pkl\n",
      "ACCQ1_colon_transverse.3_peaks\n",
      "ACCQ1_colon_transverse.3_peaks\n",
      "3\n",
      "Enterocytes\n",
      "ACCQ1_colon_transverse.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 119146 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c3_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c3_Enterocytes\n",
      "Found rules: 178384\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c3_Enterocytes.pkl\n",
      "ACCQ1_colon_transverse.2_peaks\n",
      "ACCQ1_colon_transverse.2_peaks\n",
      "2\n",
      "Goblet_cells\n",
      "ACCQ1_colon_transverse.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 178780 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c2_Goblet_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 96%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c2_Goblet_cells\n",
      "Found rules: 255592\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c2_Goblet_cells.pkl\n",
      "ACCQ1_colon_transverse.6_peaks\n",
      "ACCQ1_colon_transverse.6_peaks\n",
      "6\n",
      "Gastric_chief_cells\n",
      "ACCQ1_colon_transverse.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 283660 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "Found rules: 323192\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c6_Gastric_chief_cells.pkl\n",
      "ACCQ1_colon_transverse.1_peaks\n",
      "ACCQ1_colon_transverse.1_peaks\n",
      "1\n",
      "Gastric_chief_cells\n",
      "ACCQ1_colon_transverse.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 302030 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "Found rules: 305626\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c1_Gastric_chief_cells.pkl\n",
      "ACCQ1_colon_transverse.4_peaks\n",
      "ACCQ1_colon_transverse.4_peaks\n",
      "4\n",
      "Paneth_cells\n",
      "ACCQ1_colon_transverse.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 467440 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 42%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "Found rules: 394181\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c4_Paneth_cells.pkl\n",
      "ACCQ1_colon_transverse.7_peaks\n",
      "ACCQ1_colon_transverse.7_peaks\n",
      "7\n",
      "Fibroblasts\n",
      "ACCQ1_colon_transverse.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 147733 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "Found rules: 205018\n",
      "Saved: ./results/wp2/main/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c7_Fibroblasts.pkl\n",
      "['AZPYJ_esophagus_squamous_epithelium.5_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.4_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.10_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.2_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.3_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.6_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.1_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.8_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.7_peaks.bed', 'AZPYJ_esophagus_squamous_epithelium.9_peaks.bed']\n",
      "AZPYJ_esophagus_squamous_epithelium.5_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.5_peaks\n",
      "5\n",
      "Fibroblasts\n",
      "AZPYJ_esophagus_squamous_epithelium.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 376958 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 42%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 74%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts\n",
      "Found rules: 346934\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.4_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.4_peaks\n",
      "4\n",
      "Fibroblasts\n",
      "AZPYJ_esophagus_squamous_epithelium.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 458282 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts\n",
      "Found rules: 376264\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.10_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.10_peaks\n",
      "10\n",
      "Fibroblasts\n",
      "AZPYJ_esophagus_squamous_epithelium.10_peaks.bed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 474798 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts\n",
      "Found rules: 387330\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.2_peaks\n",
      "Could not find celltype annotation for cluster: 2, set name to na.\n",
      "AZPYJ_esophagus_squamous_epithelium.2_peaks\n",
      "2\n",
      "na\n",
      "AZPYJ_esophagus_squamous_epithelium.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1221931 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "Found rules: 487551\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c2_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.3_peaks\n",
      "Could not find celltype annotation for cluster: 3, set name to na.\n",
      "AZPYJ_esophagus_squamous_epithelium.3_peaks\n",
      "3\n",
      "na\n",
      "AZPYJ_esophagus_squamous_epithelium.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1443807 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "Found rules: 500752\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c3_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.6_peaks\n",
      "Could not find celltype annotation for cluster: 6, set name to na.\n",
      "AZPYJ_esophagus_squamous_epithelium.6_peaks\n",
      "6\n",
      "na\n",
      "AZPYJ_esophagus_squamous_epithelium.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1629732 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "Found rules: 505905\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c6_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.1_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.1_peaks\n",
      "1\n",
      "Enterocytes\n",
      "AZPYJ_esophagus_squamous_epithelium.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1629281 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "Found rules: 506259\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.8_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.8_peaks\n",
      "8\n",
      "Adipocytes\n",
      "AZPYJ_esophagus_squamous_epithelium.8_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 51%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 349339 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 16%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "Found rules: 334458\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.7_peaks\n",
      "Could not find celltype annotation for cluster: 7, set name to na.\n",
      "AZPYJ_esophagus_squamous_epithelium.7_peaks\n",
      "7\n",
      "na\n",
      "AZPYJ_esophagus_squamous_epithelium.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1263785 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "Found rules: 489458\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c7_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium.9_peaks\n",
      "AZPYJ_esophagus_squamous_epithelium.9_peaks\n",
      "9\n",
      "Enterocytes\n",
      "AZPYJ_esophagus_squamous_epithelium.9_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 51%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1774176 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "Found rules: 510338\n",
      "Saved: ./results/wp2/main/AZPYJ_esophagus_squamous_epithelium/AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes.pkl\n",
      "['CHLWW_tibial_artery.3_peaks.bed', 'CHLWW_tibial_artery.5_peaks.bed', 'CHLWW_tibial_artery.1_peaks.bed', 'CHLWW_tibial_artery.2_peaks.bed', 'CHLWW_tibial_artery.4_peaks.bed']\n",
      "CHLWW_tibial_artery.3_peaks\n",
      "CHLWW_tibial_artery.3_peaks\n",
      "3\n",
      "Adipocyte_progenitor_cells\n",
      "CHLWW_tibial_artery.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 143239 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 24%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 98%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells\n",
      "Found rules: 202509\n",
      "Saved: ./results/wp2/main/CHLWW_tibial_artery/CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells.pkl\n",
      "CHLWW_tibial_artery.5_peaks\n",
      "Could not find celltype annotation for cluster: 5, set name to na.\n",
      "CHLWW_tibial_artery.5_peaks\n",
      "5\n",
      "na\n",
      "CHLWW_tibial_artery.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 170993 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c5_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 34%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c5_na\n",
      "Found rules: 201815\n",
      "Saved: ./results/wp2/main/CHLWW_tibial_artery/CHLWW_tibial_artery_c5_na.pkl\n",
      "CHLWW_tibial_artery.1_peaks\n",
      "CHLWW_tibial_artery.1_peaks\n",
      "1\n",
      "Fibroblasts\n",
      "CHLWW_tibial_artery.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 247478 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c1_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c1_Fibroblasts\n",
      "Found rules: 266688\n",
      "Saved: ./results/wp2/main/CHLWW_tibial_artery/CHLWW_tibial_artery_c1_Fibroblasts.pkl\n",
      "CHLWW_tibial_artery.2_peaks\n",
      "CHLWW_tibial_artery.2_peaks\n",
      "2\n",
      "Adipocytes\n",
      "CHLWW_tibial_artery.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 430163 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c2_Adipocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c2_Adipocytes\n",
      "Found rules: 358739\n",
      "Saved: ./results/wp2/main/CHLWW_tibial_artery/CHLWW_tibial_artery_c2_Adipocytes.pkl\n",
      "CHLWW_tibial_artery.4_peaks\n",
      "CHLWW_tibial_artery.4_peaks\n",
      "4\n",
      "Fibroblasts\n",
      "CHLWW_tibial_artery.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 481581 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c4_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: CHLWW_tibial_artery_c4_Fibroblasts\n",
      "Found rules: 387757\n",
      "Saved: ./results/wp2/main/CHLWW_tibial_artery/CHLWW_tibial_artery_c4_Fibroblasts.pkl\n",
      "['IOBHK_tibial_artery.2_peaks.bed', 'IOBHK_tibial_artery.1_peaks.bed', 'IOBHK_tibial_artery.4_peaks.bed', 'IOBHK_tibial_artery.6_peaks.bed', 'IOBHK_tibial_artery.5_peaks.bed', 'IOBHK_tibial_artery.3_peaks.bed', 'IOBHK_tibial_artery.7_peaks.bed']\n",
      "IOBHK_tibial_artery.2_peaks\n",
      "IOBHK_tibial_artery.2_peaks\n",
      "2\n",
      "Chondrocytes\n",
      "IOBHK_tibial_artery.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 21%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 331401 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c2_Chondrocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c2_Chondrocytes\n",
      "Found rules: 325833\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c2_Chondrocytes.pkl\n",
      "IOBHK_tibial_artery.1_peaks\n",
      "IOBHK_tibial_artery.1_peaks\n",
      "1\n",
      "Fibroblasts\n",
      "IOBHK_tibial_artery.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 116759 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c1_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 14%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 42%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 84%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c1_Fibroblasts\n",
      "Found rules: 167309\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c1_Fibroblasts.pkl\n",
      "IOBHK_tibial_artery.4_peaks\n",
      "Could not find celltype annotation for cluster: 4, set name to na.\n",
      "IOBHK_tibial_artery.4_peaks\n",
      "4\n",
      "na\n",
      "IOBHK_tibial_artery.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 41%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 200144 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c4_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c4_na\n",
      "Found rules: 231985\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c4_na.pkl\n",
      "IOBHK_tibial_artery.6_peaks\n",
      "IOBHK_tibial_artery.6_peaks\n",
      "6\n",
      "Stromal_cells\n",
      "IOBHK_tibial_artery.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 481428 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c6_Stromal_cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 96%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c6_Stromal_cells\n",
      "Found rules: 381772\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c6_Stromal_cells.pkl\n",
      "IOBHK_tibial_artery.5_peaks\n",
      "IOBHK_tibial_artery.5_peaks\n",
      "5\n",
      "Fibroblasts\n",
      "IOBHK_tibial_artery.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 477190 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c5_Fibroblasts\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c5_Fibroblasts\n",
      "Found rules: 376956\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c5_Fibroblasts.pkl\n",
      "IOBHK_tibial_artery.3_peaks\n",
      "IOBHK_tibial_artery.3_peaks\n",
      "3\n",
      "Chondrocytes\n",
      "IOBHK_tibial_artery.3_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 439038 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c3_Chondrocytes\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c3_Chondrocytes\n",
      "Found rules: 356047\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c3_Chondrocytes.pkl\n",
      "IOBHK_tibial_artery.7_peaks\n",
      "Could not find celltype annotation for cluster: 7, set name to na.\n",
      "IOBHK_tibial_artery.7_peaks\n",
      "7\n",
      "na\n",
      "IOBHK_tibial_artery.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 318407 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c7_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: IOBHK_tibial_artery_c7_na\n",
      "Found rules: 306160\n",
      "Saved: ./results/wp2/main/IOBHK_tibial_artery/IOBHK_tibial_artery_c7_na.pkl\n",
      "DONE: Created market basket analysis for each cluster\n"
     ]
    }
   ],
   "source": [
    "# Create a market basket analysis for each cluster of the tissue\n",
    "for tissue in tissues:\n",
    "    path_to_clusters = f\"{path_to_tissues}{tissue}/{cluster_folder_tag}\"\n",
    "    make_mb_for_clusters(path_to_clusters=path_to_clusters, tissue=tissue)\n",
    "\n",
    "print(\"DONE: Created market basket analysis for each cluster\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce9872a4",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c16c5d",
   "metadata": {},
   "source": [
    "### Differential Analysis\n",
    "We use the differential analysis of tfcomb to identifiy the differences of tf-co-occurences between all cluster/celltypes of a tissue. For this we load all market basket analysis (tfcomb-objects) of a tissue (see point 2) into a DiffCombObj. After the differential analysis, we filter the object, so that we hopefully find tf-co-occurences that only occure in a single cluster of that tissue or only in a special celltype.  \n",
    "\n",
    "- mb = market basket analysis (CombObj of tf comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2f2b5cf7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQYCP_lower_leg_skin', 'ACCPU_upper_lobe_of_left_lung', 'A8CPH_esophagus_muscularis_mucosa', 'A8WNZ_right_lobe_of_liver', 'ACCQ1_colon_transverse', 'AZPYJ_esophagus_squamous_epithelium', 'CHLWW_tibial_artery', 'IOBHK_tibial_artery']\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "# get the tissue names by folder names of the market basket analysis \n",
    "mb_tissues = get_folder_names_in_folder(rel_folder_path=main_analysis_path)\n",
    "print(mb_tissues)\n",
    "print(len(mb_tissues))\n",
    "# Probably you have to remove the .ipynb_checkpoints file in the main folder, before you continue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "929be533",
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_differential_analysis_for_tissues(tissues=[]):\n",
    "    '''\n",
    "        Differential analysis between all clusters/celltypes of a tissue.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        tissues: array\n",
    "            tissue names by market basket analysis.\n",
    "        ---\n",
    "        DiffCombObj are saved as .pkl to differential_analysis_path\n",
    "    '''\n",
    "    for tissue_folder in tissues:\n",
    "        \n",
    "        diff_save_path = f\"{differential_analysis_path}{tissue_folder}/\"\n",
    "        # Check if  folder for differential_analysis already exists for tissue, if not create new one\n",
    "        if not os.path.exists(diff_save_path):\n",
    "             pathlib.Path(diff_save_path).mkdir(parents=True, exist_ok=True)\n",
    "        \n",
    "        # get file names of the market basket analysis\n",
    "        tissue_mb_files = read_in_file_names_of_folder(rel_path=f\"{main_analysis_path}{tissue_folder}/\")\n",
    "        \n",
    "        # holds the combobj´s\n",
    "        tissue_mbs_to_compare = []\n",
    "        for file in tissue_mb_files:\n",
    "            print(file)\n",
    "            file_name = file.split('.pkl')[0]\n",
    "            # Load the CombObj (market basket analysis) for each cluster of a tissue\n",
    "            obj = CombObj().from_pickle(f\"{main_analysis_path}{tissue_folder}/{file}\")\n",
    "            obj.set_prefix(file_name)\n",
    "            tissue_mbs_to_compare.append(obj)\n",
    "        \n",
    "        # Create DiffCombObj with all Combobj (market basket analysis) of the clusters in a tissue\n",
    "        compare_obj = DiffCombObj(tissue_mbs_to_compare, measure=\"cosine\", join=\"outer\", fillna=True)\n",
    "        # save diffcombj\n",
    "        compare_obj.to_pickle(f'{diff_save_path}{tissue_folder}.pkl')\n",
    "        # Normalize the DiffCombObj\n",
    "        compare_obj.normalize()\n",
    "        compare_obj.calculate_foldchanges()\n",
    "        \n",
    "        # Remove rules which are doubled, e.g. A-B, B-A; (B-A) is removed\n",
    "        compare_obj.simplify_rules()\n",
    "        # Save the normalized, foldchange calculated and simplified diff_comb_obj to .pkl\n",
    "        compare_obj.to_pickle(f'{diff_save_path}{tissue_folder}_normalized.pkl')\n",
    "        print(f\"Done: Diff analysis for tissue {tissue_folder}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "42f520dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQYCP_lower_leg_skin_c4_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin_c5_na.pkl\n",
      "IQYCP_lower_leg_skin_c1_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin_c2_Adipocytes.pkl\n",
      "IQYCP_lower_leg_skin_c3_Sebocytes.pkl\n",
      "IQYCP_lower_leg_skin_c6_Melanocytes.pkl\n",
      "IQYCP_lower_leg_skin_c10_na.pkl\n",
      "IQYCP_lower_leg_skin_c9_na.pkl\n",
      "IQYCP_lower_leg_skin_c7_Melanocytes.pkl\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c5_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c1_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c4_Melanocytes / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c1_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c5_na / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c1_Melanocytes / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c2_Adipocytes / IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c2_Adipocytes / IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c2_Adipocytes / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c2_Adipocytes / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c2_Adipocytes / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c3_Sebocytes / IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c3_Sebocytes / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c3_Sebocytes / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c3_Sebocytes / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c6_Melanocytes / IQYCP_lower_leg_skin_c10_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c6_Melanocytes / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c6_Melanocytes / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c10_na / IQYCP_lower_leg_skin_c9_na\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c10_na / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin_c9_na / IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue IQYCP_lower_leg_skin\n",
      "ACCPU_upper_lobe_of_left_lung_c3_na.pkl\n",
      "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells.pkl\n",
      "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes.pkl\n",
      "INFO: Calculating foldchange for contrast: ACCPU_upper_lobe_of_left_lung_c3_na / ACCPU_upper_lobe_of_left_lung_c1_Clara_cells\n",
      "INFO: Calculating foldchange for contrast: ACCPU_upper_lobe_of_left_lung_c3_na / ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: ACCPU_upper_lobe_of_left_lung_c1_Clara_cells / ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue ACCPU_upper_lobe_of_left_lung\n",
      "A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells.pkl\n",
      "A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts.pkl\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells / A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells / A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue A8CPH_esophagus_muscularis_mucosa\n",
      "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver_c2_Kupffer_cells.pkl\n",
      "A8WNZ_right_lobe_of_liver_c1_Adipocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver_c3_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver_c6_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells.pkl\n",
      "A8WNZ_right_lobe_of_liver_c4_Hepatocytes.pkl\n",
      "A8WNZ_right_lobe_of_liver_c7_Hepatocytes.pkl\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c2_Kupffer_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c1_Adipocytes\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c1_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c1_Adipocytes / A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c1_Adipocytes / A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c1_Adipocytes / A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c1_Adipocytes / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c1_Adipocytes / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c3_Hepatocytes / A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c3_Hepatocytes / A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c3_Hepatocytes / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c3_Hepatocytes / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c6_Hepatocytes / A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c6_Hepatocytes / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c6_Hepatocytes / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells / A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: Calculating foldchange for contrast: A8WNZ_right_lobe_of_liver_c4_Hepatocytes / A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue A8WNZ_right_lobe_of_liver\n",
      "ACCQ1_colon_transverse_c5_Microfold_cells.pkl\n",
      "ACCQ1_colon_transverse_c3_Enterocytes.pkl\n",
      "ACCQ1_colon_transverse_c2_Goblet_cells.pkl\n",
      "ACCQ1_colon_transverse_c6_Gastric_chief_cells.pkl\n",
      "ACCQ1_colon_transverse_c1_Gastric_chief_cells.pkl\n",
      "ACCQ1_colon_transverse_c4_Paneth_cells.pkl\n",
      "ACCQ1_colon_transverse_c7_Fibroblasts.pkl\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c3_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c2_Goblet_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c5_Microfold_cells / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c3_Enterocytes / ACCQ1_colon_transverse_c2_Goblet_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c3_Enterocytes / ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c3_Enterocytes / ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c3_Enterocytes / ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c3_Enterocytes / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c2_Goblet_cells / ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c2_Goblet_cells / ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c2_Goblet_cells / ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c2_Goblet_cells / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c6_Gastric_chief_cells / ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c6_Gastric_chief_cells / ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c6_Gastric_chief_cells / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c1_Gastric_chief_cells / ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c1_Gastric_chief_cells / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: ACCQ1_colon_transverse_c4_Paneth_cells / ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue ACCQ1_colon_transverse\n",
      "AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c2_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c3_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c6_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c7_na.pkl\n",
      "AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes.pkl\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c2_na / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c3_na / AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c3_na / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c3_na / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c3_na / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c3_na / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c6_na / AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c6_na / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c6_na / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c6_na / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes / AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes / AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: Calculating foldchange for contrast: AZPYJ_esophagus_squamous_epithelium_c7_na / AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue AZPYJ_esophagus_squamous_epithelium\n",
      "CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells.pkl\n",
      "CHLWW_tibial_artery_c5_na.pkl\n",
      "CHLWW_tibial_artery_c1_Fibroblasts.pkl\n",
      "CHLWW_tibial_artery_c2_Adipocytes.pkl\n",
      "CHLWW_tibial_artery_c4_Fibroblasts.pkl\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells / CHLWW_tibial_artery_c5_na\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells / CHLWW_tibial_artery_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells / CHLWW_tibial_artery_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells / CHLWW_tibial_artery_c4_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c5_na / CHLWW_tibial_artery_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c5_na / CHLWW_tibial_artery_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c5_na / CHLWW_tibial_artery_c4_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c1_Fibroblasts / CHLWW_tibial_artery_c2_Adipocytes\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c1_Fibroblasts / CHLWW_tibial_artery_c4_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: CHLWW_tibial_artery_c2_Adipocytes / CHLWW_tibial_artery_c4_Fibroblasts\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue CHLWW_tibial_artery\n",
      "IOBHK_tibial_artery_c2_Chondrocytes.pkl\n",
      "IOBHK_tibial_artery_c1_Fibroblasts.pkl\n",
      "IOBHK_tibial_artery_c4_na.pkl\n",
      "IOBHK_tibial_artery_c6_Stromal_cells.pkl\n",
      "IOBHK_tibial_artery_c5_Fibroblasts.pkl\n",
      "IOBHK_tibial_artery_c3_Chondrocytes.pkl\n",
      "IOBHK_tibial_artery_c7_na.pkl\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c1_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c4_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c6_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c5_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c3_Chondrocytes\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c2_Chondrocytes / IOBHK_tibial_artery_c7_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c1_Fibroblasts / IOBHK_tibial_artery_c4_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c1_Fibroblasts / IOBHK_tibial_artery_c6_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c1_Fibroblasts / IOBHK_tibial_artery_c5_Fibroblasts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c1_Fibroblasts / IOBHK_tibial_artery_c3_Chondrocytes\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c1_Fibroblasts / IOBHK_tibial_artery_c7_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c4_na / IOBHK_tibial_artery_c6_Stromal_cells\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c4_na / IOBHK_tibial_artery_c5_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c4_na / IOBHK_tibial_artery_c3_Chondrocytes\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c4_na / IOBHK_tibial_artery_c7_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c6_Stromal_cells / IOBHK_tibial_artery_c5_Fibroblasts\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c6_Stromal_cells / IOBHK_tibial_artery_c3_Chondrocytes\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c6_Stromal_cells / IOBHK_tibial_artery_c7_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c5_Fibroblasts / IOBHK_tibial_artery_c3_Chondrocytes\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c5_Fibroblasts / IOBHK_tibial_artery_c7_na\n",
      "INFO: Calculating foldchange for contrast: IOBHK_tibial_artery_c3_Chondrocytes / IOBHK_tibial_artery_c7_na\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue IOBHK_tibial_artery\n"
     ]
    }
   ],
   "source": [
    "# Do the differential analysis for all clusters in a tissue \n",
    "do_differential_analysis_for_tissues(tissues=mb_tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42d4d9b5",
   "metadata": {},
   "source": [
    "# 4. Specific analysis  :\n",
    "### Biological Question:\n",
    " Find Transcriptionfactor-co-occurences, which only occure in one (or more) \"cluster/celltype of a tissue\".\n",
    " - Solution with a hard threshold (a, all cluster) or a week threshold (b, celltypes in a tissue):\n",
    "       - a) The found TF-pair occures only in a single cluster of a tissue:\n",
    "            e.g. Cluster1 has TF1-TF1 and the other cluster in the tissue do not have TF1-TF1.\n",
    "       - b) A TF-pair occures significantly only in some clusters:\n",
    "            e.g. We have 3 endothelcells cluster in a tissue and this three clusters share a TF-pair\n",
    "            which does not occure in the other cluster of the tissue. (need celltype annotation)\n",
    "          \n",
    "###  Steps:\n",
    "1. Read in tfcomb-DiffObj´s of the Differential analysis (Point 3)\n",
    "2. Get the rules of the DiffCombObj\n",
    "3. Filter the DiffCombObj rules for each cluster in it:\n",
    "       - e.g. investigate cluster1: get all log2changes col,\n",
    "         where cluster1 is associated with and his cosine column\n",
    "       - than filter the rules (tf-pairs) for tf-pairs with a significant change in the log2changes\n",
    "       - We would like to find TF-pairs that show a significant change in the log2changes in each contrast,\n",
    "         because this indicates a TF-Pair, which is striking for the investigated cluster in that tissue.\n",
    "         And a high cosine would be a indication, that the found tf-pair is really \n",
    "         a tf-co-occurence in that cluster.\n",
    "       - To find a TF-pair, which only occures in a single cluster of a tissue,\n",
    "         we check the log2changes in each col of the significant tf-pairs, if they have the same value,\n",
    "         the tf-pair only occures in that cluster in the tissue. \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "505cb1db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQYCP_lower_leg_skin', 'ACCPU_upper_lobe_of_left_lung', 'A8CPH_esophagus_muscularis_mucosa', 'A8WNZ_right_lobe_of_liver', 'ACCQ1_colon_transverse', 'AZPYJ_esophagus_squamous_epithelium', 'CHLWW_tibial_artery', 'IOBHK_tibial_artery']\n"
     ]
    }
   ],
   "source": [
    "# Get the folder names of the tissues that have a differential analysis\n",
    "diff_mb_tissues = get_folder_names_in_folder(rel_folder_path=differential_analysis_path)\n",
    "print(diff_mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d56ce5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_significant_rules(df:pd.DataFrame, cosine_col:str, cosine_threshold=0.001, log2fc_threshold_percent=0.05):\n",
    "    '''\n",
    "    Filter a rules Dataframe (tf-pairs) by log2foldchange col´s and cosine col´s of a DiffCombObj for significant tf-co-occurences (tf-pairs).\n",
    "    Only the TF-pairs with significant log2foldchanges per contrast are kept in the returned Dataframe.\n",
    "    The same filtering is done for the cosine col of the investigated cluster.\n",
    "    \n",
    "    For Filtering: for each col (foldchanges/ cosine) the thresholds to get significant values in that col\n",
    "    are calculated by tfcomb.utils.get_threshold. All rows (tf-pairs), which are not in the threshold \n",
    "    (e.g. >95% percentile for log2foldchange or 99.9% percentile) are removed from the datframe (df). \n",
    "    Threshold are calculated on the original dataframe (df)\n",
    "    Filtering/Reduction of the results is done on a copy of the original dataframe.\n",
    "    \n",
    "    This is done step by step:\n",
    "    e.g. 1. thresholds for column \"log2change_cluster1/cluster2\" are calculated on the original df.\n",
    "         2. all rows (tf-pairs) of the dataframe,\n",
    "            where the values of the investigated column (log2change_cluster1/cluster2) are not in the thresholds,\n",
    "            are removed from the copy of the original dataframe.\n",
    "        These steps are done for each column.\n",
    "\n",
    "    ---\n",
    "    Parameter:\n",
    "    df: pd.DataFrame\n",
    "        DiffCombObj Rules\n",
    "        e.g.:\n",
    "        index   cosine_cluster1 log2change_cluster1/cluster2 log2change_cluster1/cluster3 log2change_cluster1/cluster4\n",
    "        TF1-TF2  0.20                        3.4                              1                        4\n",
    "        TF3-TF4  0.30                       - 0.3                             2                        1\n",
    "        TFX-TFY  0.15                        8                                3                        -4\n",
    "        ...\n",
    "    \n",
    "    cosine_col: string\n",
    "        Name of the cluster, which is investigated and name of the cosine col\n",
    "    \n",
    "    cosine_threshold: float,\n",
    "        Threshold value for the cosine col. All values of the cosine col are investigated\n",
    "        and the threshold value for filtering is calculated. \n",
    "        \n",
    "    log2fc_threshold_percent: float\n",
    "        Threshold value for the log2foldchanges col. All values of the log2foldchanges cols are investigated\n",
    "        and the threshold value for filtering is calculated.\n",
    "    ---\n",
    "    Return pd.Datframe, Filtered df         \n",
    "\n",
    "    '''\n",
    "    # copy of the original dataframe, which will be reduced (remove rows (TF-Pairs),  step by step for each col)\n",
    "    reduced_df = df.copy(deep=True)\n",
    "    for col in df.columns:\n",
    "        # calculate thresholds for log2fc columns or cosine_column\n",
    "        \n",
    "        if col == cosine_col:\n",
    "            # for the cosine column we only want to keep rows with a high value (no negative values),\n",
    "            # so only the upper threshold is important\n",
    "            measure_threshold = utils.get_threshold(df[col], \"both\", percent=cosine_threshold)\n",
    "            upper_threshold = measure_threshold[1]\n",
    "            \n",
    "            # removes tf-pairs (rows), which are smaller than the upper_threshold form the df.\n",
    "            reduced_df = reduced_df[(reduced_df[col] > upper_threshold)]\n",
    "            print(upper_threshold)\n",
    "        else:\n",
    "            # calculates the thresholds for the log2foldchange cols,\n",
    "            # positive and negative is possible, so both thresholds are needed.\n",
    "            measure_threshold = utils.get_threshold(df[col], \"both\", percent=log2fc_threshold_percent)\n",
    "            upper_threshold = measure_threshold[1]\n",
    "            lower_threshold = measure_threshold[0]\n",
    "             # removes tf-pairs (rows), which are smalleror higher than the thresholds form the df.\n",
    "            reduced_df = reduced_df[(reduced_df[col] > upper_threshold) | (reduced_df[col] < lower_threshold)]\n",
    "    # returns only tf-pairs, which are significant        \n",
    "    return reduced_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "540bc39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_specific_tf_cos_for_cluster(df:pd.DataFrame, cluster_name:str) -> pd.DataFrame:\n",
    "    '''\n",
    "        Find tf-co-occurences for a cluster/celltype (cluster_name), that are specific for that cluster\n",
    "        in the associated tissue.\n",
    "        \n",
    "        ---\n",
    "        Parameters:\n",
    "            df:pd.DataFrame\n",
    "                 rules of a DiffCombObj. (Differential analysis)\n",
    "\n",
    "            cluster_name: string\n",
    "                name of the cluster/celltype, that is investigated\n",
    "        ---\n",
    "        Return pd.Dataframe with tf-pairs, that show a high difference (log2foldchange) to the other clusters in the the tissue\n",
    "        and are significantly occureing (cosine) in the investigated cluster (cluster_name).   \n",
    "        \n",
    "    '''\n",
    "    # Get only the columns associated with the investigated cluster of the differential analysis,\n",
    "    # that belongs to the tissue of the cluster.\n",
    "    # reduce dataframe(df) to relevant columns assocaiated with the cluster\n",
    "    cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "    # print(cluster_cols)\n",
    "    relevant_cluster_cols = []\n",
    "    # Add cosine value for the investigated cluster\n",
    "    cluster_cosine_col_name = f\"{cluster_name}_cosine\"\n",
    "    relevant_cluster_cols.append(cluster_cosine_col_name) \n",
    "    for entry in cluster_cols:\n",
    "        if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "            relevant_cluster_cols.append(entry)\n",
    "   # print(logfc_cluster_cols)\n",
    "    print(len(relevant_cluster_cols))\n",
    "    \n",
    "    # Get only the values (cosine of cluster + log2fc with each contrast of to the cluster) for the cluster to investigate .\n",
    "    reduced_df = df[relevant_cluster_cols]\n",
    "    print(f'Initial TF-pairs-Count: {reduced_df.shape}')\n",
    "\n",
    "   # Filter out rows with 0.00   \n",
    "    # Count all entries in a row , which do not have a zero(0.00) in it.\n",
    "    # e.g. 15 cols have 0.00 => val_counts = 0, 10 cols not have a 0.00 => val_counts = 10 \n",
    "    val_counts = reduced_df[~reduced_df.isin([0])].count(axis=1).sort_values()\n",
    "    \n",
    "    # Set threshold \n",
    "    selection_threshold = len(relevant_cluster_cols) # e.g. 15, could be varyied\n",
    "    # Keep all entries, which have more/same values that are higher than the threshold\n",
    "    tfs_occ = val_counts[val_counts >= selection_threshold].index\n",
    "    result = reduced_df.loc[tfs_occ]\n",
    "    print(f'Zero filtered TF-pairs-Count: {result.shape}')\n",
    "    \n",
    "    # We would like to find TF-pairs that show a significant change in the log2changes in each contrast, because\n",
    "    # this indicates a TF-Pair, which is striking for the investigated cluster in that tissue.\n",
    "    # And a high cosine would be a indication, that the found tf-pair is really a tf-co-occurence in that cluster.\n",
    "    # Filtering: log2changes and cosine cols to get tf-pairs that have a log2change,\n",
    "    # that shows a high difference to all other clusters in the tissue and a significant cosine value.\n",
    "    significants = get_significant_rules(df=result, cosine_col=cluster_cosine_col_name, cosine_threshold=0.001, log2fc_threshold_percent=0.1)\n",
    "    \n",
    "    print(f'Cluster: {cluster_name}: {significants.shape} ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue')\n",
    "    print(f'Done: Find specific tf co-occurences for cluster{cluster_name}.')\n",
    "    \n",
    "    return significants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a5ce5809",
   "metadata": {},
   "outputs": [],
   "source": [
    "def investigate_differential_analysis_per_tissue(tissues=[]):\n",
    "    '''\n",
    "    Investigate the differential analysis of tf-comb for each tissue.\n",
    "    ---\n",
    "    Parameters:\n",
    "    tissues: []\n",
    "    Tissues that have differential analysis.\n",
    "    ---\n",
    "    Exception:\n",
    "        Catch exception for cluster analysis and continues with analysing the next cluster.\n",
    "    \n",
    "    '''\n",
    "    for tissue in tissues:\n",
    "        print(f\"Start analyse cluster for {tissue}:\")\n",
    "        answer_save_path = f\"{answers_path}{tissue}/\"\n",
    "        # Check if  folder for path already exists for tissue, if not create new one\n",
    "        if not os.path.exists(answer_save_path):\n",
    "             pathlib.Path(answer_save_path).mkdir(parents=True, exist_ok=True)\n",
    "        \n",
    "        # folder errors\n",
    "        error_path = f\"{answers_path}{tissue}/errors/\"\n",
    "        # Check if  folder for path already exists for tissue, if not create new one\n",
    "        if not os.path.exists(error_path):\n",
    "             pathlib.Path(error_path).mkdir(parents=True, exist_ok=True)\n",
    "        \n",
    "        # load the diffcombobj for the tissue \n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_path}{tissue}/{tissue}_normalized.pkl\")\n",
    "        # get the rules (tf-pairs) of the diffcombobj.\n",
    "        diff_rules = diff_obj.rules\n",
    "        \n",
    "        # read in the filenames of the original tfcomobj with the executed market basket analysis. \n",
    "        files_main_mb = read_in_file_names_of_folder(rel_path=f\"{main_analysis_path}{tissue}\")\n",
    "        \n",
    "        # Iterate over each cluster.\n",
    "        for i, file in enumerate(files_main_mb):\n",
    "            # print(file)\n",
    "            cluster_name = file.split('.pkl')[0]\n",
    "            print(f\"Start Find specific tf-cos for: {cluster_name}\")\n",
    "            try:\n",
    "                # res contains a pd.dataframe, that contains tf-pairs, \n",
    "                # that have significant log2changes/ and a significant cosine value for the investigated cluster\n",
    "                # compared to the other clusters in the tissue. This TF-pairs could be interesting for further investigation\n",
    "                # ,because they show a significant difference to all the other clusters in the tissue.\n",
    "                res = find_specific_tf_cos_for_cluster(df=diff_rules, cluster_name=cluster_name)\n",
    "\n",
    "                # save res as .pkl\n",
    "                res.to_pickle(f\"{answer_save_path}{cluster_name}.pkl\")\n",
    "            except Exception as err:\n",
    "                print(f\"ERROR: Could not proceed with analysis for cluster: {cluster_name}\")\n",
    "                print(f\"{err}\")\n",
    "                df = pd.DataFrame()\n",
    "                # just save a csv with the clustername as file name,so\n",
    "                # that we later know that an error happened for this cluster.\n",
    "                df.to_csv(f\"{error_path}{cluster_name}.csv\")\n",
    "                print(f\"Continue with next cluster.\")\n",
    "                continue\n",
    "    print(\"Done investigating diff analysis!\")\n",
    "\n",
    "           \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "378e96c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start analyse cluster for IQYCP_lower_leg_skin:\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c4_Melanocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (71760, 9)\n",
      "0.4137660819725677\n",
      "Cluster: IQYCP_lower_leg_skin_c4_Melanocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c4_Melanocytes.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c5_na\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (26184, 9)\n",
      "0.3500593599702473\n",
      "Cluster: IQYCP_lower_leg_skin_c5_na: (4, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c5_na.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c1_Melanocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (79992, 9)\n",
      "0.35484463798586696\n",
      "Cluster: IQYCP_lower_leg_skin_c1_Melanocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c1_Melanocytes.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c2_Adipocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (133436, 9)\n",
      "0.28939693379465786\n",
      "Cluster: IQYCP_lower_leg_skin_c2_Adipocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c2_Adipocytes.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c3_Sebocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (139321, 9)\n",
      "0.25140401291662456\n",
      "Cluster: IQYCP_lower_leg_skin_c3_Sebocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c3_Sebocytes.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c6_Melanocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (174565, 9)\n",
      "0.38912001136481783\n",
      "Cluster: IQYCP_lower_leg_skin_c6_Melanocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c6_Melanocytes.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c10_na\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (42659, 9)\n",
      "0.3684032505599751\n",
      "Cluster: IQYCP_lower_leg_skin_c10_na: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c10_na.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c9_na\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (36496, 9)\n",
      "0.3815773507688856\n",
      "Cluster: IQYCP_lower_leg_skin_c9_na: (16, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c9_na.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin_c7_Melanocytes\n",
      "9\n",
      "Initial TF-pairs-Count: (212263, 9)\n",
      "Zero filtered TF-pairs-Count: (151267, 9)\n",
      "0.26998748426185465\n",
      "Cluster: IQYCP_lower_leg_skin_c7_Melanocytes: (0, 9) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin_c7_Melanocytes.\n",
      "Start analyse cluster for ACCPU_upper_lobe_of_left_lung:\n",
      "Start Find specific tf-cos for: ACCPU_upper_lobe_of_left_lung_c3_na\n",
      "3\n",
      "Initial TF-pairs-Count: (49811, 3)\n",
      "Zero filtered TF-pairs-Count: (8829, 3)\n",
      "0.9377810822110996\n",
      "Cluster: ACCPU_upper_lobe_of_left_lung_c3_na: (0, 3) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCPU_upper_lobe_of_left_lung_c3_na.\n",
      "Start Find specific tf-cos for: ACCPU_upper_lobe_of_left_lung_c1_Clara_cells\n",
      "3\n",
      "Initial TF-pairs-Count: (49811, 3)\n",
      "Zero filtered TF-pairs-Count: (8166, 3)\n",
      "1.1512493394794545\n",
      "Cluster: ACCPU_upper_lobe_of_left_lung_c1_Clara_cells: (0, 3) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCPU_upper_lobe_of_left_lung_c1_Clara_cells.\n",
      "Start Find specific tf-cos for: ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
      "3\n",
      "Initial TF-pairs-Count: (49811, 3)\n",
      "Zero filtered TF-pairs-Count: (46791, 3)\n",
      "0.48294908916480966\n",
      "Cluster: ACCPU_upper_lobe_of_left_lung_c2_Adipocytes: (75, 3) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCPU_upper_lobe_of_left_lung_c2_Adipocytes.\n",
      "Start analyse cluster for A8CPH_esophagus_muscularis_mucosa:\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (150024, 12)\n",
      "0.09636973146820096\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c4_Enterocytes.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (103804, 12)\n",
      "0.13256495181636324\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes: (1, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c6_Enterocytes.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (132341, 12)\n",
      "0.1147544422917076\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c2_Enterocytes.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (190831, 12)\n",
      "0.0609428669226853\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (226579, 12)\n",
      "0.05437624641903524\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts: (10, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (204489, 12)\n",
      "0.06020949531640419\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (141114, 12)\n",
      "0.11157431291626774\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (236765, 12)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.05637225814216247\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (221257, 12)\n",
      "0.059219688101427595\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (206566, 12)\n",
      "0.11523411897659146\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (244645, 12)\n",
      "0.05885487444436921\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells.\n",
      "Start Find specific tf-cos for: A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts\n",
      "12\n",
      "Initial TF-pairs-Count: (261034, 12)\n",
      "Zero filtered TF-pairs-Count: (208622, 12)\n",
      "0.11396607911004414\n",
      "Cluster: A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts.\n",
      "Start analyse cluster for A8WNZ_right_lobe_of_liver:\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (73423, 8)\n",
      "0.13241261942897087\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes: (12, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c5_Cholangiocytes.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (104482, 8)\n",
      "0.1438808652196082\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c2_Kupffer_cells: (6, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c2_Kupffer_cells.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c1_Adipocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (133391, 8)\n",
      "0.14913701444863672\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c1_Adipocytes: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c1_Adipocytes.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c3_Hepatocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (190270, 8)\n",
      "0.13684457285923485\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c3_Hepatocytes: (5, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c3_Hepatocytes.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c6_Hepatocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (193737, 8)\n",
      "0.1244937888131607\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c6_Hepatocytes: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c6_Hepatocytes.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (177682, 8)\n",
      "0.16770781423593312\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells: (8, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c4_Hepatocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (235179, 8)\n",
      "0.0583026663858146\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c4_Hepatocytes: (12, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c4_Hepatocytes.\n",
      "Start Find specific tf-cos for: A8WNZ_right_lobe_of_liver_c7_Hepatocytes\n",
      "8\n",
      "Initial TF-pairs-Count: (249015, 8)\n",
      "Zero filtered TF-pairs-Count: (217282, 8)\n",
      "0.06547750992162102\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c7_Hepatocytes: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterA8WNZ_right_lobe_of_liver_c7_Hepatocytes.\n",
      "Start analyse cluster for ACCQ1_colon_transverse:\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c5_Microfold_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (44311, 7)\n",
      "0.1950691412503733\n",
      "Cluster: ACCQ1_colon_transverse_c5_Microfold_cells: (30, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c5_Microfold_cells.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c3_Enterocytes\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (89330, 7)\n",
      "0.19250487283161122\n",
      "Cluster: ACCQ1_colon_transverse_c3_Enterocytes: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c3_Enterocytes.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c2_Goblet_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (127984, 7)\n",
      "0.21782298447982312\n",
      "Cluster: ACCQ1_colon_transverse_c2_Goblet_cells: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c2_Goblet_cells.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c6_Gastric_chief_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (161817, 7)\n",
      "0.19768820194000064\n",
      "Cluster: ACCQ1_colon_transverse_c6_Gastric_chief_cells: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c6_Gastric_chief_cells.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c1_Gastric_chief_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (153021, 7)\n",
      "0.08256764409150376\n",
      "Cluster: ACCQ1_colon_transverse_c1_Gastric_chief_cells: (16, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c1_Gastric_chief_cells.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c4_Paneth_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (197349, 7)\n",
      "0.08456551676834458\n",
      "Cluster: ACCQ1_colon_transverse_c4_Paneth_cells: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c4_Paneth_cells.\n",
      "Start Find specific tf-cos for: ACCQ1_colon_transverse_c7_Fibroblasts\n",
      "7\n",
      "Initial TF-pairs-Count: (223653, 7)\n",
      "Zero filtered TF-pairs-Count: (102652, 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.17620211400995242\n",
      "Cluster: ACCQ1_colon_transverse_c7_Fibroblasts: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterACCQ1_colon_transverse_c7_Fibroblasts.\n",
      "Start analyse cluster for AZPYJ_esophagus_squamous_epithelium:\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (173642, 10)\n",
      "0.08172481165040126\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts: (3, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (188327, 10)\n",
      "0.07574361930091497\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts: (23, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (193859, 10)\n",
      "0.06898960319320992\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts: (1, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c2_na\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (244012, 10)\n",
      "0.09912360846608083\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c2_na: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c2_na.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c3_na\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (250621, 10)\n",
      "0.08548038039151486\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c3_na: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c3_na.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c6_na\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (253206, 10)\n",
      "0.08703245037425351\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c6_na: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c6_na.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (253372, 10)\n",
      "0.08287699147901147\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c1_Enterocytes.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (167421, 10)\n",
      "0.08890237727229841\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes: (2, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c8_Adipocytes.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c7_na\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (244965, 10)\n",
      "0.08893643062100007\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c7_na: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c7_na.\n",
      "Start Find specific tf-cos for: AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes\n",
      "10\n",
      "Initial TF-pairs-Count: (265726, 10)\n",
      "Zero filtered TF-pairs-Count: (255454, 10)\n",
      "0.08497233925943676\n",
      "Cluster: AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterAZPYJ_esophagus_squamous_epithelium_c9_Enterocytes.\n",
      "Start analyse cluster for CHLWW_tibial_artery:\n",
      "Start Find specific tf-cos for: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells\n",
      "5\n",
      "Initial TF-pairs-Count: (216105, 5)\n",
      "Zero filtered TF-pairs-Count: (101392, 5)\n",
      "0.09889247188536864\n",
      "Cluster: CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells: (0, 5) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCHLWW_tibial_artery_c3_Adipocyte_progenitor_cells.\n",
      "Start Find specific tf-cos for: CHLWW_tibial_artery_c5_na\n",
      "5\n",
      "Initial TF-pairs-Count: (216105, 5)\n",
      "Zero filtered TF-pairs-Count: (101054, 5)\n",
      "0.11027379305966983\n",
      "Cluster: CHLWW_tibial_artery_c5_na: (0, 5) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCHLWW_tibial_artery_c5_na.\n",
      "Start Find specific tf-cos for: CHLWW_tibial_artery_c1_Fibroblasts\n",
      "5\n",
      "Initial TF-pairs-Count: (216105, 5)\n",
      "Zero filtered TF-pairs-Count: (133523, 5)\n",
      "0.07217904219776763\n",
      "Cluster: CHLWW_tibial_artery_c1_Fibroblasts: (0, 5) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCHLWW_tibial_artery_c1_Fibroblasts.\n",
      "Start Find specific tf-cos for: CHLWW_tibial_artery_c2_Adipocytes\n",
      "5\n",
      "Initial TF-pairs-Count: (216105, 5)\n",
      "Zero filtered TF-pairs-Count: (179577, 5)\n",
      "0.2552160658440272\n",
      "Cluster: CHLWW_tibial_artery_c2_Adipocytes: (0, 5) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCHLWW_tibial_artery_c2_Adipocytes.\n",
      "Start Find specific tf-cos for: CHLWW_tibial_artery_c4_Fibroblasts\n",
      "5\n",
      "Initial TF-pairs-Count: (216105, 5)\n",
      "Zero filtered TF-pairs-Count: (194108, 5)\n",
      "0.21757420087389617\n",
      "Cluster: CHLWW_tibial_artery_c4_Fibroblasts: (0, 5) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCHLWW_tibial_artery_c4_Fibroblasts.\n",
      "Start analyse cluster for IOBHK_tibial_artery:\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c2_Chondrocytes\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (163112, 7)\n",
      "0.12716229872343668\n",
      "Cluster: IOBHK_tibial_artery_c2_Chondrocytes: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c2_Chondrocytes.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c1_Fibroblasts\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (83785, 7)\n",
      "0.13446771321567405\n",
      "Cluster: IOBHK_tibial_artery_c1_Fibroblasts: (3, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c1_Fibroblasts.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c4_na\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (116135, 7)\n",
      "0.138032994876152\n",
      "Cluster: IOBHK_tibial_artery_c4_na: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c4_na.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c6_Stromal_cells\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (191075, 7)\n",
      "0.11929909072984715\n",
      "Cluster: IOBHK_tibial_artery_c6_Stromal_cells: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c6_Stromal_cells.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c5_Fibroblasts\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (188661, 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.06412541103059613\n",
      "Cluster: IOBHK_tibial_artery_c5_Fibroblasts: (1, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c5_Fibroblasts.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c3_Chondrocytes\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (178223, 7)\n",
      "0.12950172312614228\n",
      "Cluster: IOBHK_tibial_artery_c3_Chondrocytes: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c3_Chondrocytes.\n",
      "Start Find specific tf-cos for: IOBHK_tibial_artery_c7_na\n",
      "7\n",
      "Initial TF-pairs-Count: (223433, 7)\n",
      "Zero filtered TF-pairs-Count: (153241, 7)\n",
      "0.1446196144617764\n",
      "Cluster: IOBHK_tibial_artery_c7_na: (0, 7) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIOBHK_tibial_artery_c7_na.\n",
      "Done investigating diff analysis!\n"
     ]
    }
   ],
   "source": [
    "investigate_differential_analysis_per_tissue(tissues=diff_mb_tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad67ae86",
   "metadata": {},
   "source": [
    "### Answer for: \n",
    "a) The found TF-pair occures only in a single cluster of a tissue:\n",
    "     e.g. Cluster1 has TF1-TF1 and the other cluster in the tissue do not have TF1-TF1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4dc4c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rows_with_same_value(df:pd.DataFrame) -> pd.DataFrame:\n",
    "    '''\n",
    "    Remove all rows of a Dataframe that do not have the same value in each column and\n",
    "    return the reduced dataframe.\n",
    "    '''\n",
    "    return df[df.apply(lambda x: min(abs(x)) == max(abs(x)), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dcde71dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQYCP_lower_leg_skin', 'ACCPU_upper_lobe_of_left_lung', 'A8CPH_esophagus_muscularis_mucosa', 'A8WNZ_right_lobe_of_liver', 'ACCQ1_colon_transverse', 'AZPYJ_esophagus_squamous_epithelium', 'CHLWW_tibial_artery', 'IOBHK_tibial_artery']\n"
     ]
    }
   ],
   "source": [
    "#  Get folder names of the tissues.\n",
    "answers_mb_tissues = get_folder_names_in_folder(rel_folder_path=answers_path)\n",
    "print(answers_mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "85d4c2df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start analysis IQYCP_lower_leg_skin:\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c4_Melanocytes.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c5_na.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c1_Melanocytes.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c2_Adipocytes.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c3_Sebocytes.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c6_Melanocytes.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c10_na.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c9_na.\n",
      "Could not find single occurences in cluster IQYCP_lower_leg_skin_c7_Melanocytes.\n",
      "Start analysis ACCPU_upper_lobe_of_left_lung:\n",
      "Could not find single occurences in cluster ACCPU_upper_lobe_of_left_lung_c3_na.\n",
      "Could not find single occurences in cluster ACCPU_upper_lobe_of_left_lung_c1_Clara_cells.\n",
      "Cluster: ACCPU_upper_lobe_of_left_lung_c2_Adipocytes, wich has tf-pairs that only occure in this cluster for tissue ACCPU_upper_lobe_of_left_lung.Found: 27\n",
      "Could not find single occurences in cluster ACCPU_upper_lobe_of_left_lung_c2_Adipocytes.\n",
      "Start analysis A8CPH_esophagus_muscularis_mucosa:\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c4_Enterocytes.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c6_Enterocytes.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c2_Enterocytes.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c7_Fibroblasts.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c1_Fibroblasts.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c3_Gastric_chief_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c9_Adipocyte_progenitor_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c5_Stromal_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c8_Gastric_chief_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c10_Stromal_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c13_Gastric_chief_cells.\n",
      "Could not find single occurences in cluster A8CPH_esophagus_muscularis_mucosa_c11_Fibroblasts.\n",
      "Start analysis A8WNZ_right_lobe_of_liver:\n",
      "Cluster: A8WNZ_right_lobe_of_liver_c5_Cholangiocytes, wich has tf-pairs that only occure in this cluster for tissue A8WNZ_right_lobe_of_liver.Found: 1\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c5_Cholangiocytes.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c2_Kupffer_cells.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c1_Adipocytes.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c3_Hepatocytes.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c6_Hepatocytes.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c4_Hepatocytes.\n",
      "Could not find single occurences in cluster A8WNZ_right_lobe_of_liver_c7_Hepatocytes.\n",
      "Start analysis ACCQ1_colon_transverse:\n",
      "Cluster: ACCQ1_colon_transverse_c5_Microfold_cells, wich has tf-pairs that only occure in this cluster for tissue ACCQ1_colon_transverse.Found: 9\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c5_Microfold_cells.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c3_Enterocytes.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c2_Goblet_cells.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c6_Gastric_chief_cells.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c1_Gastric_chief_cells.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c4_Paneth_cells.\n",
      "Could not find single occurences in cluster ACCQ1_colon_transverse_c7_Fibroblasts.\n",
      "Start analysis AZPYJ_esophagus_squamous_epithelium:\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c5_Fibroblasts.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c4_Fibroblasts.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c10_Fibroblasts.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c2_na.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c3_na.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c6_na.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c1_Enterocytes.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c8_Adipocytes.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c7_na.\n",
      "Could not find single occurences in cluster AZPYJ_esophagus_squamous_epithelium_c9_Enterocytes.\n",
      "Start analysis CHLWW_tibial_artery:\n",
      "Could not find single occurences in cluster CHLWW_tibial_artery_c3_Adipocyte_progenitor_cells.\n",
      "Could not find single occurences in cluster CHLWW_tibial_artery_c5_na.\n",
      "Could not find single occurences in cluster CHLWW_tibial_artery_c1_Fibroblasts.\n",
      "Could not find single occurences in cluster CHLWW_tibial_artery_c2_Adipocytes.\n",
      "Could not find single occurences in cluster CHLWW_tibial_artery_c4_Fibroblasts.\n",
      "Start analysis IOBHK_tibial_artery:\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c2_Chondrocytes.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c1_Fibroblasts.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c4_na.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c6_Stromal_cells.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c5_Fibroblasts.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c3_Chondrocytes.\n",
      "Could not find single occurences in cluster IOBHK_tibial_artery_c7_na.\n",
      "DONE\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "We check now the TF-pairs, that have significant log2changes/ \n",
    "and a significant cosine value for the investigated clusters.\n",
    "If the log2foldchange for each contrast (compared clusters in the tissue) has the same value,\n",
    "the tf-pair is only occureing in the investigated cluster. \n",
    "So The found TF-pair occures only in a single cluster of that tissue. \n",
    "This tf-pair could be interesting for further investigations.\n",
    "'''\n",
    "cluster_with_specials= []\n",
    "all_cluster_results= []\n",
    "for tissue in answers_mb_tissues:\n",
    "    \n",
    "    print(f\"Start analysis {tissue}:\")\n",
    "    answer_specials_save_path = f\"{answers_path}{tissue}/specials/\"\n",
    "    # Check if  folder for path already exists for tissue, if not create new one\n",
    "    if not os.path.exists(answer_specials_save_path):\n",
    "         pathlib.Path(answer_specials_save_path).mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # file names of pd.dataframes with the significant tf-pairs in it.\n",
    "    files = read_in_file_names_of_folder(rel_path=f\"{answers_path}{tissue}\")\n",
    "    for i, file in enumerate(files):\n",
    "        # print(file)\n",
    "        cluster_name = file.split('.pkl')[0]\n",
    "        \n",
    "        # Read in significant tf-pairs\n",
    "        df = pd.read_pickle(f\"{answers_path}{tissue}/{file}\")\n",
    "        \n",
    "        all_cluster_results.append(df)\n",
    "\n",
    "        # has to remove cosine from dataframe, to get only the log2foldchanges\n",
    "        # for each contrast (compared clusters in the tissue), to check if they have the same value in each column\n",
    "        df_copy = df.copy()\n",
    "        cosine_col_name = df_copy.columns[0]\n",
    "        cosine_col = df_copy[cosine_col_name]\n",
    "        df_only_log2fc = df_copy.drop(columns=[cosine_col_name])\n",
    "        \n",
    "        # find tf-pairs, that have the same value in each log2foldchange column\n",
    "        df_res = find_rows_with_same_value(df_only_log2fc)\n",
    "        \n",
    "        # check if we found a tf-co-occurence which is significant and only occures in one cluster of a tissue\n",
    "        if df_res.shape[0] > 0: \n",
    "            print(f\"Cluster: {cluster_name}, wich has tf-pairs that only occure in this cluster for tissue {tissue}.Found: {str(df_res.shape[0])}\")\n",
    "            #add cosine value back to dataframe\n",
    "            df_res[cosine_col_name]= cosine_col\n",
    "            cluster_with_specials.append(df_res)\n",
    "            # save the tf-pairs as csv, if we found some.\n",
    "            df_res.to_csv(f\"{answer_specials_save_path}{cluster_name}.csv\")\n",
    "        \n",
    "        print(f\"Could not find single occurences in cluster {cluster_name}.\")\n",
    "print(\"DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3cc72ded",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of all investigated clusters: 61\n",
      "Number of cluster, that show significant high/ changes for tf-pairs in comparison to the other cluster of theire tissue.\n",
      "18\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of all investigated clusters: {len(all_cluster_results)}')\n",
    "cluster_with_high_changes = []\n",
    "for res in all_cluster_results:\n",
    "    \n",
    "    if res.shape[0] > 0:\n",
    "        cluster_with_high_changes.append(res)\n",
    "print(f'Number of cluster, that show significant high/ changes for tf-pairs in comparison to the other cluster of theire tissue.')\n",
    "print(len(cluster_with_high_changes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42fc51ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['IQYCP_lower_leg_skin_c5_na_cosine',\n",
      "       'IQYCP_lower_leg_skin_c4_Melanocytes/IQYCP_lower_leg_skin_c5_na_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c1_Melanocytes_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c2_Adipocytes_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c3_Sebocytes_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c6_Melanocytes_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c10_na_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c9_na_cosine_log2fc',\n",
      "       'IQYCP_lower_leg_skin_c5_na/IQYCP_lower_leg_skin_c7_Melanocytes_cosine_log2fc'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "for c in cluster_with_high_changes:\n",
    "    print(c.columns)\n",
    "    break\n",
    "    cluster_names = list(filter(lambda x: f'cosine' in x , c.columns))\n",
    "    print(cluster_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8e48c3d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cluster that have tf-pairs that only occure in the investigated cluster in that tissue. (filter thresholds: cosine=0.001 , log2fc=0.10)\n",
      "3\n",
      "[                ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc  \\\n",
      "PHOX2A-BHLHE22                                          -3.273248                                               \n",
      "PHOX2A-OLIG3                                            -3.273248                                               \n",
      "E2F3-E2F4                                               -3.270808                                               \n",
      "Rhox11-ZBED1                                            -3.266179                                               \n",
      "NEUROG2-PROP1                                           -3.265071                                               \n",
      "OLIG2-PROP1                                             -3.265071                                               \n",
      "Dlx2-EMX1                                               -3.260364                                               \n",
      "Dlx2-EN1                                                -3.260364                                               \n",
      "Dlx2-PAX4                                               -3.260364                                               \n",
      "Dlx2-TLX2                                               -3.260364                                               \n",
      "Dlx2-VSX2                                               -3.260364                                               \n",
      "E2F1-E2F3                                               -3.260364                                               \n",
      "FOXF2-LHX6                                              -3.260364                                               \n",
      "Lhx8-Dlx2                                               -3.260364                                               \n",
      "Six3-MYBL2                                              -3.281143                                               \n",
      "FOXD2-LHX6                                              -3.438723                                               \n",
      "EMX1-Six3                                               -3.428891                                               \n",
      "EN1-Six3                                                -3.428891                                               \n",
      "Lhx8-Six3                                               -3.428891                                               \n",
      "PAX4-Six3                                               -3.428891                                               \n",
      "Six3-TLX2                                               -3.428891                                               \n",
      "Six3-VSX2                                               -3.428891                                               \n",
      "Atoh1-PHOX2A                                            -3.374966                                               \n",
      "EVX1-GSC2                                               -3.293265                                               \n",
      "NEUROG2-PHOX2A                                          -3.291634                                               \n",
      "OLIG2-PHOX2A                                            -3.291634                                               \n",
      "Atoh1-PROP1                                             -3.286949                                               \n",
      "\n",
      "                ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc  \\\n",
      "PHOX2A-BHLHE22                                          -3.273248                                                        \n",
      "PHOX2A-OLIG3                                            -3.273248                                                        \n",
      "E2F3-E2F4                                               -3.270808                                                        \n",
      "Rhox11-ZBED1                                            -3.266179                                                        \n",
      "NEUROG2-PROP1                                           -3.265071                                                        \n",
      "OLIG2-PROP1                                             -3.265071                                                        \n",
      "Dlx2-EMX1                                               -3.260364                                                        \n",
      "Dlx2-EN1                                                -3.260364                                                        \n",
      "Dlx2-PAX4                                               -3.260364                                                        \n",
      "Dlx2-TLX2                                               -3.260364                                                        \n",
      "Dlx2-VSX2                                               -3.260364                                                        \n",
      "E2F1-E2F3                                               -3.260364                                                        \n",
      "FOXF2-LHX6                                              -3.260364                                                        \n",
      "Lhx8-Dlx2                                               -3.260364                                                        \n",
      "Six3-MYBL2                                              -3.281143                                                        \n",
      "FOXD2-LHX6                                              -3.438723                                                        \n",
      "EMX1-Six3                                               -3.428891                                                        \n",
      "EN1-Six3                                                -3.428891                                                        \n",
      "Lhx8-Six3                                               -3.428891                                                        \n",
      "PAX4-Six3                                               -3.428891                                                        \n",
      "Six3-TLX2                                               -3.428891                                                        \n",
      "Six3-VSX2                                               -3.428891                                                        \n",
      "Atoh1-PHOX2A                                            -3.374966                                                        \n",
      "EVX1-GSC2                                               -3.293265                                                        \n",
      "NEUROG2-PHOX2A                                          -3.291634                                                        \n",
      "OLIG2-PHOX2A                                            -3.291634                                                        \n",
      "Atoh1-PROP1                                             -3.286949                                                        \n",
      "\n",
      "                ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine  \n",
      "PHOX2A-BHLHE22                                           0.503637   \n",
      "PHOX2A-OLIG3                                             0.503637   \n",
      "E2F3-E2F4                                                0.502688   \n",
      "Rhox11-ZBED1                                             0.500892   \n",
      "NEUROG2-PROP1                                            0.500463   \n",
      "OLIG2-PROP1                                              0.500463   \n",
      "Dlx2-EMX1                                                0.498643   \n",
      "Dlx2-EN1                                                 0.498643   \n",
      "Dlx2-PAX4                                                0.498643   \n",
      "Dlx2-TLX2                                                0.498643   \n",
      "Dlx2-VSX2                                                0.498643   \n",
      "E2F1-E2F3                                                0.498643   \n",
      "FOXF2-LHX6                                               0.498643   \n",
      "Lhx8-Dlx2                                                0.498643   \n",
      "Six3-MYBL2                                               0.506720   \n",
      "FOXD2-LHX6                                               0.571908   \n",
      "EMX1-Six3                                                0.567630   \n",
      "EN1-Six3                                                 0.567630   \n",
      "Lhx8-Six3                                                0.567630   \n",
      "PAX4-Six3                                                0.567630   \n",
      "Six3-TLX2                                                0.567630   \n",
      "Six3-VSX2                                                0.567630   \n",
      "Atoh1-PHOX2A                                             0.544673   \n",
      "EVX1-GSC2                                                0.511486   \n",
      "NEUROG2-PHOX2A                                           0.510842   \n",
      "OLIG2-PHOX2A                                             0.510842   \n",
      "Atoh1-PROP1                                              0.508998   ,              A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c2_Kupffer_cells_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                      \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c1_Adipocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                   \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c3_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c6_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                                   \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c4_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c7_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes_cosine  \n",
      "BARHL2-CUX2                                           0.173113   ,              ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c3_Enterocytes_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                               \n",
      "MEOX1-VSX2                                            4.533017                                               \n",
      "MEOX1-TLX2                                            4.533017                                               \n",
      "MEOX1-Lhx8                                            4.533017                                               \n",
      "MEOX1-LHX6                                            4.533017                                               \n",
      "MEOX1-EN1                                             4.533017                                               \n",
      "MEOX1-EMX1                                            4.533017                                               \n",
      "RHOXF1-EMX2                                           4.649823                                               \n",
      "PITX2-EMX2                                            4.649823                                               \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c2_Goblet_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                \n",
      "MEOX1-VSX2                                            4.533017                                                \n",
      "MEOX1-TLX2                                            4.533017                                                \n",
      "MEOX1-Lhx8                                            4.533017                                                \n",
      "MEOX1-LHX6                                            4.533017                                                \n",
      "MEOX1-EN1                                             4.533017                                                \n",
      "MEOX1-EMX1                                            4.533017                                                \n",
      "RHOXF1-EMX2                                           4.649823                                                \n",
      "PITX2-EMX2                                            4.649823                                                \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                       \n",
      "MEOX1-VSX2                                            4.533017                                                       \n",
      "MEOX1-TLX2                                            4.533017                                                       \n",
      "MEOX1-Lhx8                                            4.533017                                                       \n",
      "MEOX1-LHX6                                            4.533017                                                       \n",
      "MEOX1-EN1                                             4.533017                                                       \n",
      "MEOX1-EMX1                                            4.533017                                                       \n",
      "RHOXF1-EMX2                                           4.649823                                                       \n",
      "PITX2-EMX2                                            4.649823                                                       \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                       \n",
      "MEOX1-VSX2                                            4.533017                                                       \n",
      "MEOX1-TLX2                                            4.533017                                                       \n",
      "MEOX1-Lhx8                                            4.533017                                                       \n",
      "MEOX1-LHX6                                            4.533017                                                       \n",
      "MEOX1-EN1                                             4.533017                                                       \n",
      "MEOX1-EMX1                                            4.533017                                                       \n",
      "RHOXF1-EMX2                                           4.649823                                                       \n",
      "PITX2-EMX2                                            4.649823                                                       \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                \n",
      "MEOX1-VSX2                                            4.533017                                                \n",
      "MEOX1-TLX2                                            4.533017                                                \n",
      "MEOX1-Lhx8                                            4.533017                                                \n",
      "MEOX1-LHX6                                            4.533017                                                \n",
      "MEOX1-EN1                                             4.533017                                                \n",
      "MEOX1-EMX1                                            4.533017                                                \n",
      "RHOXF1-EMX2                                           4.649823                                                \n",
      "PITX2-EMX2                                            4.649823                                                \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                               \n",
      "MEOX1-VSX2                                            4.533017                                               \n",
      "MEOX1-TLX2                                            4.533017                                               \n",
      "MEOX1-Lhx8                                            4.533017                                               \n",
      "MEOX1-LHX6                                            4.533017                                               \n",
      "MEOX1-EN1                                             4.533017                                               \n",
      "MEOX1-EMX1                                            4.533017                                               \n",
      "RHOXF1-EMX2                                           4.649823                                               \n",
      "PITX2-EMX2                                            4.649823                                               \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells_cosine  \n",
      "PAX4-MEOX1                                           0.202569  \n",
      "MEOX1-VSX2                                           0.202569  \n",
      "MEOX1-TLX2                                           0.202569  \n",
      "MEOX1-Lhx8                                           0.202569  \n",
      "MEOX1-LHX6                                           0.202569  \n",
      "MEOX1-EN1                                            0.202569  \n",
      "MEOX1-EMX1                                           0.202569  \n",
      "RHOXF1-EMX2                                          0.220424  \n",
      "PITX2-EMX2                                           0.220424  ]\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of cluster that have tf-pairs that only occure in the investigated cluster in that tissue. (filter thresholds: cosine=0.001 , log2fc=0.10)')\n",
    "print(len(cluster_with_specials))\n",
    "print(cluster_with_specials)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "482617be",
   "metadata": {},
   "source": [
    "### Checking - The found TF-pairs, that occure only in a single cluster of a tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "30c3c481",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found cluster: 3\n"
     ]
    }
   ],
   "source": [
    "print(f'Found cluster: {len(cluster_with_specials)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "288b1493",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[                ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc  \\\n",
      "PHOX2A-BHLHE22                                          -3.273248                                               \n",
      "PHOX2A-OLIG3                                            -3.273248                                               \n",
      "E2F3-E2F4                                               -3.270808                                               \n",
      "Rhox11-ZBED1                                            -3.266179                                               \n",
      "NEUROG2-PROP1                                           -3.265071                                               \n",
      "OLIG2-PROP1                                             -3.265071                                               \n",
      "Dlx2-EMX1                                               -3.260364                                               \n",
      "Dlx2-EN1                                                -3.260364                                               \n",
      "Dlx2-PAX4                                               -3.260364                                               \n",
      "Dlx2-TLX2                                               -3.260364                                               \n",
      "Dlx2-VSX2                                               -3.260364                                               \n",
      "E2F1-E2F3                                               -3.260364                                               \n",
      "FOXF2-LHX6                                              -3.260364                                               \n",
      "Lhx8-Dlx2                                               -3.260364                                               \n",
      "Six3-MYBL2                                              -3.281143                                               \n",
      "FOXD2-LHX6                                              -3.438723                                               \n",
      "EMX1-Six3                                               -3.428891                                               \n",
      "EN1-Six3                                                -3.428891                                               \n",
      "Lhx8-Six3                                               -3.428891                                               \n",
      "PAX4-Six3                                               -3.428891                                               \n",
      "Six3-TLX2                                               -3.428891                                               \n",
      "Six3-VSX2                                               -3.428891                                               \n",
      "Atoh1-PHOX2A                                            -3.374966                                               \n",
      "EVX1-GSC2                                               -3.293265                                               \n",
      "NEUROG2-PHOX2A                                          -3.291634                                               \n",
      "OLIG2-PHOX2A                                            -3.291634                                               \n",
      "Atoh1-PROP1                                             -3.286949                                               \n",
      "\n",
      "                ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc  \\\n",
      "PHOX2A-BHLHE22                                          -3.273248                                                        \n",
      "PHOX2A-OLIG3                                            -3.273248                                                        \n",
      "E2F3-E2F4                                               -3.270808                                                        \n",
      "Rhox11-ZBED1                                            -3.266179                                                        \n",
      "NEUROG2-PROP1                                           -3.265071                                                        \n",
      "OLIG2-PROP1                                             -3.265071                                                        \n",
      "Dlx2-EMX1                                               -3.260364                                                        \n",
      "Dlx2-EN1                                                -3.260364                                                        \n",
      "Dlx2-PAX4                                               -3.260364                                                        \n",
      "Dlx2-TLX2                                               -3.260364                                                        \n",
      "Dlx2-VSX2                                               -3.260364                                                        \n",
      "E2F1-E2F3                                               -3.260364                                                        \n",
      "FOXF2-LHX6                                              -3.260364                                                        \n",
      "Lhx8-Dlx2                                               -3.260364                                                        \n",
      "Six3-MYBL2                                              -3.281143                                                        \n",
      "FOXD2-LHX6                                              -3.438723                                                        \n",
      "EMX1-Six3                                               -3.428891                                                        \n",
      "EN1-Six3                                                -3.428891                                                        \n",
      "Lhx8-Six3                                               -3.428891                                                        \n",
      "PAX4-Six3                                               -3.428891                                                        \n",
      "Six3-TLX2                                               -3.428891                                                        \n",
      "Six3-VSX2                                               -3.428891                                                        \n",
      "Atoh1-PHOX2A                                            -3.374966                                                        \n",
      "EVX1-GSC2                                               -3.293265                                                        \n",
      "NEUROG2-PHOX2A                                          -3.291634                                                        \n",
      "OLIG2-PHOX2A                                            -3.291634                                                        \n",
      "Atoh1-PROP1                                             -3.286949                                                        \n",
      "\n",
      "                ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine  \n",
      "PHOX2A-BHLHE22                                           0.503637   \n",
      "PHOX2A-OLIG3                                             0.503637   \n",
      "E2F3-E2F4                                                0.502688   \n",
      "Rhox11-ZBED1                                             0.500892   \n",
      "NEUROG2-PROP1                                            0.500463   \n",
      "OLIG2-PROP1                                              0.500463   \n",
      "Dlx2-EMX1                                                0.498643   \n",
      "Dlx2-EN1                                                 0.498643   \n",
      "Dlx2-PAX4                                                0.498643   \n",
      "Dlx2-TLX2                                                0.498643   \n",
      "Dlx2-VSX2                                                0.498643   \n",
      "E2F1-E2F3                                                0.498643   \n",
      "FOXF2-LHX6                                               0.498643   \n",
      "Lhx8-Dlx2                                                0.498643   \n",
      "Six3-MYBL2                                               0.506720   \n",
      "FOXD2-LHX6                                               0.571908   \n",
      "EMX1-Six3                                                0.567630   \n",
      "EN1-Six3                                                 0.567630   \n",
      "Lhx8-Six3                                                0.567630   \n",
      "PAX4-Six3                                                0.567630   \n",
      "Six3-TLX2                                                0.567630   \n",
      "Six3-VSX2                                                0.567630   \n",
      "Atoh1-PHOX2A                                             0.544673   \n",
      "EVX1-GSC2                                                0.511486   \n",
      "NEUROG2-PHOX2A                                           0.510842   \n",
      "OLIG2-PHOX2A                                             0.510842   \n",
      "Atoh1-PROP1                                              0.508998   ,              A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c2_Kupffer_cells_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                      \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c1_Adipocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                   \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c3_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c6_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                                   \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c4_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c7_Hepatocytes_cosine_log2fc  \\\n",
      "BARHL2-CUX2                                           4.404766                                                    \n",
      "\n",
      "             A8WNZ_right_lobe_of_liver_c5_Cholangiocytes_cosine  \n",
      "BARHL2-CUX2                                           0.173113   ,              ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c3_Enterocytes_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                               \n",
      "MEOX1-VSX2                                            4.533017                                               \n",
      "MEOX1-TLX2                                            4.533017                                               \n",
      "MEOX1-Lhx8                                            4.533017                                               \n",
      "MEOX1-LHX6                                            4.533017                                               \n",
      "MEOX1-EN1                                             4.533017                                               \n",
      "MEOX1-EMX1                                            4.533017                                               \n",
      "RHOXF1-EMX2                                           4.649823                                               \n",
      "PITX2-EMX2                                            4.649823                                               \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c2_Goblet_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                \n",
      "MEOX1-VSX2                                            4.533017                                                \n",
      "MEOX1-TLX2                                            4.533017                                                \n",
      "MEOX1-Lhx8                                            4.533017                                                \n",
      "MEOX1-LHX6                                            4.533017                                                \n",
      "MEOX1-EN1                                             4.533017                                                \n",
      "MEOX1-EMX1                                            4.533017                                                \n",
      "RHOXF1-EMX2                                           4.649823                                                \n",
      "PITX2-EMX2                                            4.649823                                                \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                       \n",
      "MEOX1-VSX2                                            4.533017                                                       \n",
      "MEOX1-TLX2                                            4.533017                                                       \n",
      "MEOX1-Lhx8                                            4.533017                                                       \n",
      "MEOX1-LHX6                                            4.533017                                                       \n",
      "MEOX1-EN1                                             4.533017                                                       \n",
      "MEOX1-EMX1                                            4.533017                                                       \n",
      "RHOXF1-EMX2                                           4.649823                                                       \n",
      "PITX2-EMX2                                            4.649823                                                       \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                       \n",
      "MEOX1-VSX2                                            4.533017                                                       \n",
      "MEOX1-TLX2                                            4.533017                                                       \n",
      "MEOX1-Lhx8                                            4.533017                                                       \n",
      "MEOX1-LHX6                                            4.533017                                                       \n",
      "MEOX1-EN1                                             4.533017                                                       \n",
      "MEOX1-EMX1                                            4.533017                                                       \n",
      "RHOXF1-EMX2                                           4.649823                                                       \n",
      "PITX2-EMX2                                            4.649823                                                       \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                                \n",
      "MEOX1-VSX2                                            4.533017                                                \n",
      "MEOX1-TLX2                                            4.533017                                                \n",
      "MEOX1-Lhx8                                            4.533017                                                \n",
      "MEOX1-LHX6                                            4.533017                                                \n",
      "MEOX1-EN1                                             4.533017                                                \n",
      "MEOX1-EMX1                                            4.533017                                                \n",
      "RHOXF1-EMX2                                           4.649823                                                \n",
      "PITX2-EMX2                                            4.649823                                                \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \\\n",
      "PAX4-MEOX1                                            4.533017                                               \n",
      "MEOX1-VSX2                                            4.533017                                               \n",
      "MEOX1-TLX2                                            4.533017                                               \n",
      "MEOX1-Lhx8                                            4.533017                                               \n",
      "MEOX1-LHX6                                            4.533017                                               \n",
      "MEOX1-EN1                                             4.533017                                               \n",
      "MEOX1-EMX1                                            4.533017                                               \n",
      "RHOXF1-EMX2                                           4.649823                                               \n",
      "PITX2-EMX2                                            4.649823                                               \n",
      "\n",
      "             ACCQ1_colon_transverse_c5_Microfold_cells_cosine  \n",
      "PAX4-MEOX1                                           0.202569  \n",
      "MEOX1-VSX2                                           0.202569  \n",
      "MEOX1-TLX2                                           0.202569  \n",
      "MEOX1-Lhx8                                           0.202569  \n",
      "MEOX1-LHX6                                           0.202569  \n",
      "MEOX1-EN1                                            0.202569  \n",
      "MEOX1-EMX1                                           0.202569  \n",
      "RHOXF1-EMX2                                          0.220424  \n",
      "PITX2-EMX2                                           0.220424  ]\n"
     ]
    }
   ],
   "source": [
    "print(cluster_with_specials)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "83b0073d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tf_paris: 27\n",
      "Index(['PHOX2A-BHLHE22', 'PHOX2A-OLIG3', 'E2F3-E2F4', 'Rhox11-ZBED1',\n",
      "       'NEUROG2-PROP1', 'OLIG2-PROP1', 'Dlx2-EMX1', 'Dlx2-EN1', 'Dlx2-PAX4',\n",
      "       'Dlx2-TLX2', 'Dlx2-VSX2', 'E2F1-E2F3', 'FOXF2-LHX6', 'Lhx8-Dlx2',\n",
      "       'Six3-MYBL2', 'FOXD2-LHX6', 'EMX1-Six3', 'EN1-Six3', 'Lhx8-Six3',\n",
      "       'PAX4-Six3', 'Six3-TLX2', 'Six3-VSX2', 'Atoh1-PHOX2A', 'EVX1-GSC2',\n",
      "       'NEUROG2-PHOX2A', 'OLIG2-PHOX2A', 'Atoh1-PROP1'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PHOX2A-BHLHE22</th>\n",
       "      <th>PHOX2A-OLIG3</th>\n",
       "      <th>E2F3-E2F4</th>\n",
       "      <th>Rhox11-ZBED1</th>\n",
       "      <th>NEUROG2-PROP1</th>\n",
       "      <th>OLIG2-PROP1</th>\n",
       "      <th>Dlx2-EMX1</th>\n",
       "      <th>Dlx2-EN1</th>\n",
       "      <th>Dlx2-PAX4</th>\n",
       "      <th>Dlx2-TLX2</th>\n",
       "      <th>...</th>\n",
       "      <th>EN1-Six3</th>\n",
       "      <th>Lhx8-Six3</th>\n",
       "      <th>PAX4-Six3</th>\n",
       "      <th>Six3-TLX2</th>\n",
       "      <th>Six3-VSX2</th>\n",
       "      <th>Atoh1-PHOX2A</th>\n",
       "      <th>EVX1-GSC2</th>\n",
       "      <th>NEUROG2-PHOX2A</th>\n",
       "      <th>OLIG2-PHOX2A</th>\n",
       "      <th>Atoh1-PROP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc</th>\n",
       "      <td>-3.273248</td>\n",
       "      <td>-3.273248</td>\n",
       "      <td>-3.270808</td>\n",
       "      <td>-3.266179</td>\n",
       "      <td>-3.265071</td>\n",
       "      <td>-3.265071</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.374966</td>\n",
       "      <td>-3.293265</td>\n",
       "      <td>-3.291634</td>\n",
       "      <td>-3.291634</td>\n",
       "      <td>-3.286949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine_log2fc</th>\n",
       "      <td>-3.273248</td>\n",
       "      <td>-3.273248</td>\n",
       "      <td>-3.270808</td>\n",
       "      <td>-3.266179</td>\n",
       "      <td>-3.265071</td>\n",
       "      <td>-3.265071</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>-3.260364</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.428891</td>\n",
       "      <td>-3.374966</td>\n",
       "      <td>-3.293265</td>\n",
       "      <td>-3.291634</td>\n",
       "      <td>-3.291634</td>\n",
       "      <td>-3.286949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine</th>\n",
       "      <td>0.503637</td>\n",
       "      <td>0.503637</td>\n",
       "      <td>0.502688</td>\n",
       "      <td>0.500892</td>\n",
       "      <td>0.500463</td>\n",
       "      <td>0.500463</td>\n",
       "      <td>0.498643</td>\n",
       "      <td>0.498643</td>\n",
       "      <td>0.498643</td>\n",
       "      <td>0.498643</td>\n",
       "      <td>...</td>\n",
       "      <td>0.567630</td>\n",
       "      <td>0.567630</td>\n",
       "      <td>0.567630</td>\n",
       "      <td>0.567630</td>\n",
       "      <td>0.567630</td>\n",
       "      <td>0.544673</td>\n",
       "      <td>0.511486</td>\n",
       "      <td>0.510842</td>\n",
       "      <td>0.510842</td>\n",
       "      <td>0.508998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    PHOX2A-BHLHE22  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...       -3.273248   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...       -3.273248   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine        0.503637   \n",
       "\n",
       "                                                    PHOX2A-OLIG3  E2F3-E2F4  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...     -3.273248  -3.270808   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...     -3.273248  -3.270808   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine      0.503637   0.502688   \n",
       "\n",
       "                                                    Rhox11-ZBED1  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...     -3.266179   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...     -3.266179   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine      0.500892   \n",
       "\n",
       "                                                    NEUROG2-PROP1  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...      -3.265071   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...      -3.265071   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine       0.500463   \n",
       "\n",
       "                                                    OLIG2-PROP1  Dlx2-EMX1  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...    -3.265071  -3.260364   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...    -3.265071  -3.260364   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine     0.500463   0.498643   \n",
       "\n",
       "                                                    Dlx2-EN1  Dlx2-PAX4  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper... -3.260364  -3.260364   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC... -3.260364  -3.260364   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine  0.498643   0.498643   \n",
       "\n",
       "                                                    Dlx2-TLX2  ...  EN1-Six3  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...  -3.260364  ... -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...  -3.260364  ... -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine   0.498643  ...  0.567630   \n",
       "\n",
       "                                                    Lhx8-Six3  PAX4-Six3  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...  -3.428891  -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...  -3.428891  -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine   0.567630   0.567630   \n",
       "\n",
       "                                                    Six3-TLX2  Six3-VSX2  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...  -3.428891  -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...  -3.428891  -3.428891   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine   0.567630   0.567630   \n",
       "\n",
       "                                                    Atoh1-PHOX2A  EVX1-GSC2  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...     -3.374966  -3.293265   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...     -3.374966  -3.293265   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine      0.544673   0.511486   \n",
       "\n",
       "                                                    NEUROG2-PHOX2A  \\\n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...       -3.291634   \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...       -3.291634   \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine        0.510842   \n",
       "\n",
       "                                                    OLIG2-PHOX2A  Atoh1-PROP1  \n",
       "ACCPU_upper_lobe_of_left_lung_c3_na/ACCPU_upper...     -3.291634    -3.286949  \n",
       "ACCPU_upper_lobe_of_left_lung_c1_Clara_cells/AC...     -3.291634    -3.286949  \n",
       "ACCPU_upper_lobe_of_left_lung_c2_Adipocytes_cosine      0.510842     0.508998  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ACCPU_upper_lobe_of_left_lung_c2_Adipocytes\n",
    "c0 = cluster_with_specials[0].T\n",
    "print(f'Number of tf_paris: {len(c0.columns)}')\n",
    "print(c0.columns)\n",
    "c0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7ad25447",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tf_paris: 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BARHL2-CUX2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c2_Kupffer_cells_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c1_Adipocytes_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c3_Hepatocytes_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c6_Hepatocytes_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c8_Adipocyte_progenitor_cells_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c4_Hepatocytes_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8WNZ_right_lobe_of_liver_c7_Hepatocytes_cosine_log2fc</th>\n",
       "      <td>4.404766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A8WNZ_right_lobe_of_liver_c5_Cholangiocytes_cosine</th>\n",
       "      <td>0.173113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    BARHL2-CUX2\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes/A8W...     4.404766\n",
       "A8WNZ_right_lobe_of_liver_c5_Cholangiocytes_cosine     0.173113"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#A8WNZ_right_lobe_of_liver_c5_Cholangiocytes\n",
    "c1 = cluster_with_specials[1].T\n",
    "print(f'Number of tf_paris: {len(c1.columns)}')\n",
    "c1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6da30636",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tf_paris: 9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAX4-MEOX1</th>\n",
       "      <th>MEOX1-VSX2</th>\n",
       "      <th>MEOX1-TLX2</th>\n",
       "      <th>MEOX1-Lhx8</th>\n",
       "      <th>MEOX1-LHX6</th>\n",
       "      <th>MEOX1-EN1</th>\n",
       "      <th>MEOX1-EMX1</th>\n",
       "      <th>RHOXF1-EMX2</th>\n",
       "      <th>PITX2-EMX2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c3_Enterocytes_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c2_Goblet_cells_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc</th>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.533017</td>\n",
       "      <td>4.649823</td>\n",
       "      <td>4.649823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells_cosine</th>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.202569</td>\n",
       "      <td>0.220424</td>\n",
       "      <td>0.220424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    PAX4-MEOX1  MEOX1-VSX2  \\\n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells_cosine      0.202569    0.202569   \n",
       "\n",
       "                                                    MEOX1-TLX2  MEOX1-Lhx8  \\\n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017    4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells_cosine      0.202569    0.202569   \n",
       "\n",
       "                                                    MEOX1-LHX6  MEOX1-EN1  \\\n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017   4.533017   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells_cosine      0.202569   0.202569   \n",
       "\n",
       "                                                    MEOX1-EMX1  RHOXF1-EMX2  \\\n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.533017     4.649823   \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells_cosine      0.202569     0.220424   \n",
       "\n",
       "                                                    PITX2-EMX2  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1...    4.649823  \n",
       "ACCQ1_colon_transverse_c5_Microfold_cells_cosine      0.220424  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#ACCQ1_colon_transverse_c5_Microfold_cells\n",
    "c5 = cluster_with_specials[2].T\n",
    "print(f'Number of tf_paris: {len(c5.columns)}')\n",
    "c5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8a3d1f74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <th>TF1_count</th>\n",
       "      <th>TF2_count</th>\n",
       "      <th>cosine</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>88393.000000</td>\n",
       "      <td>88393.000000</td>\n",
       "      <td>88393.000000</td>\n",
       "      <td>88393.000000</td>\n",
       "      <td>8.839300e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.531456</td>\n",
       "      <td>150.783705</td>\n",
       "      <td>150.783705</td>\n",
       "      <td>0.033024</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>19.553938</td>\n",
       "      <td>261.108407</td>\n",
       "      <td>261.108407</td>\n",
       "      <td>0.031063</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002529</td>\n",
       "      <td>-8.621031e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>0.016346</td>\n",
       "      <td>-2.108654e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>57.000000</td>\n",
       "      <td>57.000000</td>\n",
       "      <td>0.025318</td>\n",
       "      <td>9.278559e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>138.000000</td>\n",
       "      <td>138.000000</td>\n",
       "      <td>0.040161</td>\n",
       "      <td>2.295559e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1298.000000</td>\n",
       "      <td>1884.000000</td>\n",
       "      <td>1884.000000</td>\n",
       "      <td>1.186732</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF1_TF2_count     TF1_count     TF2_count        cosine        zscore\n",
       "count   88393.000000  88393.000000  88393.000000  88393.000000  8.839300e+04\n",
       "mean        3.531456    150.783705    150.783705      0.033024           inf\n",
       "std        19.553938    261.108407    261.108407      0.031063           NaN\n",
       "min         1.000000      1.000000      1.000000      0.002529 -8.621031e+00\n",
       "25%         1.000000     24.000000     24.000000      0.016346 -2.108654e-02\n",
       "50%         1.000000     57.000000     57.000000      0.025318  9.278559e-01\n",
       "75%         2.000000    138.000000    138.000000      0.040161  2.295559e+00\n",
       "max      1298.000000   1884.000000   1884.000000      1.186732           inf"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking the original tf-comb market basket analysis for the tf-co-occurences of ACCQ1_colon_transverse_c5_Microfold_cells . \n",
    "comb = CombObj().from_pickle(f\"{main_analysis_path}/ACCQ1_colon_transverse/ACCQ1_colon_transverse_c5_Microfold_cells.pkl\")\n",
    "comb.rules.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2493c4c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1</th>\n",
       "      <th>TF2</th>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <th>TF1_count</th>\n",
       "      <th>TF2_count</th>\n",
       "      <th>cosine</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PAX4-MEOX1</th>\n",
       "      <td>PAX4</td>\n",
       "      <td>MEOX1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-VSX2</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>VSX2</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-Lhx8</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>Lhx8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-LHX6</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>LHX6</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-EN1</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>EN1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-EMX1</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>EMX1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-Lhx8</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>Lhx8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1-Lhx8</th>\n",
       "      <td>MEOX1</td>\n",
       "      <td>Lhx8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.218218</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RHOXF1-EMX2</th>\n",
       "      <td>RHOXF1</td>\n",
       "      <td>EMX2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0.235702</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PITX2-EMX2</th>\n",
       "      <td>PITX2</td>\n",
       "      <td>EMX2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0.235702</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                TF1    TF2  TF1_TF2_count  TF1_count  TF2_count    cosine  \\\n",
       "PAX4-MEOX1     PAX4  MEOX1              1          3          7  0.218218   \n",
       "MEOX1-VSX2    MEOX1   VSX2              1          7          3  0.218218   \n",
       "MEOX1-Lhx8    MEOX1   Lhx8              1          7          3  0.218218   \n",
       "MEOX1-LHX6    MEOX1   LHX6              1          7          3  0.218218   \n",
       "MEOX1-EN1     MEOX1    EN1              1          7          3  0.218218   \n",
       "MEOX1-EMX1    MEOX1   EMX1              1          7          3  0.218218   \n",
       "MEOX1-Lhx8    MEOX1   Lhx8              1          7          3  0.218218   \n",
       "MEOX1-Lhx8    MEOX1   Lhx8              1          7          3  0.218218   \n",
       "RHOXF1-EMX2  RHOXF1   EMX2              1          9          2  0.235702   \n",
       "PITX2-EMX2    PITX2   EMX2              1          9          2  0.235702   \n",
       "\n",
       "             zscore  \n",
       "PAX4-MEOX1      inf  \n",
       "MEOX1-VSX2      inf  \n",
       "MEOX1-Lhx8      7.0  \n",
       "MEOX1-LHX6      inf  \n",
       "MEOX1-EN1       7.0  \n",
       "MEOX1-EMX1      7.0  \n",
       "MEOX1-Lhx8      7.0  \n",
       "MEOX1-Lhx8      7.0  \n",
       "RHOXF1-EMX2     inf  \n",
       "PITX2-EMX2      inf  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Intersting is that mostly MEOX1 is occureing.\n",
    "pd.DataFrame([\n",
    "    comb.rules.loc['PAX4-MEOX1'],\n",
    "    comb.rules.loc['MEOX1-VSX2'],\n",
    "    comb.rules.loc['MEOX1-Lhx8'],\n",
    "    comb.rules.loc['MEOX1-LHX6'],\n",
    "    comb.rules.loc['MEOX1-EN1'],\n",
    "    comb.rules.loc['MEOX1-EMX1'],\n",
    "    comb.rules.loc['MEOX1-Lhx8'],\n",
    "    comb.rules.loc['MEOX1-Lhx8'],\n",
    "    comb.rules.loc['RHOXF1-EMX2'],\n",
    "    comb.rules.loc['PITX2-EMX2'],\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "40fb3969",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<tfcomb.objects.DiffCombObj at 0x7f241dd8d750>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_path}ACCQ1_colon_transverse/ACCQ1_colon_transverse_normalized.pkl\")\n",
    "diff_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "47c79d3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1</th>\n",
       "      <th>TF2</th>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c3_Enterocytes_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c2_Goblet_cells_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c4_Paneth_cells_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c7_Fibroblasts_cosine</th>\n",
       "      <th>ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c3_Enterocytes_cosine_log2fc</th>\n",
       "      <th>...</th>\n",
       "      <th>ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c1_Gastric_chief_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c1_Gastric_chief_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc</th>\n",
       "      <th>ACCQ1_colon_transverse_c4_Paneth_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PAX3-PAX7</th>\n",
       "      <td>PAX3</td>\n",
       "      <td>PAX7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162118</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.026379</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.227107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.957778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.957778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.957778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.957778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUX2-Mecom</th>\n",
       "      <td>CUX2</td>\n",
       "      <td>Mecom</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123812</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.010103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-3.861854</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.073630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.073630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.073630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.073630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUX2-Smad4</th>\n",
       "      <td>CUX2</td>\n",
       "      <td>Smad4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.112165</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-3.729600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nr2f6var.2-Rbpjl</th>\n",
       "      <td>Nr2f6var.2</td>\n",
       "      <td>Rbpjl</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.107709</td>\n",
       "      <td>0.028727</td>\n",
       "      <td>0.046417</td>\n",
       "      <td>0.02256</td>\n",
       "      <td>0.071752</td>\n",
       "      <td>0.010684</td>\n",
       "      <td>-3.675600</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552961</td>\n",
       "      <td>0.256430</td>\n",
       "      <td>-1.094958</td>\n",
       "      <td>0.933550</td>\n",
       "      <td>0.809391</td>\n",
       "      <td>-0.541997</td>\n",
       "      <td>1.486511</td>\n",
       "      <td>-1.351388</td>\n",
       "      <td>0.677120</td>\n",
       "      <td>2.028508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRF7-IRF9</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>IRF9</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102462</td>\n",
       "      <td>0.034539</td>\n",
       "      <td>0.020203</td>\n",
       "      <td>0.03250</td>\n",
       "      <td>0.023097</td>\n",
       "      <td>0.016534</td>\n",
       "      <td>-3.609327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.573815</td>\n",
       "      <td>0.068956</td>\n",
       "      <td>0.438154</td>\n",
       "      <td>0.766507</td>\n",
       "      <td>-0.504859</td>\n",
       "      <td>-0.135661</td>\n",
       "      <td>0.192692</td>\n",
       "      <td>0.369197</td>\n",
       "      <td>0.697551</td>\n",
       "      <td>0.328353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUX1-MAF</th>\n",
       "      <td>CUX1</td>\n",
       "      <td>MAF</td>\n",
       "      <td>0.272398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.009225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.944256</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.006295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.006295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.006295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.006295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POU6F1-LHX5</th>\n",
       "      <td>POU6F1</td>\n",
       "      <td>LHX5</td>\n",
       "      <td>0.281344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007344</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.015734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.989382</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.850419</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-0.593499</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.443918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POU6F1-HOXB6</th>\n",
       "      <td>POU6F1</td>\n",
       "      <td>HOXB6</td>\n",
       "      <td>0.281344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007344</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.015734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.989382</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.850419</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-0.593499</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.443918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NKX6-1-POU6F1</th>\n",
       "      <td>NKX6-1</td>\n",
       "      <td>POU6F1</td>\n",
       "      <td>0.281344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007344</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.015734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.989382</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.850419</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-0.593499</td>\n",
       "      <td>0.850419</td>\n",
       "      <td>-1.443918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.443918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POU6F1-MEOX2</th>\n",
       "      <td>POU6F1</td>\n",
       "      <td>MEOX2</td>\n",
       "      <td>0.293653</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.020458</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.049252</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.694716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.694716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.694716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.694716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223653 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         TF1     TF2  \\\n",
       "PAX3-PAX7               PAX3    PAX7   \n",
       "CUX2-Mecom              CUX2   Mecom   \n",
       "CUX2-Smad4              CUX2   Smad4   \n",
       "Nr2f6var.2-Rbpjl  Nr2f6var.2   Rbpjl   \n",
       "IRF7-IRF9               IRF7    IRF9   \n",
       "...                      ...     ...   \n",
       "CUX1-MAF                CUX1     MAF   \n",
       "POU6F1-LHX5           POU6F1    LHX5   \n",
       "POU6F1-HOXB6          POU6F1   HOXB6   \n",
       "NKX6-1-POU6F1         NKX6-1  POU6F1   \n",
       "POU6F1-MEOX2          POU6F1   MEOX2   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c5_Microfold_cells_cosine  \\\n",
       "PAX3-PAX7                                                 0.000000   \n",
       "CUX2-Mecom                                                0.000000   \n",
       "CUX2-Smad4                                                0.000000   \n",
       "Nr2f6var.2-Rbpjl                                          0.000000   \n",
       "IRF7-IRF9                                                 0.000000   \n",
       "...                                                            ...   \n",
       "CUX1-MAF                                                  0.272398   \n",
       "POU6F1-LHX5                                               0.281344   \n",
       "POU6F1-HOXB6                                              0.281344   \n",
       "NKX6-1-POU6F1                                             0.281344   \n",
       "POU6F1-MEOX2                                              0.293653   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c3_Enterocytes_cosine  \\\n",
       "PAX3-PAX7                                             0.162118   \n",
       "CUX2-Mecom                                            0.123812   \n",
       "CUX2-Smad4                                            0.112165   \n",
       "Nr2f6var.2-Rbpjl                                      0.107709   \n",
       "IRF7-IRF9                                             0.102462   \n",
       "...                                                        ...   \n",
       "CUX1-MAF                                              0.000000   \n",
       "POU6F1-LHX5                                           0.000000   \n",
       "POU6F1-HOXB6                                          0.000000   \n",
       "NKX6-1-POU6F1                                         0.000000   \n",
       "POU6F1-MEOX2                                          0.000000   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c2_Goblet_cells_cosine  \\\n",
       "PAX3-PAX7                                              0.000000   \n",
       "CUX2-Mecom                                             0.000000   \n",
       "CUX2-Smad4                                             0.000000   \n",
       "Nr2f6var.2-Rbpjl                                       0.028727   \n",
       "IRF7-IRF9                                              0.034539   \n",
       "...                                                         ...   \n",
       "CUX1-MAF                                               0.000000   \n",
       "POU6F1-LHX5                                            0.000000   \n",
       "POU6F1-HOXB6                                           0.000000   \n",
       "NKX6-1-POU6F1                                          0.000000   \n",
       "POU6F1-MEOX2                                           0.000000   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine  \\\n",
       "PAX3-PAX7                                                  0.000000      \n",
       "CUX2-Mecom                                                 0.000000      \n",
       "CUX2-Smad4                                                 0.000000      \n",
       "Nr2f6var.2-Rbpjl                                           0.046417      \n",
       "IRF7-IRF9                                                  0.020203      \n",
       "...                                                             ...      \n",
       "CUX1-MAF                                                   0.000000      \n",
       "POU6F1-LHX5                                                0.007344      \n",
       "POU6F1-HOXB6                                               0.007344      \n",
       "NKX6-1-POU6F1                                              0.007344      \n",
       "POU6F1-MEOX2                                               0.000000      \n",
       "\n",
       "                  ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine  \\\n",
       "PAX3-PAX7                                                   0.00000      \n",
       "CUX2-Mecom                                                  0.00000      \n",
       "CUX2-Smad4                                                  0.00000      \n",
       "Nr2f6var.2-Rbpjl                                            0.02256      \n",
       "IRF7-IRF9                                                   0.03250      \n",
       "...                                                             ...      \n",
       "CUX1-MAF                                                    0.00000      \n",
       "POU6F1-LHX5                                                 0.00000      \n",
       "POU6F1-HOXB6                                                0.00000      \n",
       "NKX6-1-POU6F1                                               0.00000      \n",
       "POU6F1-MEOX2                                                0.00000      \n",
       "\n",
       "                  ACCQ1_colon_transverse_c4_Paneth_cells_cosine  \\\n",
       "PAX3-PAX7                                              0.026379   \n",
       "CUX2-Mecom                                             0.010103   \n",
       "CUX2-Smad4                                             0.000000   \n",
       "Nr2f6var.2-Rbpjl                                       0.071752   \n",
       "IRF7-IRF9                                              0.023097   \n",
       "...                                                         ...   \n",
       "CUX1-MAF                                               0.009225   \n",
       "POU6F1-LHX5                                            0.015734   \n",
       "POU6F1-HOXB6                                           0.015734   \n",
       "NKX6-1-POU6F1                                          0.015734   \n",
       "POU6F1-MEOX2                                           0.020458   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c7_Fibroblasts_cosine  \\\n",
       "PAX3-PAX7                                             0.000000   \n",
       "CUX2-Mecom                                            0.000000   \n",
       "CUX2-Smad4                                            0.000000   \n",
       "Nr2f6var.2-Rbpjl                                      0.010684   \n",
       "IRF7-IRF9                                             0.016534   \n",
       "...                                                        ...   \n",
       "CUX1-MAF                                              0.000000   \n",
       "POU6F1-LHX5                                           0.000000   \n",
       "POU6F1-HOXB6                                          0.000000   \n",
       "NKX6-1-POU6F1                                         0.000000   \n",
       "POU6F1-MEOX2                                          0.000000   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c5_Microfold_cells/ACCQ1_colon_transverse_c3_Enterocytes_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                 -4.227107                                               \n",
       "CUX2-Mecom                                                -3.861854                                               \n",
       "CUX2-Smad4                                                -3.729600                                               \n",
       "Nr2f6var.2-Rbpjl                                          -3.675600                                               \n",
       "IRF7-IRF9                                                 -3.609327                                               \n",
       "...                                                             ...                                               \n",
       "CUX1-MAF                                                   4.944256                                               \n",
       "POU6F1-LHX5                                                4.989382                                               \n",
       "POU6F1-HOXB6                                               4.989382                                               \n",
       "NKX6-1-POU6F1                                              4.989382                                               \n",
       "POU6F1-MEOX2                                               5.049252                                               \n",
       "\n",
       "                  ...  \\\n",
       "PAX3-PAX7         ...   \n",
       "CUX2-Mecom        ...   \n",
       "CUX2-Smad4        ...   \n",
       "Nr2f6var.2-Rbpjl  ...   \n",
       "IRF7-IRF9         ...   \n",
       "...               ...   \n",
       "CUX1-MAF          ...   \n",
       "POU6F1-LHX5       ...   \n",
       "POU6F1-HOXB6      ...   \n",
       "NKX6-1-POU6F1     ...   \n",
       "POU6F1-MEOX2      ...   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c6_Gastric_chief_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                                    \n",
       "CUX2-Mecom                                                 0.000000                                                    \n",
       "CUX2-Smad4                                                 0.000000                                                    \n",
       "Nr2f6var.2-Rbpjl                                          -0.552961                                                    \n",
       "IRF7-IRF9                                                  0.573815                                                    \n",
       "...                                                             ...                                                    \n",
       "CUX1-MAF                                                   0.000000                                                    \n",
       "POU6F1-LHX5                                               -0.850419                                                    \n",
       "POU6F1-HOXB6                                              -0.850419                                                    \n",
       "NKX6-1-POU6F1                                             -0.850419                                                    \n",
       "POU6F1-MEOX2                                               0.000000                                                    \n",
       "\n",
       "                  ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                                    \n",
       "CUX2-Mecom                                                 0.000000                                                    \n",
       "CUX2-Smad4                                                 0.000000                                                    \n",
       "Nr2f6var.2-Rbpjl                                           0.256430                                                    \n",
       "IRF7-IRF9                                                  0.068956                                                    \n",
       "...                                                             ...                                                    \n",
       "CUX1-MAF                                                   0.000000                                                    \n",
       "POU6F1-LHX5                                                0.000000                                                    \n",
       "POU6F1-HOXB6                                               0.000000                                                    \n",
       "NKX6-1-POU6F1                                              0.000000                                                    \n",
       "POU6F1-MEOX2                                               0.000000                                                    \n",
       "\n",
       "                  ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                 -1.957778                                             \n",
       "CUX2-Mecom                                                -1.073630                                             \n",
       "CUX2-Smad4                                                 0.000000                                             \n",
       "Nr2f6var.2-Rbpjl                                          -1.094958                                             \n",
       "IRF7-IRF9                                                  0.438154                                             \n",
       "...                                                             ...                                             \n",
       "CUX1-MAF                                                  -1.006295                                             \n",
       "POU6F1-LHX5                                               -1.443918                                             \n",
       "POU6F1-HOXB6                                              -1.443918                                             \n",
       "NKX6-1-POU6F1                                             -1.443918                                             \n",
       "POU6F1-MEOX2                                              -1.694716                                             \n",
       "\n",
       "                  ACCQ1_colon_transverse_c2_Goblet_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                            \n",
       "CUX2-Mecom                                                 0.000000                                            \n",
       "CUX2-Smad4                                                 0.000000                                            \n",
       "Nr2f6var.2-Rbpjl                                           0.933550                                            \n",
       "IRF7-IRF9                                                  0.766507                                            \n",
       "...                                                             ...                                            \n",
       "CUX1-MAF                                                   0.000000                                            \n",
       "POU6F1-LHX5                                                0.000000                                            \n",
       "POU6F1-HOXB6                                               0.000000                                            \n",
       "NKX6-1-POU6F1                                              0.000000                                            \n",
       "POU6F1-MEOX2                                               0.000000                                            \n",
       "\n",
       "                  ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c1_Gastric_chief_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                                           \n",
       "CUX2-Mecom                                                 0.000000                                                           \n",
       "CUX2-Smad4                                                 0.000000                                                           \n",
       "Nr2f6var.2-Rbpjl                                           0.809391                                                           \n",
       "IRF7-IRF9                                                 -0.504859                                                           \n",
       "...                                                             ...                                                           \n",
       "CUX1-MAF                                                   0.000000                                                           \n",
       "POU6F1-LHX5                                                0.850419                                                           \n",
       "POU6F1-HOXB6                                               0.850419                                                           \n",
       "NKX6-1-POU6F1                                              0.850419                                                           \n",
       "POU6F1-MEOX2                                               0.000000                                                           \n",
       "\n",
       "                  ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                 -1.957778                                                    \n",
       "CUX2-Mecom                                                -1.073630                                                    \n",
       "CUX2-Smad4                                                 0.000000                                                    \n",
       "Nr2f6var.2-Rbpjl                                          -0.541997                                                    \n",
       "IRF7-IRF9                                                 -0.135661                                                    \n",
       "...                                                             ...                                                    \n",
       "CUX1-MAF                                                  -1.006295                                                    \n",
       "POU6F1-LHX5                                               -0.593499                                                    \n",
       "POU6F1-HOXB6                                              -0.593499                                                    \n",
       "NKX6-1-POU6F1                                             -0.593499                                                    \n",
       "POU6F1-MEOX2                                              -1.694716                                                    \n",
       "\n",
       "                  ACCQ1_colon_transverse_c6_Gastric_chief_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                                   \n",
       "CUX2-Mecom                                                 0.000000                                                   \n",
       "CUX2-Smad4                                                 0.000000                                                   \n",
       "Nr2f6var.2-Rbpjl                                           1.486511                                                   \n",
       "IRF7-IRF9                                                  0.192692                                                   \n",
       "...                                                             ...                                                   \n",
       "CUX1-MAF                                                   0.000000                                                   \n",
       "POU6F1-LHX5                                                0.850419                                                   \n",
       "POU6F1-HOXB6                                               0.850419                                                   \n",
       "NKX6-1-POU6F1                                              0.850419                                                   \n",
       "POU6F1-MEOX2                                               0.000000                                                   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c1_Gastric_chief_cells/ACCQ1_colon_transverse_c4_Paneth_cells_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                 -1.957778                                                    \n",
       "CUX2-Mecom                                                -1.073630                                                    \n",
       "CUX2-Smad4                                                 0.000000                                                    \n",
       "Nr2f6var.2-Rbpjl                                          -1.351388                                                    \n",
       "IRF7-IRF9                                                  0.369197                                                    \n",
       "...                                                             ...                                                    \n",
       "CUX1-MAF                                                  -1.006295                                                    \n",
       "POU6F1-LHX5                                               -1.443918                                                    \n",
       "POU6F1-HOXB6                                              -1.443918                                                    \n",
       "NKX6-1-POU6F1                                             -1.443918                                                    \n",
       "POU6F1-MEOX2                                              -1.694716                                                    \n",
       "\n",
       "                  ACCQ1_colon_transverse_c1_Gastric_chief_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \\\n",
       "PAX3-PAX7                                                  0.000000                                                   \n",
       "CUX2-Mecom                                                 0.000000                                                   \n",
       "CUX2-Smad4                                                 0.000000                                                   \n",
       "Nr2f6var.2-Rbpjl                                           0.677120                                                   \n",
       "IRF7-IRF9                                                  0.697551                                                   \n",
       "...                                                             ...                                                   \n",
       "CUX1-MAF                                                   0.000000                                                   \n",
       "POU6F1-LHX5                                                0.000000                                                   \n",
       "POU6F1-HOXB6                                               0.000000                                                   \n",
       "NKX6-1-POU6F1                                              0.000000                                                   \n",
       "POU6F1-MEOX2                                               0.000000                                                   \n",
       "\n",
       "                  ACCQ1_colon_transverse_c4_Paneth_cells/ACCQ1_colon_transverse_c7_Fibroblasts_cosine_log2fc  \n",
       "PAX3-PAX7                                                  1.957778                                           \n",
       "CUX2-Mecom                                                 1.073630                                           \n",
       "CUX2-Smad4                                                 0.000000                                           \n",
       "Nr2f6var.2-Rbpjl                                           2.028508                                           \n",
       "IRF7-IRF9                                                  0.328353                                           \n",
       "...                                                             ...                                           \n",
       "CUX1-MAF                                                   1.006295                                           \n",
       "POU6F1-LHX5                                                1.443918                                           \n",
       "POU6F1-HOXB6                                               1.443918                                           \n",
       "NKX6-1-POU6F1                                              1.443918                                           \n",
       "POU6F1-MEOX2                                               1.694716                                           \n",
       "\n",
       "[223653 rows x 30 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_obj.rules"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81cbfe9b",
   "metadata": {},
   "source": [
    "####  Check what could be a good threshold\n",
    "For filtering the cluster with tf-pairs, that show significant differences between all\n",
    "cluster in that tissue. \n",
    "At the moment we just look that a row of the cluster_with_high_changes has the same value in each column of the log2fc. Because of this we cant find that much single occurences.\n",
    "What would be a good threshold e.g. for the cluster ACCQ1_colon_transverse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "924d78f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_if_row_has_min_number_of_same_values(row, min_same_val:int):\n",
    "    '''\n",
    "    Check if a row of a dataframe has a minumum count of same values in each column.\n",
    "    Returns true if a minmum count of same values is reached.\n",
    "    ---\n",
    "    Parameters:\n",
    "        row: pd.dataframe row,\n",
    "            Contains the row of a dataframe, with the rows values.\n",
    "        min_same_val: int,\n",
    "            Threshold for minimum number of same values in the row.\n",
    "    ---\n",
    "    Return True or False\n",
    "    '''\n",
    "    l = len(row.index)\n",
    "    # all values in the row should be positiv\n",
    "    abs_row = row.abs()\n",
    "    \n",
    "    # finds number of values in the row that occure once. \n",
    "    count_of_distinct_value = abs_row.nunique()\n",
    "    \n",
    "    # calculates the number of values that are the same in the row. \n",
    "    count_same_val = (l + 1) - count_of_distinct_value\n",
    "    #print(count_same_val)\n",
    "    \n",
    "    return (count_same_val >= min_same_val)\n",
    "\n",
    "def find_rows_with_min_same_values(df:pd.DataFrame, min_same_val:int) -> pd.DataFrame:\n",
    "    '''\n",
    "    Remove all rows of a Dataframe that do not have a minimum count of same value in each column and\n",
    "    return the reduced dataframe.\n",
    "    '''\n",
    "    \n",
    "    return df[df.apply(lambda x: check_if_row_has_min_number_of_same_values(x, min_same_val=min_same_val), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "493b9813",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ACCQ1_colon_transverse_c5_Microfold_cells.pkl', 'ACCQ1_colon_transverse_c3_Enterocytes.pkl', 'ACCQ1_colon_transverse_c2_Goblet_cells.pkl', 'ACCQ1_colon_transverse_c6_Gastric_chief_cells.pkl', 'ACCQ1_colon_transverse_c1_Gastric_chief_cells.pkl', 'ACCQ1_colon_transverse_c4_Paneth_cells.pkl', 'ACCQ1_colon_transverse_c7_Fibroblasts.pkl']\n"
     ]
    }
   ],
   "source": [
    "files = read_in_file_names_of_folder(rel_path=f\"{answers_path}ACCQ1_colon_transverse/\")\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a7720790",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCQ1_colon_transverse_c5_Microfold_cells.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 30\n",
      "ACCQ1_colon_transverse_c3_Enterocytes.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 0\n",
      "ACCQ1_colon_transverse_c2_Goblet_cells.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 0\n",
      "ACCQ1_colon_transverse_c6_Gastric_chief_cells.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 0\n",
      "ACCQ1_colon_transverse_c1_Gastric_chief_cells.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 16\n",
      "ACCQ1_colon_transverse_c4_Paneth_cells.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 0\n",
      "ACCQ1_colon_transverse_c7_Fibroblasts.pkl\n",
      "Number of significant, high log2fc TF-pairs found in Cluster: 0\n"
     ]
    }
   ],
   "source": [
    "cluster = []\n",
    "for file in files:\n",
    "    print(file)\n",
    "    df = pd.read_pickle(f\"{answers_path}ACCQ1_colon_transverse/{file}\")\n",
    "    print(f\"Number of significant, high log2fc TF-pairs found in Cluster: {df.shape[0]}\")\n",
    "    cluster.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4892fcd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_paris_c5 = cluster[0]\n",
    "#tf_paris_c5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2177aa3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "max_same_values = len(tf_paris_c5.columns)\n",
    "print(max_same_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5077eac9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "minN: 1, tf_pairs: (30, 7)\n",
      "minN: 2, tf_pairs: (23, 7)\n",
      "minN: 3, tf_pairs: (21, 7)\n",
      "minN: 4, tf_pairs: (19, 7)\n",
      "minN: 5, tf_pairs: (16, 7)\n",
      "minN: 6, tf_pairs: (9, 7)\n"
     ]
    }
   ],
   "source": [
    "x = range(1, max_same_values)\n",
    "y = []\n",
    "for minN in range(1, max_same_values):\n",
    "    df = find_rows_with_min_same_values(df=tf_paris_c5, min_same_val=minN)\n",
    "    print(f\"minN: {minN}, tf_pairs: {df.shape}\")\n",
    "    y.append(df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "60644843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIwAAAFNCAYAAABi2vQZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAABfP0lEQVR4nO3dd3xUdfb/8ffJpEx6CAlJSKODiHQIdhB13bUXRM3ud7vr9l7ELbpr2eJ3+3eruz/XNYDi2repIHZCFxFEVCAJvSUQINTP7497A5MwSQZkMimv5+ORR2ZuPXPn3oE5OZ9zzTknAAAAAAAAoFFcrAMAAAAAAABAx0LCCAAAAAAAAE2QMAIAAAAAAEATJIwAAAAAAADQBAkjAAAAAAAANEHCCAAAAAAAAE2QMAKALs7Mys3smXbe55tmNrGV+XPN7FPvY/vOzAb4j/9gZt8LmfdZM9tsZvVm1tPMzjaz1f7zq052n9HSPD4z+7eZfdSf9zEze7md45loZjXtsJ8+/vsYfxLrthqjmd1vZne2Mv9OM9tmZptOdN/wmNntZvZgrONA+4nWvyXm+X9mttPM5kew/NHP/zDzovaZ2fwz6/3+OwYAnQEJIwCIAjMr8RMAjT/OzPaEPD/X/1J7oNlyU091LM65Cufcxad6u23s83Tn3Fwp+l8snXO3OOd+5O8rQdLPJV3snEtzzm2X9ENJv/WfPx6tOMJpK3HhaxKfc+6Dzrm/tbC9Fr8ovY8YT/k2OzIzK5b0dUlDnXP5p3C7ZmbvmdmKFuZ9ycyW+58DNWY2y8zOCFlmvJn9y8xqzWyHmc03s4+HzM8ys9+b2SYz22tmbzQmFkOW+YKZLTSz/WZ2/6l6bbHQXonLriQkoRH6b8r3Iljvfn+9K5pN/6U//WNSVP8tOUfSRZKKnHPjo7B9AMBJImEEAFHgnKvyEwBpzrk0f/KIkGkv+dN+Grqcc+6h9ozzZKo7Org8SUFJb4ZMK232PGLtdHxOOr4TEa3XYmaBaGw3ikolbXfObTnRFds4hudJ6iWpn5mNazbvV5K+LOlLkrIlDZL0uKRL/e2eKWmOpBckDZDUU9JnJX3Qn58o6Tk/9jMlZUr6pqSfmtmXQvazQdKdkv56oq+tM+oMn18xijEr5N+UH0W4ztuSjiYg/binSHr3/QYTwTEolbTWObfn/e4LAHBqkTACgE7Ar9J5xMweMrPdZrbYzEaEzP+Omb3rz1thZleHzGtSou//xfjzZrZa0mq/+uEXZrbFzOrMbJmZDQsTwyQzeyPk+XOhwwfM7GXzh3yZ2Vozu9DMLpE0TdJU/6/dr4dsstTMXvFjfsbMclp5/d80s41mtsHMPtFs3v3mDTEaJGmVP7nWzOaY2buS+kl6yt9/kpllmtlf/O2t99cNhByrV/zjsUPS7f4695pZlXlD3f5gZsn+8hP9apGv+8dvY2NViJndLKlc0rf8fT8V5nWFiy/sMAcze9F/+LqFVKOZ2WVmttS8ypRXzWx4yDprzezbZrZM0p7mX9xa2qY/77jXFHK8f29eNcweSZPMrLeZ/cPMtprZmtAkhnmVMwvNbJd//H7e7KWV+8d2m5ndFrJeknkVDhv8n1+aWVLz4+IvO8q/Jnab2UPykobhlrtQ0rOSevuv935/+hXmDaOs9Y//aZEewxAflfSEpH+p6RfvgZI+L+lG59wc59x+59xev1rjx/5iP5P0N+fcT5xz25xnkXPuen/+RySVSJrinFvjnDvonPuPvATUnWaWLknOuUf9KrrtLcQYlpld6Z9Du8z7HLnEn97bzJ40r+LpHTP7dCvbaOsYfsO8z5Y68z7Hwr5H/vKpkv6tY+9TvR9L4+fgg2a2S9LH/PPrNX+/G83st+Yl2Bq35czsFvOGfe40s/8zM/PnDTCzF/yYtvnnTuMw13ubxfSEmX0t5Li0dL63FGPYa8DMJvjXba2ZvW6tDOUNWSfbvCFcG/zX9Hhb60TgKUlnm1kP//klkpZJOjps047/t+R0M3vWPz82m9m0Vo5B2HPJzD4p6T5JZ/rv8x3+9E/7y+3w1+vdwrHo6c/fZd6/R/0jebGtxB5nx/493W5mD5tZdgTbC3suAUCn55zjhx9++OEnyj+SnKQBzabdL+nOCNe/XdJBSddJSpD0DUlrJCX486dI6i3vDwFTJe2RVODP+5ikl5vF8qy8SodkSR+QtEhSliSTdFrjus1iCEraJylHUry8LxIbJKX729knqae/7FpJF4bE/mCzbc2V95frQf66cyX9uIXXfomkzZKGSUqVND30eIYeR0l9/HnxIesfjcV//rikP/rb6iVpvqTPhByrQ5K+6L/GZEm/lPSkf7zS5X2xusdffqK//A/99+VDkvZK6hHpexwmvrmSPtXKezcg5PloSVsklUkKyEtUrJWUFLLtpZKKJSVHcm5G+JrqJJ0t73xLkXf+fF9SorwE2HuSPuAv/5qkj/iP0yRNaPZe/dk/ziMk7Zd0mj//h5Lm+e9RrqRXJf0oJMYa/3GipHWSvurHe528ayXscQ9d138+SN71cpG//rckvSMp8QSOYYqkXf6xulbStpD1b5G0rpX3P0XSYUmTWllmpryEUvPp8f57dVGz6XdKuj/Cz5bx/vt5kf9+Fkoa4s97QdLv5F37IyVtlTS5+XUd4TGcL+8zKlvSSkm3tBFXk/ep2efgVX6syZLGSJrgH4s+/ra/0uz8flre51uJ/xou8efNkHSbv62gpHP86edJqpZk/vMe8j7fGj9jWzvfw8XY0jVQKC+59yF/2Yv857ltHJt/SnrIjytB0vnNrqn1kmok/T9JORGcA/f758yfJH3Wn/awpBslvSzpY80/j+R9Fm6UN7wz6D8va+UYtHYuHd2u//wCedfQaElJkn4j6cVwn1nyro2H5X2eD/Nf+8ttvN7WYv+KvM+dIn/ff5Q0o9nxjQ/zWR32XOKHH3746ew/VBgBQGx9w//Lcq2ZbWtj2UXOuUeccwfl9ekJyvuiJOfcLOfcBufcEecNa1st74tgS+5xzu1wzu2T9x/7dElD5H1BWumc29h8Bedcg6SF8r5MjZX31+eX5SUOJkha7byeQZH6f865t/0YHpb3JSKc6/1llztvyMLtJ7CPJswsT94wn6845/Y4b1jSLyTdELLYBufcb5xzhyQ1SPq0pK/6x2u3pLubLX9Q0g+dV/XxL0n1kgafbIwn6NOS/uicq3TOHXZe76P98s8L36+dc9X+cY5UW6/pCefcK865I5LOkPcF94fOuQPOuffkJYFuCNnWADPLcc7VO+fmNdvXHc65fc651yW9Li9xJHnVWT90zm1xzm2VdIe8SpvmJsj70vxLP95HJC04gdc6VdI/nXPP+tfWvfK+4J4Vskxbx/Aaecf9GXnJiXj5w83kDS877noK0UPel8zWlskJN98/R7fJS6idrE9K+qv/+o8459Y7594yr9fTOZK+7ZxrcM4tlVcJEu49iPQYbnDO7ZCXdB15kvG+5rxeX0f882aRc26ec+6Qc26tvC/45zdb58fOuVrnXJWk50P2fVDecKje/mtsrJ55SV5i4Fz/+XX+fjdIGqfWz/fjYlTL18CHJf3LOfcvf9ln5X3GfqilF29mBfI+w25xzu30z/kX/Nnb/PhK5SXS0iVVtH1Ij3pA0v+YWaa8Y/h4K8teJmmTc+5//WO32zlXGe4YyDt/Iz2XJO/a/6tzbrFzbr+kW+VVIPUJXci8ytBrJX3f/zxfLils/7cTiP0zkm5zztX4+75d0nXW9rC6ls4lAOjUSBgBQGzd65zL8n9ypKN3omkchvHvkGWrGx/4/wmvkfcXb5nZ/9ixYUm18v7S2uIQr2bbmiPpt5L+T9JmM/uTmWW0sN4L8v7yf57/eK68Lxbn+89PROgdqvbK+8t7OL1D45VXTXKySuUlFzaGHKs/yqtiaRS6r1z5FTQhy/9HTb+gb/e/uDdq7bWcaqWSvh6SdKyVVwkTOnyjOuyarWvrNYVus1Te0KHQGKbJ6ycleQmJQZLeMrMFZnZZs321dB70VtP3ep2avi6FLLfeOeeaLRupJvvxr61qedUfjdo6hh+V9LCftNgv6VEdG5a2XVJBK+vulHSkjWW2hZvvf4nNkVetcbKKFb5PTW9JjUnSRuvU9LiELtvWMYz0em9Lk/fCzAaZ2dPmNQPfJS+h2/yzr6V9f0teVeV884bTfcKP38mrXLnRX+4mHUu8tHW+HxejWr4GSiVNabatc9T6uVAs733Z2XyGn4xa6J+HmyV9QdLFrXyeN1//ZXmfbd+V9HQbSeaWzptGocfgRM6lxuVDz6d6eddR8+Vz5SVnT/Tfh9ZiL5X0WMj7sVJeBWBeC8s3CnsuAUBnR8IIADoY5/U2aWxY+sGQWcWND8wsTl7J/AYzK5X3F+4vyBsSliVpubz/vLa4m2b7/LVzboyk0+V9sflmC+s1Txi9oLYTRq6F6ZHaqJDXLm9YycmqllcJkhOSqMtwzp0eskxovNvkDUU5PWT5THeskXlb3u9rb0u1pLtCYstyzqU452ZEOYbQbVZLWtMshnTn3IckyTm32jl3o7yk3E8kPWJej5q2bJD35a1RiT+tuY2SCs3Mmi0bqSb78bdTLG9oS6MWj6GZFckbQvNhP2mxSV5FyofM68s1W1KRmY0Nt75zbq+8IUvXthLjc5I+GOa4XSuvsqHNW5G3olrh+75skJRtfn8kX4maHpfQZds6hieqpWPefPrvJb0laaBzLkNe8qa1z75jG3Juk3Pu08653vIqS35nx+4YOENeZUmpvCGf//Cnt3q+h4uxlWugWtLfm20r1R3rbRVOtbz3JSuSl+j/juh4+B6UN1TrgTaWa+m8ab5v6cTOpcblQ8+nVHmVes2X3ypvSOaJ/vvQWuzVkj7Y7D0JOudaPZfbOJcAoNMiYQQAnccYM7vGryr4irzExzx5vRuc/CoD8xoUH9e0uiVmNs7Mysy7Jf0eecOwDrew+KvyhiaNlzTfOfemvP/Yl0l6sYV1Nkvq4ye5TsbD8pqmDjWzFEk/OMntyHlD7Z6R9L9mluE3OO1vZs2HsDQuf0ReMu4XZtZLksys0Mw+EOEuN8vrcXKqNN/enyXd4r9/ZmapZnZpsy9m0Y5xvqRd5jWGTjazgJkNM/9OYWb2YTPL9Y9lrb9OS+dXqBmSvmtmuX7i5fvyvsw295q8L41fMrN4M7tGrQ/HbO5hSZea2WT/Gvi6vGvr1QjX/4i8O0wNljfUaaS8pGuNvEbXq+X1bplhXpP0RDMLmtkNZvYdfxvfkneOf9PMekqSmY0ws5n+/L/725tl3q3TE/xz8Nfy7rRY568Tb14z6YCkgL+ftobS/EXSx/3XH+ef30Occ9X+MbjH385weZUy4YY4vd9jGM5mST3NGx7VmnR5/aPqzWyIvLvLRcTMpvgJP8mr9HLyz03n3BJ5n6n3Sfqvc67WX67V872F/bR0DTwo6XIz+4C/naB/jhS1tC3/M+zf8hISPfxz4Tx/P2VmNth/H3vKOz/mNp4fEfq1vF5KLX2eN3paUr6ZfcW8BvXpZlbWQswnci5JXp+6j5vZSPMa3d8tqdJ5Qw5Dt3tYXjXf7WaWYmZDFdJw/iRj/4Oku/xEofzPnyvb2mBr5xIAdGYkjACg83hCXq+QnfK+pF7j969YIel/5X1x3iyvp8wrJ7DdDHmJh53yyvm3y+tBchzn9RBaLOlN59wBf/Jr8pr6tnSb8ln+7+1mtvgE4mrc57/lNZ6eI6+R7pwT3UYz/yOvWe0Kea/5EbU+BOTb/n7nmTfk5TlF3qPoL5KGmje84fGTjviY2yX9zd/e9c65hfL6GP1W3mt5R14D2ZPe5okG5H9pu1xeomSNvKqs++Td+l3ympa/aWb18m4vf4Pz+mG15U55/VyWSXpD3nl3Z5j9H5DXQ+hj8o7BVHlfIiONf5W8XjK/8WO/XNLlIed3Wz4q6Xd+hcHRH3lfPBu/vH5Jx4Z91sobDnO1vF4+cs69Kq9K6QJJ75l3h74/ybvjmvxhbhfKq36olFf19h9518UdIbF815/3Hf817fOntfb650v6uLxeXnXyKgUbqztulNfod4OkxyT9wHl9dppv4/0ew3BxvSUvafief26GvUuWvBsA3CRpt7zPsRO5O9U4SZX+ufmkpC8759aEzJ8h77hPD4mrrfM9nLDXgJ9IuVJeVdRWee/vN9X2/88/Iq+y7C15Te+/4k/vJ++82C2vynS/jg2ri4jzerXNbjbEM9xyu+Ulli6XN+RvtaRJrawS0bnkb3u2pO/Jq+raKK8a6IZwy8qrrE3zY7hfXqPvVrUR+6/knQvPmNlueX+UCZsIa6atcwkAOiVr498DAEAHYGa3y7srzIdjHQuA2PKreP4tb4jOx9r6cg8AAHAyqDACAADoRJx3J7Jr5VUqtdcd+QAAQDdDwggAAKCTcc7VOe/W7m+1tayZTbNjd14M/fl3W+tGU0eNqyNo4bjUm9m5J7m9N1vYXvmpjr0jMLNzWzqGsY4NADoThqQBAAAAAACgCSqMAAAAAAAA0AQJIwAAAAAAADQRH+sAIpGTk+P69OkT6zBOiT179ig1NTXWYQAdHtcK0DauEyAyXCtAZLhWgMh0pWtl0aJF25xzueHmdYqEUZ8+fbRw4cJYh3FKzJ07VxMnTox1GECHx7UCtI3rBIgM1woQGa4VIDJd6Voxs3UtzWNIGgAAAAAAAJogYQQAAAAAAIAmSBgBAAAAAACgCRJGAAAAAAAAaIKEEQAAAAAAAJogYQQAAAAAAIAmSBgBAAAAAACgiagljMwsaGbzzex1M3vTzO7wp2eb2bNmttr/3SNaMQAAAAAAAODERbPCaL+kC5xzIySNlHSJmU2Q9B1Js51zAyXN9p8DAAAAAACgg4hawsh56v2nCf6Pk3SlpL/50/8m6apoxdDR/L9X1mjr3iOxDgMAAAAAAKBV5pyL3sbNApIWSRog6f+cc982s1rnXFbIMjudc8cNSzOzmyXdLEl5eXljZs6cGbU428OWvUf0nZf2yTmnYTnxmlQcrxG5AQXiLNahAR1SfX290tLSYh0G0KFxnQCR4VoBIsO1AkSmK10rkyZNWuScGxtuXlQTRkd3YpYl6TFJX5T0ciQJo1Bjx451CxcujGqM7WFj3T79ZNZLmrcloE27GpSfEdTUccW6YXyxCjKTYx0e0KHMnTtXEydOjHUYQIfGdQJEhmsFiAzXChCZrnStmFmLCaN2uUuac65W0lxJl0jabGYFfmAFkra0RwwdQUFmsq4akKiXvz1Jf/rIGA3OT9ev56zW2T+eo08/sFBzV23RkSPRT+ABAAAAAAC0Jj5aGzazXEkHnXO1ZpYs6UJJP5H0pKSPSvqx//uJaMXQUcUH4nTx6fm6+PR8Ve/Yqxnzq/Twwmo9u2KzirOTdeP4Ek0ZU6zc9KRYhwoAAAAAALqhqCWMJBVI+pvfxyhO0sPOuafN7DVJD5vZJyVVSZoSxRg6vOLsFH3rkiH6yoWD9MyKTaqYV6Wf/meVfvHs2/rA6fkqLyvVhH7ZMqPXEQAAAAAAaB9RSxg555ZJGhVm+nZJk6O1384qMT5Olw3vrcuG99Y7W+o1Y36VHllUo6eXbVS/3FSVl5Xq2tGFykpJjHWoAAAAAACgi2uXHkY4MQN6pel7lw1V5bTJ+t8pI5SVnKAfPb1CZXfP1tcffl2L1u1UezQrBwAAAAAA3VM0h6ThfQomBHTtmCJdO6ZIKzbs0vT56/T4kg36x+IaDclPV/mEUl01srfSgwmxDhUAAAAAAHQhVBh1EkN7Z+jOq85Q5bTJuueaMxSIM33v8eUqu3u2bn30DS1fXxfrEAEAAAAAQBdBhVEnk5oUrxvHl+iGccVaVlOnisp1emxJjWbMr9KI4iyVl5Xo8uG9lZwYiHWoAAAAAACgkyJh1EmZmUYUZ2lEcZZuu3SoHltco4rKKn3rkWX60dMrdO3oIpWXlWhgXnqsQwUAAAAAAJ0MCaMuIDM5QR87u68+elYfLVi7UxWV6zS9skr3v7pW4/tkq3xCiS4Zlq+keKqOAAAAAABA20gYdSFmpvF9szW+b7a+f9l+PbKoRtPnV+nLM5cqOzVRU8YU6cbxJeqTkxrrUAEAAAAAQAdGwqiL6pmWpM+c31+fPrefXnl3m6ZXVum+l9fojy++p3MH5qi8rESTT8tTQoC+5wAAAAAAoCkSRl1cXJzp3IG5OndgrjbvatDDC6o1Y36VbnlwsXqlJ+mGccWaOr5EhVnJsQ4VAAAAAAB0ECSMupG8jKC+OHmgPjdpgOau2qKKyir95vl39Nvn39EFQ3qpvKxU5w3KVSDOYh0qAAAAAACIIRJG3VAgzjT5tDxNPi1PNTv3aub8as1cUK3nVi5QYVaybhxfrOvHFatXejDWoQIAAAAAgBiggU03V9QjRd/4wGC9dusF+l35aPXNSdW9z7yts+6Zo89VLNIr72zTkSMu1mECAAAAAIB2RIURJEkJgTh96IwCfeiMAq3Ztkcz5ldp1sJq/euNTeqbk6qbxpfoujFF6pGaGOtQAQAAAABAlFFhhOP0zUnVtA+dptdunaxfTh2pnLRE3fWvlSq7Z7a++tBSLVy7Q85RdQQAAAAAQFdFhRFaFEwI6KpRhbpqVKFWbdqt6ZXr9Oji9XpsyXoNyktTeVmprh5dqIxgQqxDBQAAAAAApxAVRojI4Px03XHlMFXeNlk/ufYMBRMC+sGTb6rsrtn69iPLtKymNtYhAgAAAACAU4QKI5yQlMR4TR1XoqnjSrSsplbTK6v0xNINemhhtc4ozFR5WYmuGNlbKYmcWgAAAAAAdFZ8q8dJG16UpeFFWZp26Wl6Ysl6PTivSt959A3d9c+Vunp0oW4qK9GQ/IxYhwkAAAAAAE4QCSO8bxnBBH3kzD768IRSLa7aqYp5VZq5oFoPvLZOY0t7qHxCiT44rEDBhECsQwUAAAAAABGghxFOGTPTmNJs/XzqSFXeOlnfvfQ0bd9zQF996HVNuGe27nx6hd7bWh/rMAEAAAAAQBuoMEJU9EhN1KfO7adPntNXr727XRWVVbr/1bW67+U1Oqt/T5WXleqioXlKjCdnCQAAAABAR0PCCFFlZjprQI7OGpCjLbsbNGthjaZXVunz0xcrJy1JU8cV6YZxJSrOTol1qAAAAAAAwEfCCO2mV3pQn580QLec318vrt6qinlV+v3cd/W7ue9q4qBclZeVatKQXgrEWaxDBQAAAACgWyNhhHYXiDNNGtxLkwb30obafZq5oFoz51fpUw8sVO/MoKaOK9HUccXKzwzGOlQAAAAAALolEkaIqd5ZyfraRYP0xQsGaPbKLaqoXKdfPPe2fj1ntS48rZfKy0p1zoAcxVF1BAAAAABAuyFhhA4hIRCnS4bl65Jh+Vq3fY+mz6/SrIU1+u+bm1WSnaKbyko0ZUyReqYlxTpUAAAAAAC6PG5RhQ6ntGeqbv3gaXrt1gv0qxtGKj8zqB//+y2dec8cfWnGElW+t13OuViHCQAAAABAl0WFETqspPiArhxZqCtHFmr15t2aPr9K/1hUoydf36ABvdJUXlaia0YVKTMlIdahAgAAAADQpVBhhE5hYF66fnD56aqcdqF+dt1wpSXF646nVqjsnuf0zVmva0nVTqqOAAAAAAA4RagwQqeSnBjQlLHFmjK2WMvX12n6/Co9vmS9Zi2q0dCCDJVPKNGVIwuVlsSpDQAAAADAyaLCCJ3WsMJM3X31GaqcNll3XjVMTtJtjy1X2V3P6bbH3tCKDbtiHSIAAAAAAJ0SZRjo9NKDCfrwhFKVl5VoSXWtpldW6ZFFNaqorNKokiyVl5XqsuEFCiYEYh0qAAAAAACdAhVG6DLMTKNLeujeKSM0f9qF+v5lQ7Vr30F9Y9brGn/Xc/rhUyv0zpb6WIcJAAAAAECHR4URuqTMlAR94py++vjZfVS5ZocqKqv093lr9ddX1qisb7bKJ5TqA6fnKSmeqiMAAAAAAJojYYQuzcw0oV9PTejXU9vqh2rWwhpNn79OX5qxRD1TEzVlbLFuGl+ikp4psQ4VAAAAAIAOg4QRuo2ctCR9dmJ/fea8fnrpnW2qmLdOf37pPf3hhXd13qBclZeVaPKQXooPMFITAAAAANC9kTBCtxMXZzp/UK7OH5SrjXX79NCCas2cX63P/H2R8jKSdMO4Et0wvlgFmcmxDhUAAAAAgJggYYRurSAzWV+5cJC+MGmAnl+1VRWV6/TrOav1mzmrNfm0PJWXlei8gbmKi7NYhwoAAAAAQLshYQRIig/E6aKhebpoaJ6qd+zVjPlVenhhtZ5dsVlFPZJ14/gSXT+2WLnpSbEOFQAAAACAqKNZC9BMcXaKvnXJEL36ncn67U2jVNwjRT/77yqd9ePZ+vz0xXr13W1yzsU6TAAAAAAAooYKI6AFifFxumx4b102vLfe2VKvGfOr9MiiGv1z2Ub1y0nVTWUlum5MkbJSEmMdKgAAAAAApxQVRkAEBvRK0/cuG6rKaZP18+tHqEdqou7850qNv3u2vvbwUi1at5OqIwAAAABAl0GFEXACggkBXTO6SNeMLtLKjbs0vbJKjy1Zr0cXr9eQ/HSVl5XoqlGFSg8mxDpUAAAAAABOGhVGwEk6rSBDP7pqmCqnTdY915yhQJzpe0+8qbK7Z+vWR5dp+fq6WIcIAAAAAMBJocIIeJ9Sk+J14/gS3TCuWMtq6lRRuU6PLVmvGfOrNaIoU+VlpbpsRIFSErncAAAAAACdA99ggVPEzDSiOEsjirN026VD9djiGlVUVulb/1imH/1zha4dXaSbyko0KC891qECAAAAANAqEkZAFGQmJ+hjZ/fVR8/qo4Xrdqpi3jpNr6zS/a+u1fg+2bqprESXDMtXMCEQ61ABAAAAADgOCSMgisxM4/pka1yfbH3/8gN6ZFG1Kiqr9JWHlqrHUwmaMrZYN44vUd+c1FiHCgAAAADAUSSMgHaSnZqom8/rr0+d00+vvrtdFZXr9JeX1+hPL76ncwbkqLysRBcOzVNCgF70AAAAAIDYilrCyMyKJT0gKV/SEUl/cs79ysxul/RpSVv9Rac55/4VrTiAjiYuznTOwBydMzBHm3c16OEF1Zoxv0qfrVis3PQk3TCuWDeML1FhVnKsQwUAAAAAdFPRrDA6JOnrzrnFZpYuaZGZPevP+4Vz7t4o7hvoFPIygvri5IH63KQBmrtqiyoqq/Tb59/R/z3/jiYN7qXyCSU6f1AvBeIs1qECAAAAALqRqCWMnHMbJW30H+82s5WSCqO1P6AzC8SZJp+Wp8mn5alm5149tKBaMxdUa/b9C1WYlawbxxfr+rHF6pURjHWoAAAAAIBuoF2apZhZH0mjJFX6k75gZsvM7K9m1qM9YgA6i6IeKfr6xYP16ncu0O/LR6tvTqrufeZtnfXjOfrsg4v08uptOnLExTpMAAAAAEAXZs5F94unmaVJekHSXc65R80sT9I2SU7SjyQVOOc+EWa9myXdLEl5eXljZs6cGdU420t9fb3S0tJiHQY6mU17jmhu9SG9vP6g6g9KeSmmicUJOqcwXumJXXO4GtcK0DauEyAyXCtAZLhWgMh0pWtl0qRJi5xzY8PNi2rCyMwSJD0t6b/OuZ+Hmd9H0tPOuWGtbWfs2LFu4cKF0Qmync2dO1cTJ06MdRjopBoOHtZ/lm9SReU6LVi7U4mBOH3ojHyVTyjV2NIeMus6ySOuFaBtXCdAZLhWgMhwrQCR6UrXipm1mDCK5l3STNJfJK0MTRaZWYHf30iSrpa0PFoxAF1NMCGgq0YV6qpRhXp7825Nr6zSPxbV6PGlGzQoL03lZaW6alShMpMTYh0qAAAAAKATi+Zd0s6W9BFJb5jZUn/aNEk3mtlIeUPS1kr6TBRjALqsQXnpuv2K0/WtSwbr6dc3qqJynX7w5Ju6598rdcWI3iovK9WI4qxYhwkAAAAA6ISieZe0lyWFGx/zr2jtE+iOUhLjdf24Yl0/rlhv1NRp+vx1enzJBj28sEbDCjNUXlaqK0b0VmpSNPPDAAAAAICupF3ukgagfZxRlKl7rhmuytsm60dXnq5Dh51uffQNld09W997fLlWbtwV6xABAAAAAJ0AJQdAF5QRTNBHzuyjD08o1eKqnaqYV6WHFlbr7/PWaUxpD5WXlehDZxQomBCIdagAAAAAgA6ICiOgCzMzjSnN1s+njlTlrZP13UtP0849B/S1h1/XhHtm686nV+jdrfWxDhMAAAAA0MFQYQR0Ez1SE/Wpc/vpk+f01WvvbVdFZZXuf3Wt7nt5jc7s11PlE0p08dB8JcaTRwYAAACA7o6EEdDNmJnO6p+js/rnaMvuBs1aWKPplVX6wvQlyklL1PVji3Xj+BIVZ6fEOlQAAAAAQIyQMAK6sV7pQX1+0gDdcn5/vbh6qyrmVekPL7yr37/wrs4flKvyslJNGpyr+ABVRwAAAADQnZAwAqBAnGnS4F6aNLiXNtTu08wF1Zo5v0qffmChCjKDumFciaaOK1Z+ZjDWoQIAAAAA2gEJIwBN9M5K1tcuGqQvXjBAs1duUUXlOv3iubf16zmrNXlIL5VPKNW5A3IUF2exDhUAAAAAECUkjACElRCI0yXD8nXJsHyt275HM+ZXa9bCaj2zYrNKslN04/gSTRlbpJy0pFiHCgAAAAA4xWhMAqBNpT1T9Z0PDtGrt16gX984SgWZQf3kP2/pzHtm64szlmjee9vlnIt1mAAAAACAU4QKIwARS4oP6IoRvXXFiN56Z8tuVVRW6R+LavTU6xvUPzdV5WWlunZ0kTJTEmIdKgAAAADgfaDCCMBJGdArXT+4/HRVTrtQP7tuuNKDCfrh0ys0/u7n9I1Zr2tJ1U6qjgAAAACgk6LCCMD7kpwY0JSxxZoytlhvbqjT9MoqPb5kvR5ZVKOhBRm6qaxEV40qVFoSHzcAAAAA0FlQYQTglDm9d6buuvoMVd52oe66epicpO8+vlxldz2naY+9oTc31MU6RAAAAABABPiTP4BTLi0pXuVlpbppfImWVtce7XU0vbJKI4uzVF5WosuG91ZyYiDWoQIAAAAAwqDCCEDUmJlGlfTQvVNGaP60C/X9y4Zqd8NBffORZSq7+znd8dSbemfL7liHCQAAAABohgojAO0iMyVBnzinrz5+dh9VrtmhisoqPThvnf7fK2tV1jdbN5WV6JJh+UqKp+oIAAAAAGKNhBGAdmVmmtCvpyb066lt9UP1iD9U7cszlyo7NVFTxhbppvElsQ4TAAAAALq1E0oYmVmcpDTn3K4oxQOgG8lJS9It5/fXzef208vvbFNF5Trd99Ia/fGF95QSL5UsfVG9s5LVOyvo/c5MPvo8LyOohACjagEAAAAgGtpMGJnZdEm3SDosaZGkTDP7uXPuZ9EODkD3EBdnOm9Qrs4blKtNdQ16etkGvfbGallqstbXNmhx1U7V7j3YdB2TeqUHjyWTspLVOzPkcVayeqQkyMxi9KoAAAAAoPOKpMJoqHNul5mVS/qXpG/LSxyRMAJwyuVnBvWpc/tpwOEqTZw47uj0PfsPaWPdPq2vbdDG2n3aUOs/rtun5evr9MyKzTpw6EiTbQUT4ppUJRVkJqswK7lJ1VIwgZ5JAAAAANBcJAmjBDNLkHSVpN865w6amYtuWADQVGpSvAb0SteAXulh5zvntH3PAW3wk0kbahu833Xe47mrtmrL7v3HrZedmtgsmeQ97p3lPc9NT1IgjiolAAAAAN1LJAmjP0haK+l1SS+aWakkehgB6FDMTDlpScpJS9Lwoqywy+w/dFib6/b7SST/p85LLK3bvkevvbtd9fsPNVknPs6UlxFUYVayCkKGvxWGJJYygvEMfQMAAADQpbSaMPKbXG92zhWGTKuSNCnagQHAqZYUH1BJzxSV9ExpcZldDQe1oXafNtY2aL2fVNpY5z1etG6n/rlsow4daVpkmZYUr4LMlnopBZWfGVRSPEPfAAAAAHQerSaMnHNHzOwLkh4OmeYkHWp5LQDovDKCCcrIT9CQ/Iyw8w8fcdpWv/9YMqlZYmn5+jpt33PguPVy05OOa8od+rxnaqLiGPoGAAAAoIOIZEjas2b2DUkPSdrTONE5tyNqUQFABxXwh6jlZQQ1uqRH2GUaDh7WRn+oW2hiaUPdPq3avFtzV23VvoOHm6yTGIjzhrxlekPfGptzF2Q2DodLVlpSJB/ZAAAAAPD+RfLt4xP+78+HTHOS+p36cACg8wsmBNQ3J1V9c1LDznfOqXbvQa33q5Ka91N67d3t2ryrQc1GvikjGH+0GXfvrOMTS3kZQSUE4trhFQIAAADo6tpMGDnn+rZHIADQXZiZeqQmqkdqooYVZoZd5tDhI9q8e//xd33zE0sL1+1U3b6DTdaJMykvI3i0n1Khn0gKHQbXIyWBBt0AAAAA2tRiwsjMLnDOzTGza8LNd849Gr2wAKB7iw/EqdBP+rRkz/5D2li3T+v9ZNLG2mOPl6+v0zNvbtaBw0earBNMiAufTMr0GnT3zkpWMIEG3QAAAEB311qF0fmS5ki6PMw8J4mEEQDEUGpSvAb0SteAXulh5x854rR9zwG/IXdIYsl//Namrdq6e/9x62WnJnrJo8xjd3rzhr15iabc9CQFaNANAAAAdGktJoyccz/wf3+8/cIBAJwqcXGm3PQk5aYnaURxVthl9h86rM11+/1+Svv8Rt0N2li3T2u379Er72zTngNNG3THx5nyM4NNqpIKspJVGJJYygjGM/QNAAAA6MQiuuWOmV0q6XRJwcZpzrkfRisoAED7SIoPqKRnikp6poSd75zTroZDTZNJIX2VFq7bqU3LNupQsw7daUnx6p0VVIFfpVTY5HGy8jODSoynQTcAAADQUbWZMDKzP0hKkTRJ0n2SrpM0P8pxAQA6ADNTZnKCMpMTNCQ/I+wyh484bd29Xxvq9h3fpLtun95YX6cdew40266Uk5YUJpl07HFOWiJVSgAAAECMRFJhdJZzbriZLXPO3WFm/yv6FwEAfAF/iFp+ZlCjS3qEXWbfgcN+lVJDk8TSxroGvbVpt+a8tUUNB5s26E6Mj/MaczfrpRSaWEpNiqhQFgAAAMAJiuR/2g3+771m1lvSdkl9oxcSAKCrSU4MqF9umvrlpoWd75xT7d6DWh+SSPKGwHnPX313mzbvalCzkW/KTE7w7/IWcse3kMRSXnqS4gMMfQMAAABOVCQJo6fMLEvSzyQtlneHtD9HMygAQPdiZuqRmqgeqYkaVpgZdpmDh49o866GJsmkjbXHHi9ct1N1+w42WSfOpLyMkGRSSGKpIDOowqxkZaUkMPQNAAAAaKbVhJGZjZK0UlKBc+4fZva0pKBzrq5dogMAwJcQiFNRjxQV9QjfoFuS6vcf8ppy+0ml0MTSsppa/Xd5gw4cbjr0LTkhoIIsL3nUOzNZBX6FUqGfVOqdlaxgQiDaLw8AAADoUFpMGJnZ9yV9WNIiST81s3ucc3+WtL+9ggMA4ESkJcVrYF66Bualh51/5IjT9j0HjjXnDkksbahr0Fubtmjr7uP/meuZmni0KuloMikksZSTlqRAHFVKAAAA6DpaqzCaKmmkc26vmfWU9B8xFA0A0InFxZly05OUm56kEcVZYZfZf+iwNtftP9o/KTSxtGbbHr3yzjbtOXC4yTrxfuPv1vopZQQT2uEVAgAAAKdGawmjBufcXklyzm03M7qGAgC6vKT4gEp6pqikZ/ihb8457Wo45Dfn3qf1fh+ljbXeXeAWrN2pzbs26lCzDt3pSfFHq5KOSyxlJis/M6jEeP6pBQAAQMfQWsKov5k96T+2Zs/lnLsiqpEBANABmZkykxOUmZyg0woywi5z+IjT1t37Q+765iWT1vuPl9XUaceeA822K+WmJakgK1mFWUG/n5L/2E8s9UxNpEE3AAAA2kVrCaMrmz2/N5qBAADQVQT8IWr5mUGNKe0Rdpl9Bw4fTSR5w94ah8A16K1NuzXnrS1qONi0QXdifNzRyqSCzGQdrDugTSlVRxNLBZnJSk2K5AaoAAAAQOta/F+lc+6F5tPMbLRzbnF0QwIAoOtLTgyoX26a+uWmhZ3vnNPOvQeP9VFq1qT7lXe2afOug3ry3TearJeZnOA34/YSSI29lLxG3cnKS09SfIChbwAAAGjdif4Z8j5Jo6MRCAAAOMbMlJ2aqOzURA0rzAy7zHNznteQUWXaUNvg91M6VqVUs3Of5q/ZoV0Nh5qsE2dSfkZQBaGNuTObPs5KSWDoGwAAQDd3ogkj/vcIAEAHER9nKuqRoqIe4Rt0S1L9/kPaWNuYTGqaWFpWU6v/Lm/QgcNNh74lJwSO3eEt8/g7vhVkBhVMCET75QEAACCGTjRhdEdUogAAAFGRlhSvgXnpGpiXHnb+kSNO2/bs10a/l1JoYmlD7T6t3Lhb2+r3H7dez9TEo4mkgsxkFTYmk/zhb7lpSYqL4+9MAAAAnVWLCSMz+4Jz7rf+49Odc2865x5vt8gAAEDUxcWZeqUH1Ss9qBHFWWGX2X/osDbV+Xd5C2nSvb62Qe9t3aOXV2/TngOHm6yTEPAafx9LJh2fWMoIJrTDKwQAAMDJaK3C6BOSfus//rvoXQQAQLeUFB9Qac9UlfZMDTvfOadd+w6F3OmtaYPu+Wt2aNOuBh0+4pqsl54UfzR55DXqbppYyssIKjGeBt0AAACxEOmQNGrKAQBAWGamzJQEZaYk6LSCjLDLHD7itGV3w9Gm3M0TS69X12rn3oPNtivlpiUdTSYVZB7ro9TYU6lnaiINugEAAKKgtYRRlpldLSlOUoaZXRM60zn3aFQjAwAAXUYgzlSQmayCzGSNKQ2/zL4Dh49WKW2sbTjanHtjXYNWbtyl51Zu1v5DTRt0J8XHHW3EfTSZ1CyxlJJ4oi0bAQAA0Nr/oF6QdIX/+EVJl4fMc5JaTRiZWbGkByTlSzoi6U/OuV+ZWbakhyT1kbRW0vXOuZ0nEzwAAOg6khMD6p+bpv65aWHnO+e0Y88BbaxrmkxqfPzy6m3avLtBrunIN2WlJPh3e2t6p7fGfkq90pMUH2DoGwAAQKjWEkZPvc8qokOSvu6cW2xm6ZIWmdmzkj4mabZz7sdm9h1J35H07fexHwAA0A2YmXqmJalnWpKGFWaGXebg4SPaVNegjXWhd33zEks1O/epcs0O7W441GSdQJwpLz3pWDLJv9Nb78xjjzOTExj6BgAAupXWEkbfVRtVRK1xzm2UtNF/vNvMVkoqlHSlpIn+Yn+TNFckjAAAwCmQEIhTcXaKirNTWlxmd8PBo5VJG0P6Ka2v3ael1bX69/J9Oni4aZlSSmLg6LC3wjBVSvmZQQUTAtF+eQAAAO2mXQb1m1kfSaMkVUrK85NJcs5tNLNe7REDAACAJKUHE5QeTNCgvPSw848ccdpWv7/Jnd6ONuqu26eVG3drW/3+49bLSUts0k/Ja9TtDYUrzEpWTlqS4uKoUgIAAJ2DueYD/RtnmO2V9E64WZKcc254RDswS5PXD+ku59yjZlbrnMsKmb/TOdcjzHo3S7pZkvLy8sbMnDkzkt11ePX19UpLC9+bAcAxXCtA27hOYufAYaedDU47Gpy2NxzR9n1O2/3nO/Yd0fYGp/2Hm64TMCk7aMoOmnomx3m/g6aeyabsYJx6JpuS40koRQPXChAZrhUgMl3pWpk0adIi59zYcPNaqzBao6aNrk+YmSVI+oekipB+SJvNrMCvLiqQtCXcus65P0n6kySNHTvWTZw48f2E0mHMnTtXXeW1ANHEtQK0jeuk43LOade+QyE9lPZpvV+ltLFun9bVNqhyU4MOH2n6h7v0YPxxDbp7ZwX9ad7QtwQadJ8wrhUgMlwrQGS6y7XSWsLogHNu3clu2LzOkH+RtNI59/OQWU9K+qikH/u/nzjZfQAAAHREZqbMlARlpiRoaO+MsMscOnxEW3bvb5pMqj32eGl1rXbuPdhsu1KvxgbdIYmlgszG3kpBZacm0qAbAAC8b60ljF55n9s+W9JHJL1hZkv9adPkJYoeNrNPSqqSNOV97gcAAKDTiQ/EHa0iGlMafpm9Bw5pQ22DNtY1NuY+VqW0YuMuPbtysw4cOtJknaT4uKOVSQV+ZVJhSGKpd1ZQKYnt0sYSAAB0Yi3+b8E594X3s2Hn3Mvy+h2FM/n9bBsAAKA7SEmM14BeaRrQK3yfBOecduw5oA21/l3f6kKadNft00urt2rL7v1q3rKyR0pCk2RSQVbTxFKv9KACNOgGAKBb489LAAAAnZSZqWdaknqmJemMosywyxw4dESbdx27y9vRO77V7lP1jr2qXLNduxsONVknEGfKzwgeV6V07HGyMpLjGfoGAEAX1mLCyMymOOdmmVlf59ya9gwKAAAAp0ZifJyKs1NUnJ3S4jK7Gg5qo1+V1JhMakwsLaneqX8v36iDh5uWKaUkBo415s48vkl3fmZQwYRAtF8eAACIktYqjG6VNEveXc5Gt084AAAAaG8ZwQRl5CdocH562PlHjjhtq9/v3/WtwW/U3XgHuAat2FCnbfUHjlsvJy2pyV3emtz9LTOonLQkxTH0DQCADqm1hNF2M3teUl8ze7L5TOfcFdELCwAAAB1FXJypV0ZQvTKCGlUSfpmGg4e1qa7Bb87dNLG0estuvfD2Vu07eLjJOgkBO9qI+1hSKVkFWUEVZiWrIDOo9GBCO7xCAADQXGsJo0vlVRb9XdL/tk84AAAA6IyCCQH1yUlVn5zUsPOdc6rbd9Brzu0Pfzv6uHaf5r23XZt2NehIswbd6cF4FTYmkvyhb4Uhj/Mzg0oIxLXDKwQAoHtp7S5pByTNM7OznHNbzSzdm+zq2y88AAAAdAVmpqyURGWlJOr03uEbdB86fERbdu8/WqW0sa6hSU+lxVU7Vbv3YLPtSnnpQRVkHZ9Manzsmt8mDgAAtCmSu6TlmdkzkrIlmZltlfRR59zy6IYGAACA7iQ+EHd0WNrYFpbZe+BQkzu9eXd/856/ub5Oz67YrAOHjjRZJzEgDXnzZQ3OS9fg/GM/uWlJ3OkNAIAWRJIw+pOkrznnnpckM5voTzsremEBAAAAx0tJjNeAXmka0Cst7HznnLbvOdDkTm+vvfG29ibE6/lVWzRrUc3RZXukeI2+vURShgbnp2lQXjp9kwAAUGQJo9TGZJEkOefmmln4wekAAABADJmZctKSlJOWpOFF3rR+h9Zp4sQJkqRt9fv19qbdemvTbr292fs9a1GN9h441pC7MCtZg/PTNSgvXUP83/17pSopPhCLlwQAQExEkjB6z8y+J6/5tSR9WNKa6IUEAAAAREdOWpJyBiTprAE5R6cdOeK0vnafVm3arVWbd3u/N+3Wi29v1SG/C3cgztQvJ1WD8tM1JC/d+52fruIeKYqLY1gbAKDriSRh9AlJd0h61H/+oqSPRy0iAAAAoB3FxZmKs1NUnJ2iC4fmHZ1+4NARrdm2x08i7dKqTbu1rKZW/1y28egyyQkBDcrzhrLRHwkA0JW0mTByzu2U9KV2iAUAAADoMBLj444mgDSi99Hp9fsPaXVjJZL/u6X+SEPyM44mkwblpdEfCQDQaURSYQQAAADAl5YUr1ElPTSqpEeT6eH6Iz28sDpsf6RjzbbT1S+X/kgAgI6HhBEAAABwCrTWHyk0ifR2s/5I8XGmvvRHAgB0MG0mjMzsbOfcK21NAwAAANBUaH+ki8L0R3pr0y697Q9ra6k/0rE7tmVoUH4a/ZEAAO0ikgqj30gaHcE0AAAAABFo0h8pRLj+SLNXbtHDC4/1R8pOTdSgvDT6IwEAoqrFhJGZnSnpLEm5Zva1kFkZkhhkDQAAAJxirfVHWrVp97Gfza33RxriVyX1z01TYnxce78MAEAX0FqFUaKkNH+Z0D997JJ0XTSDAgAAAHBMY3+ks0+yP1Jok+3B9EcCAESgxYSRc+4FSS+Y2f3OuXXtGBMAAACANrTWH+m9bfVHq5He3rxbS6tr9XQb/ZEG56crJy2R/kgAAEmR9TBKMrM/SeoTurxz7oJoBQUAAADg5CTGx2lIfoaG5Gc0mV6//5De3uxVITVWJYXrjxRaidTYIyktiZsrA0B3E8kn/yxJf5B0n6TDbSwLAAAAoANKS4rX6JIeGn2S/ZGG5KdrEP2RAKDbiCRhdMg59/uoRwIAAACg3bXUH6lm5z7/Tm27tGpzvVZt2qUXmvVH6peb6g9pOza0rahHMv2RAKALiCRh9JSZfU7SY5L2N050zu2IWlQAAAAAYiYuzlTSM0UlPU+8P1JKYkAD89I1OC9Ng/Mzjg5xy01PisVLAQCcpEgSRh/1f38zZJqT1O/UhwMAAACgo2qrP1Lo0LbnmvVH6pmaeLQnUmiPJPojAUDH1Oans3Oub3sEAgAAAKBzaqk/0tbd+/X2Zr/J9qbdeitMf6SiHslNGm0Pzk9Xvxz6IwFArLWZMDKz/wk33Tn3wKkPBwAAAEBXkZuepNz01vsjNd6xrbX+SI1D2+iPBADtJ5L6z3Ehj4OSJktaLImEEQAAAIATEml/pFWbWu6PNCSv6R3b6I8EAKdeJEPSvhj63MwyJf09ahEBAAAA6HZa6o+0u+GgVm9pmkh6duVmPbSw+ugyof2RhuR7yST6IwHA+3Myn6B7JQ081YEAAAAAQHPpwYQ2+yOt2rRLqzbX66EF1dp3sGl/pMYqJPojAcCJiaSH0VPy7oomSQFJp0l6OJpBAQAAAEBrWuuP9NamXceabW/erbmrju+P5PVFStPg/AwNyU9XYRb9kQAgVCQVRveGPD4kaZ1zrqalhQEAAAAgFkL7I118ev7R6fsPHdaabXuaDGtbvG6nnnp9w9FlUv3+SM3v2JaTRn8kAN1TJD2MXjCzPB1rfr06uiEBAAAAwKmTFB9osT/S25vr9fbm1vsjDfaHtdEfCUB3EsmQtOsl/UzSXEkm6Tdm9k3n3CNRjg0AAAAAoiY9mKAxpT00pvRYfyTnnLbVH/ASSJtb7o9UnJ18tBrJSyZlqG9OKv2RAHQZkaTFb5M0zjm3RZLMLFfSc5JIGAEAAADoUszsaH+kcwY27Y9UvXPvsWFtflXS86u26nBIf6T+uWka1Hi3Nr8qif5IADqjSBJGcY3JIt92SaTNAQAAAHQbcXGm0p6pKu2Zelx/pPe27jnWZDuC/kiNQ9vojwSgI4skYfQfM/uvpBn+86mS/h29kAAAAACgc0iKD+i0ggydVpChK0OmN++P9NamXXpmxaYm/ZFy0hI1qLHJdsjwtlT6IwHoACJpev1NM7tG0jnyehj9yTn3WNQjAwAAAIBOqqX+SFvr9+vtTfV6a9Ouo8mkmfPpjwSg42kxYWRmAyTlOedecc49KulRf/p5ZtbfOfduewUJAAAAAJ2dmalXelC90oMn3B8pIWDql5PmVSOFVCTRHwlAtLRWYfRLSdPCTN/rz7s8CvEAAAAAQLfSVn+k0CTSonU79WSY/kihTbbpjwTgVGgtYdTHObes+UTn3EIz6xO9kAAAAAAAof2RQjX2R1q1abffbHuX/vvmJs1cEL4/UmMyif5IAE5Ea58WwVbmJZ/qQAAAAAAAbYukP1JjMil8f6QMDc5P0+D8DA3OS1e/3FQlBOiPBKCp1hJGC8zs0865P4dONLNPSloU3bAAAAAAAJFqqz/SW5t26+1Nu/XWZu/386u2HNcfKTd+v5KKt2tCv2yZ0RcJ6O5aSxh9RdJjZlauYwmisZISJV0d5bgAAAAAAO9TaH+kD7TRH+m1d3brxj/PU//cVN1UVqprRxcqKyUxhtEDiKUWE0bOuc2SzjKzSZKG+ZP/6Zyb0y6RAQAAAACiIlx/pP/Ofl67MgeoorJKP3p6hX76n7d02fDeKp9QolHFWVQdAd1Mmx3PnHPPS3q+HWIBAAAAAMRIUsA0ZWyxpowt1psb6jS9skqPL1mvfyyu0WkFGSovK9FVowqVRuNsoFugsxkAAAAAoInTe2fqrqvPUOVtF+quq70BJ999fLnK7npO0x57Q8vX18U4QgDRRmoYAAAAABBWWlK8ystKddP4Ei2trlVFZZX+sahG0yurNKI4S+VlJbp8eG8lJwZiHSqAU4wKIwAAAABAq8xMo0p66N4pIzR/2oX6/mVDVd9wUN96ZJnG3/2cbn/yTa3evDvWYQI4hagwAgAAAABELDMlQZ84p68+fnYfzV+zQw9WVqmicp3uf3WtxvfNVnlZiS4Zlq+keKqOgM6MhBEAAAAA4ISZmcr69VRZv57aVj9Uj/hD1b48c6myUxM1ZWyRbhpfotKeqbEOFcBJiNqQNDP7q5ltMbPlIdNuN7P1ZrbU//lQtPYPAAAAAGgfOWlJuuX8/pr7jYl64BPjNa5PD9330hqd/7O5+shfKvWf5Zt08PCRWIcJ4AREs8Lofkm/lfRAs+m/cM7dG8X9AgAAAABiIC7OdN6gXJ03KFeb6hr00IJqzVxQpVseXKS8jCRNHVusG8aXqHdWcqxDBdCGqCWMnHMvmlmfaG0fAAAAANBx5WcG9eULB+rzk/rr+VVbVVG5Tr95/h399vl3dMGQXiovK9V5g3IViLNYhwogjFj0MPqCmf2PpIWSvu6c2xmDGAAAAAAA7SA+EKeLhubpoqF5qt6xVzMXVOmhBdV6buUWFWYl66ayEk0ZW6Re6cFYhwoghDnnordxr8LoaefcMP95nqRtkpykH0kqcM59ooV1b5Z0syTl5eWNmTlzZtTibE/19fVKS0uLdRhAh8e1ArSN6wSIDNcKEJn2vFYOHXFavOWwnq86qJU7jihg0ui8gCYVJ+i07DiZUXWEjqsr/bsyadKkRc65seHmtWvCKNJ5zY0dO9YtXLjw1AcYA3PnztXEiRNjHQbQ4XGtAG3jOgEiw7UCRCZW18q7W+s1o7JKsxbVqG7fQfXLSdWN40t03Zgi9UhNbPd4gLZ0pX9XzKzFhFHU7pLWQiAFIU+vlrS8pWUBAAAAAF1f/9w0ffeyoaqcNlk/v36EeqQm6q5/rVTZPbP11YeWauHaHYpmoQOA8KLWw8jMZkiaKCnHzGok/UDSRDMbKW9I2lpJn4nW/gEAAAAAnUcwIaBrRhfpmtFFemvTLk2vrNKji9frsSXrNTgvXeUTSnTVqEJlBBNiHSrQLUTzLmk3hpn8l2jtDwAAAADQNQzJz9APrxymb18yRE+9vkEPVq7T9594U/f86y1dObK3ystKdUZRZqzDBLq0WNwlDQAAAACANqUmxeuG8SW6YXyJltXUqmJelR5ful4zF1RreFGmystKdPmI3kpJ5KstcKq1aw8jAAAAAABOxvCiLP3kuuGqnHah7rjidDUcPKxv/+MNld01Wz94YrlWbdod6xCBLoU0LAAAAACg08hMTtBHz+qj/zmzVAvX7VTFvHWaMb9af3ttncaW9lD5hBJ9cFiBggmBWIcKdGokjAAAAAAAnY6ZaVyfbI3rk63vX35Ajyyq1vTKKn31odd1x1MrNGVMkW4qK1XfnNRYhwp0SiSMAAAAAACdWnZqom4+r78+dU4/vfbedlVUrtP/e2Wt/vzSGp09oKfKy0p10dA8JQToygJEioQRAAAAAKBLiIsznT0gR2cPyNGWXQ16eGG1Zsyv1ucqFis3PUlTxxbrhvHFKuqREutQgQ6PhBEAAAAAoMvplRHUFy4YqM9OHKAX3t6iinlV+r+57+j/5r6jiYNyVV5WqklDeikQZ7EOFeiQSBgBAAAAALqsQJzpgiF5umBInmp27tVDC6r10IJqfeqBheqdGdQN40s0dVyx8jKCsQ4V6FBIGAEAAAAAuoWiHin6+sWD9aXJAzV75WZVVFbp58++rV/NXq2LTstT+YQSnd0/R3FUHQEkjAAAAAAA3UtCIE6XDCvQJcMKtHbbHs2YX6WHF1brP29uUmnPFN00vkTXjSlSz7SkWIcKxAwt4gEAAAAA3VafnFTd+qHT9Nqtk/WrG0YqLz2oe/79ls68Z46+PHOJ5q/ZIedcrMME2h0VRgAAAACAbi+YENCVIwt15chCvb15t6ZXVukfi2v0xNINGtArTeVlJbpmdJEykxNiHSrQLqgwAgAAAAAgxKC8dN1+xemaP+1C/fS64UpNitcdT61Q2d3P6ZuzXtfS6lqqjtDlUWEEAAAAAEAYyYkBXT+2WNePLdby9XWqqKzSE0vXa9aiGp3eO0PlZaW6cmRvpSbx1RpdDxVGAAAAAAC0YVhhpu655gxVTpusH101TIePOE177A2V3T1b3338Da3YsCvWIQKnFGlQAAAAAAAilB5M0EcmlOrDZSVaXFWrisp1mrWwRg/Oq9KokiyVl5XqsuEFCiYEYh0q8L5QYQQAAAAAwAkyM40p7aGfXz9SldMm63uXDVXdvoP6xqzXVXb3bP3wqRV6Z0t9rMMEThoVRgAAAAAAvA9ZKYn65Dl99Ymz+2jeeztUUblOf5+3Vn99ZY0m9MtWeVmpPnB6vhLjqdlA50HCCAAAAACAU8DMdGb/njqzf09t3b1fsxZVa3pllb44Y4ly0hI1ZWyxbhxXopKeKbEOFWgTCSMAAAAAAE6x3PQkfW7iAN1yXn+9uHqrKiqr9McX3tUfXnhX5w3MVXlZiS4Y0kvxAaqO0DGRMAIAAAAAIEri4kwTB/fSxMG9tLFun2bOr9bMBVW6+e+LlJ8R1NRxxbphfLEKMpNjHSrQBAkjAAAAAADaQUFmsr560SB98YIBmvPWFlVUVunXc1brN3NWa/JpeSovK9F5A3MVF2exDhUgYQQAAAAAQHuKD8Tp4tPzdfHp+aravlczFlTp4QXVenbFZhVnJ+vG8SWaMqZYuelJsQ4V3RiDJQEAAAAAiJGSnin69iVD9OqtF+g3N45SYVayfvqfVTrrx7P1hemL9dq72+Wci3WY6IaoMAIAAAAAIMaS4gO6fERvXT6it97ZUq/plVV6ZFG1nl62Uf1yU3XT+BJdN6ZIWSmJsQ4V3QQVRgAAAAAAdCADeqXp+5cP1fzbLtS9U0YoKzlBd/5zpcrunq2vPbxUi9btpOoIUUeFEQAAAAAAHVAwIaDrxhTpujFFWrFhl6bPX6fHFq/Xo4vXa0h+usonlOqqkb2VHkyIdajogqgwAgAAAACggxvaO0N3XnWGKm+7UHdffYbizPS9x5er7O7ZuvXRN7R8fV2sQ0QXQ4URAAAAAACdRFpSvG4qK9GN44v1ek2dKuat02NLajRjfpVGFGepvKxElw/vreTEQKxDRSdHwggAAAAAgE7GzDSyOEsji7P03UuH6tElNaqorNK3HlmmHz29QteOLlJ5WYkG5qXHOlR0UiSMAAAAAADoxDJTEvTxs/vqY2f10fw1O1RRWaXplVW6/9W1Gt8nW+UTSnTJsHwlxVN1hMiRMAIAAAAAoAswM5X166myfj21vX6/HllUo+nzq/TlmUuVnZqoKWOKdOP4EvXJSY11qOgESBgBAAAAANDF9ExL0mfO769Pn9tPr7y7TRXzqnTfy2v0xxff07kDc1ReVqLJp+UpIcC9sBAeCSMAAAAAALqouDjTuQNzde7AXG3e1aCHFlRrxvwq3fLgYvVKT9IN44o1dXyJCrOSYx0qOhgSRgAAAAAAdAN5GUF9afJAfW5if81dtVUVlev0m+ff0W+ff0cXDOml8rJSnTcoV4E4i3Wo6ABIGAEAAAAA0I3EB+J04dA8XTg0T9U79uqhBdWauaBaz61coMKsZN04vljXjytWr/RgrENFDDFYEQAAAACAbqo4O0Xf+MBgvfqdC/S78tHqk5Oie595W2fdM0efq1ikV97ZpiNHXKzDRAxQYQQAAAAAQDeXGB+nD51RoA+dUaD3ttZrxvwqzVpUo3+9sUl9c1J10/gSXTumSNmpibEOFe2ECiMAAAAAAHBUv9w03XbpUM27dbJ+MXWEeqYm6q5/rdSEe2brqw8t1cK1O+QcVUddHRVGAAAAAADgOMGEgK4eVaSrRxVp1abdml65To8uXq/HlqzXoLw0lZeV6urRhcoIJsQ6VEQBFUYAAAAAAKBVg/PTdceVw1R522T95NozFEwI6AdPvqmyu2br248s07Ka2liHiFOMCiMAAAAAABCRlMR4TR1XoqnjSrSsplbTK6v0xNINemhhtc4ozFR5WYmuGNlbKYmkGzo73kEAAAAAAHDChhdlaXhRlqZdepoeX7JeFfOq9J1H39Bd/1ypq0cX6qayEg3Jz4h1mDhJJIwAAAAAAMBJywgm6H/O7KOPTCjVonU7VVFZpZkLqvXAa+s0prSHystK9KEzChRMCMQ6VJwAehgBAAAAAID3zcw0tk+2fjF1pCpvnazvXnqaduw5oK89/Lom3DNbdz69Qu9trY91mIgQFUYAAAAAAOCU6pGaqE+d20+fPKevXnt3uyoqq3T/q2t138trdFb/niovK9VFQ/OUGE8dS0dFwggAAAAAAESFmemsATk6a0COtuxu0KyFNZpeWaXPT1+snLQkTR1XpBvGlag4OyXWoaIZEkYAAAAAACDqeqUH9flJA3TL+f314ttbVVG5Tr+f+65+N/ddTRyUq/KyUk0a0kuBOIt1qBAJIwAAAAAA0I4CcaZJQ3pp0pBe2lC7TzMXVGvm/Cp96oGF6p0Z1NRxJZo6rlj5mcFYh9qtkTACAAAAAAAx0TsrWV+7aJC+eMEAzV65RRWV6/SL597Wr+es1oWn9VJ5WanOGZCjOKqO2l3UEkZm9ldJl0na4pwb5k/LlvSQpD6S1kq63jm3M1oxAAAAAACAji8hEKdLhuXrkmH5Wrttj2YsqNKshTX675ubVZKdopvKSjRlTJF6piXFOtRuI5rtyO+XdEmzad+RNNs5N1DSbP85AAAAAACAJKlPTqpu/eBpeu3WC/SrG0YqPzOoH//7LZ15zxx9acYSVb63Xc65WIfZ5UWtwsg596KZ9Wk2+UpJE/3Hf5M0V9K3oxUDAAAAAADonJLiA7pyZKGuHFmo1Zt3q6KySv9YXKMnX9+gAb3SdNP4El07ukiZKQmxDrVLimaFUTh5zrmNkuT/7tXO+wcAAAAAAJ3MwLx03X7F6Zo/7UL99LrhSkuK1w+fXqHxdz+nb8x6XUuqdlJ1dIpZNA+oX2H0dEgPo1rnXFbI/J3OuR4trHuzpJslKS8vb8zMmTOjFmd7qq+vV1paWqzDADo8rhWgbVwnQGS4VoDIcK2gs1m367Cerz6k1zYc0v7DUkl6nCYVx2tC73glx0evSXZXulYmTZq0yDk3Nty89k4YrZI00Tm30cwKJM11zg1uaztjx451CxcujFqc7Wnu3LmaOHFirMMAOjyuFaBtXCdAZLhWgMhwraCz2t1wUE8s3aAH563TW5t2KzUxoKtGFaq8rFRDe2ec8v11pWvFzFpMGEWth1ELnpT0UUk/9n8/0c77BwAAAAAAXUh6MEEfnlCq8rISLamuVcW8Kj2yqEYVlVUaVZKl8rJSXTa8QMGEQKxD7VSi1sPIzGZIek3SYDOrMbNPyksUXWRmqyVd5D8HAAAAAAB4X8xMo0t66H+vH6HKaZP1vcuGate+g/rGrNc1/q7n9MOnVuidLfWxDrPTiOZd0m5sYdbkaO0TAAAAAAAgKyVRnzynrz5xdh9Vrtmhisoq/X3eWv31lTUq65ut8gml+sDpeUqKp+qoJe09JA0AAAAAAKBdmJkm9OupCf16alv9UM1aWKPp89fpSzOWqGdqoqaMLdZN40tU0jMl1qF2OCSMAAAAAABAl5eTlqTPTuyvz5zXTy+9s00V89bpzy+9pz+88K7OG5Sr8rISTR7SS/GBqHXv6VRIGAEAAAAAgG4jLs50/qBcnT8oVxvr9umhBdWaOb9an/n7IuVlJOmGcSW6YXyxCjKTYx1qTJEwAgAAAAAA3VJBZrK+cuEgfWHSAD2/aqsqKtfp13NW6zdzVmvyaXkqLyvReQNzFRdnsQ613ZEwAgAAAAAA3Vp8IE4XDc3TRUPzVL1jr2bMr9LDC6v17IrNKuqRrBvHl+j6scXKTU+KdajthoQRAAAAAACArzg7Rd+6ZIi+cuEgPbNikyrmVeln/12lXz73ti4+PV/Dkg7rfOdk1rWrjkgYAQAAAAAANJMYH6fLhvfWZcN7650t9Zoxv0qPLKrRcwcO6sOXHlJ6MCHWIUYVrb8BAAAAAABaMaBXmr532VBVTpusb44NdvlkkUTCCAAAAAAAICLBhIAG9AjEOox2QcIIAAAAAAAATZAwAgAAAAAAQBMkjAAAAAAAANAECSMAAAAAAAA0QcIIAAAAAAAATZAwAgAAAAAAQBMkjAAAAAAAANAECSMAAAAAAAA0QcIIAAAAAAAATZAwAgAAAAAAQBPmnIt1DG0ys62S1sU6jlMkR9K2WAcBdAJcK0DbuE6AyHCtAJHhWgEi05WulVLnXG64GZ0iYdSVmNlC59zYWMcBdHRcK0DbuE6AyHCtAJHhWgEi012uFYakAQAAAAAAoAkSRgAAAAAAAGiChFH7+1OsAwA6Ca4VoG1cJ0BkuFaAyHCtAJHpFtcKPYwAAAAAAADQBBVGAAAAAAAAaIKEUTsxs7+a2RYzWx7rWICOysyKzex5M1tpZm+a2ZdjHRPQEZlZ0Mzmm9nr/rVyR6xjAjoyMwuY2RIzezrWsQAdlZmtNbM3zGypmS2MdTxAR2VmWWb2iJm95X9vOTPWMUULQ9LaiZmdJ6le0gPOuWGxjgfoiMysQFKBc26xmaVLWiTpKufcihiHBnQoZmaSUp1z9WaWIOllSV92zs2LcWhAh2RmX5M0VlKGc+6yWMcDdERmtlbSWOfctljHAnRkZvY3SS855+4zs0RJKc652hiHFRVUGLUT59yLknbEOg6gI3PObXTOLfYf75a0UlJhbKMCOh7nqfefJvg//AUICMPMiiRdKum+WMcCAOjczCxD0nmS/iJJzrkDXTVZJJEwAtBBmVkfSaMkVcY4FKBD8ofYLJW0RdKzzjmuFSC8X0r6lqQjMY4D6OicpGfMbJGZ3RzrYIAOqp+krZL+nz/U+T4zS411UNFCwghAh2NmaZL+IekrzrldsY4H6Iicc4edcyMlFUkab2YMdwaaMbPLJG1xzi2KdSxAJ3C2c260pA9K+rzfUgNAU/GSRkv6vXNulKQ9kr4T25Cih4QRgA7F78fyD0kVzrlHYx0P0NH5ZdBzJV0S20iADulsSVf4vVlmSrrAzB6MbUhAx+Sc2+D/3iLpMUnjYxsR0CHVSKoJqex+RF4CqUsiYQSgw/Ab+f5F0krn3M9jHQ/QUZlZrpll+Y+TJV0o6a2YBgV0QM65W51zRc65PpJukDTHOffhGIcFdDhmlurfcET+8JqLJXF3Z6AZ59wmSdVmNtifNFlSl71BT3ysA+guzGyGpImScsysRtIPnHN/iW1UQIdztqSPSHrD780iSdOcc/+KXUhAh1Qg6W9mFpD3x5+HnXPcLhwAcLLyJD3m/e1O8ZKmO+f+E9uQgA7ri5Iq/DukvSfp4zGOJ2rMOW6qAgAAAAAAgGMYkgYAAAAAAIAmSBgBAAAAAACgCRJGAAAAAAAAaIKEEQAAAAAAAJogYQQAAAAAAIAmSBgBABADZpZvZjPN7F0zW2Fm/zKzQad4HxPN7KwTXCfJzJ4zs6VmNvV97r/e/93HzJa/n2012+59ZjbUfzwtZHpE+zGzqxrX95/PNbOxpyq+9vJ+4jazr/nn3TIzm21mpSHzfmZmb5rZz05dtO3rZM79VraVZWafOxXbAgCgMyFhBABAOzMzk/SYpLnOuf7OuaGSpknKO8W7mijpRL80j5KU4Jwb6Zx76BTHc0o45z7lnFvhP53W6sLhXSVpaFsLdXFLJI11zg2X9Iikn4bM+4yk0c65b0Y7CDOLj9KmJ6qFc/8k9pkliYQRAKDbIWEEAED7myTpoHPuD40TnHNLnXMvmednZrbczN5orPLxKyaeblzezH5rZh/zH681szvMbLG/zhAz6yPpFklf9auFzg0NwMyyzexxv8JknpkNN7Nekh6UNNJfp3+zdQb41Uev+/vq70//ppkt8Ld1R2sv3MxON7P5/vaXmdnAZvOvN7Of+4+/bGbv+Y/7m9nL/uO5ZjbWzH4sKdnfVoW/iYCZ/dmvkHnGzJKbbf8sSVdI+lmz1zjFj+vtxmNlZgH/vWh8bZ8J83pSzeyf/jFZHvJ+fd9fb7mZ/clPEjbG/gsze9HMVprZODN71MxWm9mdIdv9cMhx+qOZBdo4rjf67/1yM/tJyPRP+q9prn9cfitJzrnnnXN7/cXmSSryl39SUqqkSjObamZ5ZvaY//peD1e1Y2b1Zva//jkx28xyQ96z/5jZIjN7ycyG+NPvN7Ofm9nzkn7SbFsBM7vXfy3LzOyL/vTJZrbEn/5XM0vyp0d07jffp5mNN7NX/W2+amaD/e2FOz9/LKm/P63TVl0BAHCiSBgBAND+hkla1MK8aySNlDRC0oXyEhsFEWxzm3NutKTfS/qGc26tpD9I+oVfLfRSs+XvkLTErzCZJukB59wWSZ+S9JK/zrvN1qmQ9H/OuRHyqjc2mtnFkgZKGu/HPcbMzmslzlsk/co5N1LSWEk1zea/KKkxuXWupO1mVijpHElNXoNz7juS9vmxlvuTB/oxni6pVtK1zdZ5VdKTkr7Z7DXGO+fGS/qKpB/40z4pqc45N07SOEmfNrO+zeK9RNIG59wI59wwSf/xp//WOTfOn5Ys6bKQdQ44586T9/48Ienz8s6Jj5lZTzM7TdJUSWf7x+mwpHK1wMx6y0u8XCDvPRhn3rC73pK+J2mCpIskDWlhE5+U9G//+FyhY8f0IUm/lvSC/56PlvRmmPVTJS32z78XdOz4/UnSF51zYyR9Q9LvQtYZJOlC59zXm23rZkl9JY3yz80KMwtKul/SVOfcGZLiJX02ZJ1Iz/3Qfb4l6Tzn3ChJ35d0t79MuPPzO5Le9bcV9aorAAA6imiVAQMAgJNzjqQZzrnDkjab2QvykhW72ljvUf/3InlJp0j2c60kOefm+ImKzJYWNrN0SYXOucf8dRr86RdLuljeECdJSpOXtHmxhU29Juk2MyuS9KhzbnXoTOfcJjNL8/dXLGm6pPPkJY8ePW5rx1vjnFvqP14kqU8E60hNj1/jOhdLGm5m1/nPM+W9tjUh670h6V6/qufpkOTEJDP7lqQUSdnyEi1P+fOeDFn3TefcRkkyr5qqWN57M0bSAr8wKVnSllZiHydveONWfzsV8o6Z5CV7dvjTZ8lLmhxlZh+Wlxg5v4VtXyDpfyTJPyfrwixzRFLj8MUHJT1qZmnykoqz/NcgSUkh68zyt9fchZL+4Jw75O9zh5mNkPe+vu0v8zd5SbZf+s8jPfdD95kp6W9+BZGTlOBPP+78DIkfAIBuhYQRAADt701J17Uwr6Vvp4fUtDI42Gz+fv/3YUX273u4/bgTXL5x+j3OuT9GsE8556abWaWkSyX918w+5Zyb02yx1yR9XNIqeVVFn5B0pqTm1Sjh7A95fFhesiUS4Y6fyauQ+W9LKznn3jazMZI+JOkeM3tGXj+g38nrEVRtZrer6fvVuK8jzeI94u/bJP3NOXdrhLG39t60vJLZhZJuk3S+c25/a8ueICfvXK31K3XC2dNSWDr+PGwrYxPpuR+6zx9Jet45d7U/hG2uFP78lPReG/sHAKBLYkgaAADtb46kJDP7dOMEv5fN+fIqc6b6vVxy5VWKzJe0TtJQ8+5ililpcgT72S0pvYV5L8of5mRmE+UN62mxismfV2NmV/nrJJlZiqT/SvqEX1EiMys0rxdSWGbWT9J7zrlfy6u0Gd5CbN/wfy+R1/Npv3MuXHXLQTNLCDO9Na0dl1D/lfTZxu2b2SAzSw1dwB/2tdc596Cke+UN22pMDm3zj0tLycGWzJZ0XeNxNK/fVGkry1dKOt/McszrdXSjvKFh8/3pPcxr9Hx0eJ6ZjZL0R0lX+EMRW4vls/46ATPLCLNMnI69xpskveyfL2vMbIq/rvmVQm15RtItfrwys2x5w8f6mNkAf5mP+K+vNW29x5mS1vuPP9Y4sYXzM9LzBQCALoWEEQAA7cw55yRdLekiM3vXzN6UdLukDfLunrZM0uvyEkvfcs5tcs5VS3rYn1ehY0PAWvOUpKstTNNrf39jzWyZvKa+H41gex+R9CV/nVcl5TvnnpE3bOw1M3tD3h23WvtyPVXScjNbKq+nzgNhlnlJ3tCsF/0hRNWSXm5he3+StMyONb2OxExJ3/QbHvdvZbn7JK2QtNjMlstLsDSvYDlD0nz/9dwm6U7nXK2kP8sbcva4pAUnEJv8O8B9V9Iz/rF+VlKLfaz8IW23Snpe3nmz2Dn3hHNuvbzePJWSnvNfS2PS7Wfyhg/O8s+PJ4/fsiTpy/KG170hb8jX6WGW2SPpdDNbJG8I2w/96eWSPmlmr8urqrsygpd/n6Qqee/p65Ju8oc/ftyP9Q15lVh/aGUbUuvnvuRVgd1jZq9ICm0oftz56ZzbLukV8xqK0/QaANBtmPd/VgAAAHQ1ZpbmnKv3K3Yek/TXxj5Up3Af9c65tFO5TQAAEHtUGAEAAHRdt/vVMsvlNet+PKbRAACAToMKIwAAAAAAADRBhREAAAAAAACaIGEEAAAAAACAJkgYAQAAAAAAoAkSRgAAAAAAAGiChBEAAAAAAACaIGEEAAAAAACAJv4/ocvvasDwwLEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1440x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Data for plotting\n",
    "t = np.arange(0.0, 2.0, 0.01)\n",
    "s = 1 + np.sin(2 * np.pi * t)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 5),)\n",
    "ax.plot(x, y)\n",
    "\n",
    "ax.set(xlabel='Count of cells with the same log2fc per contrast', ylabel='Count of TF-Pairs',\n",
    "       title=' TF-pairs with different filter threshold for ACCQ1_colon_transverse_c5_Microfold_cells')\n",
    "ax.grid()\n",
    "\n",
    "fig.savefig(\"varying_thresholds.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6642cd7",
   "metadata": {},
   "source": [
    "# OLD\n",
    "# -----------------------------------------------------------------------------------------\n",
    "# OLD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ececb4a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9597b944",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analyses with whole diffcombj\n",
    "diff_file_names=read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "\n",
    "normalized_diff_objects = []\n",
    "diff_objects = []\n",
    "\n",
    "for file in diff_file_names:\n",
    "    obj = DiffCombObj().from_pickle(f\"{differential_analysis_path}{file}\")\n",
    "    if \"normalized\" in file:\n",
    "        normalized_diff_objects.append(obj)\n",
    "    else:\n",
    "        diff_objects.append(obj)\n",
    "\n",
    "print(normalized_diff_objects)\n",
    "print(diff_objects) \n",
    "\n",
    "normalized_dfs = []\n",
    "for obj in normalized_diff_objects:\n",
    "    normalized_dfs.append(obj.rules)\n",
    "print(\"Done: Preparing rules of DiffObj\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a11d6ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj = normalized_diff_objects[0]\n",
    "t = normalized_diff_objects[0].rules\n",
    "t[t['right-lobe-of-liver.10_cosine'] > 0.7]\n",
    "obj.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3836030f",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df = []\n",
    "for df in normalized_dfs:\n",
    "\n",
    "    for i, file in enumerate(files_main_mb):\n",
    "        # print(file)\n",
    "        cluster_name = file.split('.pkl')[0]\n",
    "        print(cluster_name)\n",
    "        \n",
    "        # reduce to relevant columns of cluster\n",
    "        cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "        # NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "        # This is important: for right-lob-of-liver-1 ,     #right-lobe-of-liver.10_cosine\n",
    "        logfc_cluster_cols = []\n",
    "        for entry in cluster_cols:\n",
    "            if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "                logfc_cluster_cols.append(entry)\n",
    "        \n",
    "        #print(logfc_cluster_cols)\n",
    "        print(len(logfc_cluster_cols))\n",
    "        #print(logfc_cluster_cols)\n",
    "        \n",
    "        reduced_df = df[logfc_cluster_cols]\n",
    "        print(f'Initial Count: {reduced_df.shape}')\n",
    "        \n",
    "        # Count all entries in a row , which do not have a zero(0.00) in it.\n",
    "        # e.g. 15 cols have 0.00 => val_counts = 0, 10 cols not have a 0.00 => val_counts = 10 \n",
    "        val_counts = reduced_df[~reduced_df.isin([0])].count(axis=1).sort_values()\n",
    "        #print(tmp_val_counts)\n",
    "        # Set threshold \n",
    "        selection_threshold = len(logfc_cluster_cols) # z.B 15 could be varyied\n",
    "        \n",
    "        ## Keep all entries,which do have more values different higher than the threshold\n",
    "        tfs_occ = val_counts[val_counts >= selection_threshold].index\n",
    "        result = reduced_df.loc[tfs_occ]\n",
    "        print(f'Zero filtered: {result.shape}')\n",
    "        \n",
    "        significants = get_significant_log2fc_rules(result, threshold_percent=0.05)\n",
    "        results_df.append(significants)\n",
    "        print(f'Cluster: {cluster_name}: {significants.shape} ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue: {tissue_name} ')\n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af3a075",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df1 = results_df[0]\n",
    "print(df1.shape)\n",
    "df1\n",
    "# found tf-pairs, which have significant changes in comparison to all the other clusters in the tissue\n",
    "# this 10 tf-pairs of cluster 10, show significant changes in comparion to all the other clusters in the tissue right lob of liver. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42fd4e9f",
   "metadata": {},
   "source": [
    "#####  nur cosine werte vergleichbar, daher zum finden von spezifischen tf-cos nur cosines nutzen.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b900b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = compare_obj.rules\n",
    "df = pd.DataFrame(td.loc['EN2-MYBL2'])\n",
    "df.head(50)\n",
    "test\n",
    "\n",
    "# nur cosine werte vergleichbar, daher zum finden von spezifischen tf-cos nur cosines nutzen.  \n",
    "np.percentile(test['right-lobe-of-liver.11_cosine'], 75)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16a033b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c684a4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "test['right-lobe-of-liver.11_cosine'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a48bb61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster 6\n",
    "df1 = results_df[13]\n",
    "print(df1.shape)\n",
    "df1.head(50)\n",
    "df1.loc['Sox17-Dlx4':'TP63-FOSJUNB']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f4388d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filtering if, nan values occure\n",
    "filtered = df[logfc_cluster_cols]\n",
    "\n",
    "val_counts = filtered.count(axis=1).sort_values()\n",
    "#print(val_counts)\n",
    "tfs_occ = val_counts[val_counts >=16].index\n",
    "final = filtered.loc[tfs_occ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b10cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = normalized_dfs[0]\n",
    "cluster_name = \"right-lobe-of-liver.1\"\n",
    "cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "# NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "# This is important: for right-lob-of-liver-1 \n",
    "logfc_cluster_cols = []\n",
    "for entry in cluster_cols:\n",
    "    if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "        logfc_cluster_cols.append(entry)\n",
    "        \n",
    "tmp = df[logfc_cluster_cols]\n",
    "# wtf, tmp[~tmp.isin([0])]??\n",
    "tmp_val_counts = tmp[~tmp.isin([0])].count(axis=1).sort_values()\n",
    "#print(tmp_val_counts)\n",
    "tmp_tfs_occ = tmp_val_counts[tmp_val_counts == len(logfc_cluster_cols)].index\n",
    "result = tmp.loc[tmp_tfs_occ]\n",
    "print(result.shape)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed56edb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = normalized_dfs[0]\n",
    "cluster_name = \"right-lobe-of-liver.1\"\n",
    "cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "# NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "# This is important: for right-lob-of-liver-1 \n",
    "logfc_cluster_cols = []\n",
    "for entry in cluster_cols:\n",
    "    if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "        logfc_cluster_cols.append(entry)\n",
    "        \n",
    "tmp = df[logfc_cluster_cols]\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89acdd63",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure_threshold_1 = utils.get_threshold(result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc']\n",
    "                                        , \"both\", percent=0.05)\n",
    "measure_threshold_2 = utils.get_threshold(result['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc']\n",
    "                                        , \"both\", percent=0.05)\n",
    "print(f'1: {measure_threshold_1}')\n",
    "print(f'1: {measure_threshold_2}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31975340",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure_threshold_1[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3dda76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#t[t['right-lobe-of-liver.10_cosine'] > 0.7]\n",
    "reduced_result_1 = result[(result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc'] > measure_threshold_1[1]) | (result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc'] < measure_threshold_1[0])]\n",
    "reduced_result = reduced_result_1[(reduced_result_1['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc'] > measure_threshold_2[1]) | (reduced_result_1['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc'] < measure_threshold_2[0])]\n",
    "reduced_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70edc13d",
   "metadata": {},
   "outputs": [],
   "source": [
    "f.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed9893bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "m1 = pd.DataFrame(result.mean(axis=1), columns=['mean'])\n",
    "m1['sum'] = result.sum(axis=1)\n",
    "m1.plot.hist(by='mean', bins=100)\n",
    "m1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1782c3ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = filtered.fillna(0)\n",
    "tmp_val_counts = tmp[~tmp.isin([0])].count(axis=1).sort_values()\n",
    "\n",
    "print(tmp_val_counts)\n",
    "tmp_tfs_occ = tmp_val_counts[tmp_val_counts >=16].index\n",
    "result = filtered.loc[tmp_tfs_occ]\n",
    "\n",
    "m1 = pd.DataFrame(result.mean(axis=1), columns=['mean'])\n",
    "m1['sum'] = result.sum(axis=1)\n",
    "m1.plot.hist(by='mean', bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf6b31b0",
   "metadata": {},
   "source": [
    "## Old: Self implemented - Differential analysis - comparing each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f9bb030",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mb market basket analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "print(f\"Files: {files_main_mb}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172636a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diff analysis for all clusters of a tissue:\n",
    "# TODO: what should be compared ? All of a Tissue? or All? Naming?\n",
    "tissue_name = 'right-lobe-of-liver'\n",
    "combObj_to_compare = []\n",
    "for i, file in enumerate(files_main_mb):\n",
    "    print(file)\n",
    "    name_i = file.split('.pkl')[0]\n",
    "    tissue_name = file.split('.')[0]\n",
    "    obj = CombObj().from_pickle(f\"{main_analysis_path}{file}\")\n",
    "    obj.set_prefix(name_i)\n",
    "    #print(obj)\n",
    "    combObj_to_compare.append(obj)\n",
    "    \n",
    "compare_obj = DiffCombObj(combObj_to_compare, measure=\"cosine\", join=\"outer\", fillna=True)\n",
    "compare_obj.to_pickle(f'{differential_analysis_path}{tissue_name}.pkl')\n",
    "compare_obj.normalize()\n",
    "compare_obj.calculate_foldchanges()\n",
    "compare_obj.simplify_rules()\n",
    "compare_obj.to_pickle(f'{differential_analysis_path}{tissue_name}_normalized.pkl')\n",
    "#selection does not work?\n",
    "#selected_std = compare_obj.select_rules()\n",
    "#selected_std.to_pickle(f'{differential_analysis_selection_path}{tissue_name}.pkl')\n",
    "print(\"Done differential analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb1f7b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "td = compare_obj.rules\n",
    "\n",
    "tmp_val_counts = td[td.isin([0])].count(axis=1).sort_values()\n",
    "#tmp_val_counts = td[td.isna()].count(axis=1).sort_values()\n",
    "\n",
    "print(tmp_val_counts)\n",
    "#tmp_tfs_occ = tmp_val_counts[tmp_val_counts >=16].index\n",
    "#result = filtered.loc[tmp_tfs_occ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5266b634",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check original market basket analysis for count of tf-cooccurences\n",
    "print(compare_obj.rules.shape)\n",
    "count_of_all = 0\n",
    "count_of_all_significant = 0\n",
    "for obj in combObj_to_compare:\n",
    "    print(obj.prefix)\n",
    "    print(obj.rules.shape)\n",
    "    obj.simplify_rules()\n",
    "    print(obj.rules.shape)\n",
    "    count_of_all = count_of_all + obj.rules.shape[0]\n",
    "    sig = obj.select_significant_rules()\n",
    "    print(sig.rules.shape)\n",
    "    count_of_all_significant = count_of_all_significant + sig.rules.shape[0]\n",
    "\n",
    "print(\"all simplyfied for tissue\")\n",
    "print(count_of_all)    \n",
    "print(\"all simplyfied for tissue and significant\")\n",
    "print(count_of_all_significant) \n",
    "   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47f3d2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diff analysis between each cluster:\n",
    "for i, file in enumerate(files_main_mb):\n",
    "    print(file)\n",
    "    name_i = file.split('.pkl')[0]\n",
    "    \n",
    "    for j in range(i + 1, len(files_main_mb), 1):\n",
    "        file_j = files_main_mb[j]\n",
    "        name_j = file_j.split('.pkl')[0]\n",
    "        print(j)\n",
    "        print(name_j)\n",
    "        A = CombObj().from_pickle(f\"{main_analysis_path}{file}\")\n",
    "        print(A)\n",
    "        A.set_prefix(name_i)\n",
    "        B = CombObj().from_pickle(f\"{main_analysis_path}{file_j}\")\n",
    "        print(B)\n",
    "        B.set_prefix(name_j)\n",
    "        compare_obj = A.compare(B)\n",
    "        compare_obj.to_pickle(f'{differential_analysis_path}{name_i}__{name_j}.pkl')\n",
    "        \n",
    "        selected_std = compare_obj.select_rules()\n",
    "        \n",
    "        #TODO: Save autamatically generated thresholds\n",
    "        # utils.get_threshold(new.rules.iloc[:,4], 'both', percent=0.05)\n",
    "        # logfc threshold (-xxx , +xxx)\n",
    "        #  utils.get_threshold(new.rules.iloc[:,2:4].mean(axis=1), 'upper', percent=0.05)\n",
    "        # cosine threshold\n",
    "        selected_std.to_pickle(f'{differential_analysis_selection_path}{name_i}__{name_j}.pkl')\n",
    "        \n",
    "        break;\n",
    "        \n",
    "        \n",
    "print(\"Done differential analysis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "777b6701",
   "metadata": {},
   "source": [
    "## OLD implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f3afe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_diff_obj_dataframe(diff_obj: DiffCombObj) -> pd.DataFrame:\n",
    "    \n",
    "    # possible prefix names ['right-lobe-of-liver.10', 'right-lobe-of-liver.16']\n",
    "    df = diff_obj.rules\n",
    "    tissue_name_c1 , cluster_nr_c1  = diff_obj.prefixes[0].split('.')\n",
    "    tissue_name_c2 , cluster_nr_c2  = diff_obj.prefixes[1].split('.')\n",
    "    suff = \"\"\n",
    "    if tissue_name_c1 == tissue_name_c2:\n",
    "        suff += f\"_{tissue_name_c1}\"\n",
    "    else:\n",
    "        suff += f\"_{tissue_name_c1}_{tissue_name_c2}\"\n",
    "\n",
    "    if cluster_nr_c1 == cluster_nr_c2:\n",
    "        suff += f\"_{cluster_nr_c1}\"\n",
    "    else:\n",
    "        suff += f\"_{cluster_nr_c1}_{cluster_nr_c2}\"\n",
    "\n",
    "    df['log2fc_class'] = df.apply(lambda x: 'negativ' if x[4] < 0 else 'positiv', axis=1)\n",
    "    df.columns = [f'{x}{suff}' for x in df.columns]\n",
    "    \n",
    "    return df.copy(deep=True)\n",
    "\n",
    "\n",
    "# Find the specific tf_cooccurences of a tissue that unique for the specific cluster in the tissue.\n",
    "# 1. Diff analyse , \n",
    "# 2. Read in Diffanalyse for the specific cluster\n",
    "# 3. Find tf-cooccurence of the diffob , which are occuring in each cluster\n",
    "# Read in file Names of all analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "#print(f\"Files: {files_main_mb}\")\n",
    "\n",
    "files_diff= read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "print(f\"Count of Files: {len(files_diff)}\")\n",
    "\n",
    "test = \"\"\n",
    "for file_mb in files_main_mb:\n",
    "    cluster_name = file_mb.split('.pkl')[0]\n",
    "    print(cluster_name)\n",
    "    diffs = list(filter(lambda x: cluster_name in x, files_diff))\n",
    "    print(len(diffs))\n",
    "    print(diffs)\n",
    "    \n",
    "    # Keeps the read in DiffCombObj diff_objects:\n",
    "    diff_objects = []\n",
    "    \n",
    "    for diff in diffs:\n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}{diff}\")\n",
    "        diff_objects.append(diff_obj)\n",
    "    \n",
    "    erg = None\n",
    "    for i in range(len(diff_objects)-1):\n",
    "        \n",
    "        if erg is None:\n",
    "            obj_1= diff_objects[i]\n",
    "            obj_2 = diff_objects[i + 1]\n",
    "            df1 = prepare_diff_obj_dataframe(diff_obj = obj_1)\n",
    "            df2 = prepare_diff_obj_dataframe(diff_obj = obj_2)\n",
    "            \n",
    "            erg = df1.merge(df2, how='outer', left_index=True, right_index=True)\n",
    "        else:\n",
    "            obj_2 = diff_objects[i + 1] \n",
    "            df2 = prepare_diff_obj_dataframe(diff_obj = obj_2)\n",
    "            erg = erg.merge(df2, how='outer', left_index=True, right_index=True)\n",
    "       \n",
    "    test = erg\n",
    "    erg.to_pickle(path=f\"{answers_path}{cluster_name}.pkl\")\n",
    "    \n",
    "print(\"Done\")    \n",
    "test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bde5900",
   "metadata": {},
   "outputs": [],
   "source": [
    "answer_file_names=read_in_file_names_of_folder(rel_path=answers_path)\n",
    "print(answer_file_names)\n",
    "cluster_dfs = []\n",
    "df = None\n",
    "for name in answer_file_names:\n",
    "    df = pd.read_pickle(f\"{answers_path}{name}\")\n",
    "    cluster_dfs.append(name)\n",
    "    df = df\n",
    "\n",
    "    #df.groupby(['class', 'value']).count()\n",
    "    break;\n",
    "filter_columns = list(filter(lambda x: 'log2fc_class' in x , df.columns))\n",
    "#len(filter_columns)\n",
    "filtered_df = df[df[filter_columns].notna().all(1)] #\n",
    "filtered_df\n",
    "df\n",
    "#df3.iloc[:, 2:3]\n",
    "#df = pd.read_pickle(f\"{answers_path}right-lobe-of-liver.6.pkl\")\n",
    "#df = pd.read_pickle(f\"{differential_analysis_selection_path}{right-lobe-of-liver.6.pkl}\")\n",
    "\n",
    "\n",
    "#original = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.6.pkl\")\n",
    "#original.rules.loc[df.index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee683c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle(f\"{answers_path}right-lobe-of-liver.6.pkl\")\n",
    "selection = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.10__right-lobe-of-liver.16.pkl\")\n",
    "selection_orig = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.10__right-lobe-of-liver.16.pkl\")\n",
    "selection_orig\n",
    "selection\n",
    "\n",
    "original = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.6.pkl\")\n",
    "original.rules.loc[df.index]\n",
    "selection.prefixes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "245c7d57",
   "metadata": {},
   "source": [
    "### Try and Error section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c1e564",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#mb_obj = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.10.pkl\")\n",
    "obj_1 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.1.pkl\")\n",
    "obj_2 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.2.pkl\")\n",
    "obj_3 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.3.pkl\")\n",
    "obj_4 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.4.pkl\")\n",
    "obj_5 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.5.pkl\")\n",
    "obj_6 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.6.pkl\")\n",
    "#mb_obj.rules\n",
    "#type(diff_obj.rules)\n",
    "#diff_obj2.rules\n",
    "df_diff = pd.concat([obj_1.rules, obj_2.rules, obj_3.rules, obj_4.rules, obj_5.rules, obj_6.rules], join=\"inner\")\n",
    "\n",
    "\n",
    "\n",
    "#df_diff2 = pd.concat([diff_obj_1_1.rules,diff_obj2.rules, diff_obj1_2.rules])\n",
    "\n",
    "unified_duplicates = df_diff[df_diff.duplicated(subset=['TF1', 'TF2'], keep='first')]\n",
    "\n",
    "df_diff2 = pd.concat([unified_duplicates, diff_obj1_2.rules])\n",
    "\n",
    "unified_duplicates2 = df_diff2[df_diff2.duplicated(subset=['TF1', 'TF2'], keep='first')]\n",
    "\n",
    "#df_diff = df_diff.drop_duplicates(subset=['TF1', 'TF2'])\n",
    "#unified_duplicates\n",
    "#unified_duplicates\n",
    "#unified_duplicates2\n",
    "#diff_obj_1_1.rules\n",
    "#unified_duplicates\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3318406",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_1 = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.16__right-lobe-of-liver.1.pkl\")\n",
    "obj_2 = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.16__right-lobe-of-liver.2.pkl\")\n",
    "\n",
    "obj_1.simplify_rules()\n",
    "obj_2.simplify_rules()\n",
    "obj3 = obj_1.rules.merge(obj_2.rules, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "obj3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85338848",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['log2fc_class'] = df.apply(lambda x: 'negativ' if x[4] < 0 else 'positiv', axis=1)\n",
    "\n",
    "#removedNAN = df[df.notna().all(1)]\n",
    "\n",
    "#df2 = removedNAN[(removedNAN[filter_columns] > 0.0) | (removedNAN[filter_columns] < 0.0)]\n",
    "#df2[df2.notna().all(1)]\n",
    "#filtered_df = df[df[filter_columns].notna().all(1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d35c51b",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_obj_1_1.rules.loc['Foxd3-ONECUT2']\n",
    "diff_obj2.rules.loc['Foxd3-ONECUT2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "732674cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C = selectedC.select_top_rules(n=30)\n",
    "top30C.rules.head(31)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "bcd725aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value1</th>\n",
       "      <th>value2</th>\n",
       "      <th>value3</th>\n",
       "      <th>value4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>my1</th>\n",
       "      <td>-1.00</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my2</th>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my3</th>\n",
       "      <td>3.00</td>\n",
       "      <td>3.40</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my4</th>\n",
       "      <td>6.00</td>\n",
       "      <td>9.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my5</th>\n",
       "      <td>4.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>9.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my6</th>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     value1  value2  value3  value4\n",
       "my1   -1.00   -1.00    1.00    1.00\n",
       "my2   -3.33    3.33   -3.33    3.33\n",
       "my3    3.00    3.40    3.00    3.00\n",
       "my4    6.00    9.00    2.00    1.00\n",
       "my5    4.00    2.00    4.00    9.00\n",
       "my6    7.00    7.00    7.00    7.00"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame({\n",
    "                    'value1': [-1, -3.33, 3, 6,4,7],\n",
    "                    'value2': [-1, 3.33, 3.4, 9,2,7],\n",
    "                    'value3': [1, -3.33, 3, 2,4,7],\n",
    "                    'value4': [1, 3.33, 3, 1,9,7]},\n",
    "                   index=['my1', 'my2', 'my3', 'my4', 'my5', 'my6'])\n",
    "df2 = pd.DataFrame({'rkey': ['foo', 'bar', 'baz', 'foo','test'],\n",
    "                    'value': [5, 6, 7, 8, 9]},\n",
    "                  index=['my1', 'not2', 'my3', 'not4', 'my5'])\n",
    "df3 = pd.DataFrame({'rkey': ['new', 'lol'],\n",
    "                    'value': [5, 6]},\n",
    "                  index=['my1', 'not2'])\n",
    "df1\n",
    "\n",
    "#df1.merge(df2, left_on='lkey', right_on='rkey')\n",
    "#erg = df1.merge(df2, left_index=True, right_index=True, suffixes=(\"_test\", \"_test2\"))\n",
    "#erg = df1.merge(df2, how='outer', left_index=True, right_index=True, suffixes=(\"_x\", \"_y\"))\n",
    "#erg = erg.merge(df3, how='outer', left_index=True, right_index=True, suffixes=(\"_x\", \"_\"))\n",
    "#erg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "607bc563",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value1</th>\n",
       "      <th>value2</th>\n",
       "      <th>value3</th>\n",
       "      <th>value4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>my1</th>\n",
       "      <td>-1.00</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my2</th>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my6</th>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     value1  value2  value3  value4\n",
       "my1   -1.00   -1.00    1.00    1.00\n",
       "my2   -3.33    3.33   -3.33    3.33\n",
       "my6    7.00    7.00    7.00    7.00"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1[df1.apply(lambda x: min(abs(x)) == max(abs(x)), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "e06cf753",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_abs = df1.abs()\n",
    "df_abs\n",
    "\n",
    "df_abs['num_uniq'] = df_abs.apply(pd.Series.nunique, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "255f88c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value1</th>\n",
       "      <th>value2</th>\n",
       "      <th>value3</th>\n",
       "      <th>value4</th>\n",
       "      <th>num_uniq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>my1</th>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my2</th>\n",
       "      <td>3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my3</th>\n",
       "      <td>3.00</td>\n",
       "      <td>3.40</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my4</th>\n",
       "      <td>6.00</td>\n",
       "      <td>9.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my5</th>\n",
       "      <td>4.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>9.00</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my6</th>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     value1  value2  value3  value4  num_uniq\n",
       "my1    1.00    1.00    1.00    1.00         1\n",
       "my2    3.33    3.33    3.33    3.33         1\n",
       "my3    3.00    3.40    3.00    3.00         2\n",
       "my4    6.00    9.00    2.00    1.00         4\n",
       "my5    4.00    2.00    4.00    9.00         3\n",
       "my6    7.00    7.00    7.00    7.00         1"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_abs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "f158220c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rows_with_same_value(df:pd.DataFrame) -> pd.DataFrame:\n",
    "    '''\n",
    "    Removes all rows of a Dataframe that do not have the same value in each column and\n",
    "    returns the reduced dataframe.\n",
    "    '''\n",
    "    return df[df.apply(lambda x: find_value(x) , 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "69fae29e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_if_row_has_min_number_of_same_values(row, min_same_val=4):\n",
    "    l = len(row.index)\n",
    "    #print(l)\n",
    "    count_same_val = 0\n",
    "    i = 0\n",
    "    for i in range(0, l - 1):\n",
    "        j = i + 1\n",
    "        while j < l:\n",
    "            if abs(row[i]) == abs(row[j]):\n",
    "                count_same_val = count_same_val + 1\n",
    "            j = j+1\n",
    "            \n",
    "    #print(f'Count of same values in a row: {count_same_val}')   \n",
    "    return (count_same_val >= min_same_val )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "a2195c08",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_if_row_has_min_number_of_same_values(row, min_same_val:int):\n",
    "    '''\n",
    "    Check if a row of a dataframe has a minumum count of same values in each column.\n",
    "    Returns true if a minmum count of same values is reached.\n",
    "    ---\n",
    "    Parameters:\n",
    "        row: pd.dataframe row,\n",
    "            Contains the row of a dataframe, with the rows values.\n",
    "        min_same_val: int,\n",
    "            Threshold for minimum number of same values in the row.\n",
    "    ---\n",
    "    Return True or False\n",
    "    '''\n",
    "    l = len(row.index)\n",
    "    abs_row = row.abs()\n",
    "    count_of_distinct_value = abs_row.nunique()\n",
    "    count_same_val = (l + 1) - count_of_distinct_value\n",
    "    print(count_same_val)\n",
    "    \n",
    "    return (count_same_val >= min_same_val)\n",
    "\n",
    "def find_rows_with_min_same_values(df:pd.DataFrame, min_same_val:int) -> pd.DataFrame:\n",
    "    '''\n",
    "    Remove all rows of a Dataframe that do not have a minimum count of same value in each column and\n",
    "    return the reduced dataframe.\n",
    "    '''\n",
    "    \n",
    "    return df[df.apply(lambda x: check_if_row_has_min_number_of_same_values(x, min_same_val=min_same_val), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "3a3a9784",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "4\n",
      "3\n",
      "1\n",
      "2\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value1</th>\n",
       "      <th>value2</th>\n",
       "      <th>value3</th>\n",
       "      <th>value4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>my1</th>\n",
       "      <td>-1.00</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my2</th>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "      <td>-3.33</td>\n",
       "      <td>3.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>my6</th>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>7.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     value1  value2  value3  value4\n",
       "my1   -1.00   -1.00    1.00    1.00\n",
       "my2   -3.33    3.33   -3.33    3.33\n",
       "my6    7.00    7.00    7.00    7.00"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_rows_with_min_same_values(df1, min_same_val=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c4db0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.columns = [f'{x}_df2' for x in df2.columns]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8024c034",
   "metadata": {},
   "outputs": [],
   "source": [
    "[x+ 1 for x in df2.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "615d5305",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1['class'] = df1.apply(lambda x: 'niedrig' if x[1] < 5 else 'hoch', axis=1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d1c3d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1[df1['class'] == 'hoch']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa2c8394",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.groupby(['class', 'value']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa0f86e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4cd1175",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3.iloc[:, 2:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91b30ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C.plot_bubble()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa096c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C.plot_network()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2788d143",
   "metadata": {},
   "outputs": [],
   "source": [
    "selectedC.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1edc217",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34cfa1b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### OLD \n",
    "# Find the specific tf_cooccurences of a tissue that unique for the specific cluster in the tissue.\n",
    "# 1. Diff analyse , \n",
    "# 2. Read in Diffanalyse for the specific cluster\n",
    "# 3. Find tf-cooccurence of the diffob , which are occuring in each cluster\n",
    "# Read in file Names of all analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "#print(f\"Files: {files_main_mb}\")\n",
    "\n",
    "files_diff= read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "print(f\"Count of Files: {len(files_diff)}\")\n",
    "#print(f\"Files: {files_diff}\")\n",
    "test = \"\"\n",
    "for file_mb in files_main_mb:\n",
    "    cluster_name = file_mb.split('.pkl')[0]\n",
    "    print(cluster_name)\n",
    "    diffs = list(filter(lambda x: cluster_name in x, files_diff))\n",
    "    print(len(diffs))\n",
    "    print(diffs)\n",
    "    \n",
    "    # Keeps the read in DiffCombObj diff_objects:\n",
    "    diff_objects = []\n",
    "    \n",
    "    for diff in diffs:\n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}{diff}\")\n",
    "        diff_objects.append(diff_obj)\n",
    "    \n",
    "    # erste DiffObj dataframe \n",
    "    initial_df = diff_objects[0].rules\n",
    "    \n",
    "    #has neg and pos foldchange\n",
    "    cross_product_merged = initial_df\n",
    "    \n",
    "    # only pos foldchange\n",
    "    pos_merged = initial_df[initial_df.iloc[:,4] > 0.00]\n",
    "    \n",
    "    # only neg foldchange\n",
    "    neg_merged = initial_df[initial_df.iloc[:,4] < 0.00]\n",
    "    for i in range(len(diff_objects)-1):\n",
    "        obj_1= diff_objects[i]\n",
    "        obj_2 = diff_objects[i + 1]\n",
    "            \n",
    "        # cross_product merge rules-dataframe by index (TF´s)\n",
    "        cross_product = cross_product_merged.merge(obj_2.rules, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        cross_product_merged = cross_product.copy(deep=True)\n",
    "        \n",
    "        # pos merge rules-dataframe by index (TF´s)\n",
    "        obj2_df_pos = obj_2.rules[obj_2.rules.iloc[:,4] > 0.00]\n",
    "        df_pos_merged = pos_merged.merge(obj2_df_pos, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        pos_merged = df_pos_merged.copy(deep=True)\n",
    "        \n",
    "        # neg merge rules-dataframe by index (TF´s)\n",
    "        obj2_df_neg = obj_2.rules[obj_2.rules.iloc[:,4] < 0.00]\n",
    "        df_neg_merged = neg_merged.merge(obj2_df_neg, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        neg_merged = df_neg_merged.copy(deep=True)\n",
    "        \n",
    "        \n",
    "    cross_product_merged.to_pickle(path=f\"{answers_path}{cluster_name}_cross.pkl\")\n",
    "    pos_merged.to_pickle(path=f\"{answers_path}{cluster_name}_pos.pkl\")\n",
    "    neg_merged.to_pickle(path=f\"{answers_path}{cluster_name}_neg.pkl\")\n",
    "    \n",
    "print(\"Done\")    \n",
    "test\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tfcomb_env",
   "language": "python",
   "name": "tfcomb_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
