{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fa877c14",
   "metadata": {},
   "source": [
    "# TF-co-occurences for WP2 - Data\n",
    "### Outline of this notebook:\n",
    "    1. Constants, Path and Interface Definitions \n",
    "    2. Market Basket analysis with tf-comb, for all cluster/celltypes of a tissue\n",
    "    3. Differential analysis with all market basket analysis (CombObj ´s) of step 2 for the clusters/celltypes of a single tissue. (One DiffCombObj for each tissue)\n",
    "    4. Analysis for biological questions \n",
    "    \n",
    "The aim is to find transcription-factor-co-occurences for cluster/celltypes of human tissues with the help of the python-library tf-comb. For the data of WP2. The data basis comes from the \"cell atlas of chromatin accessibility across 25 adult human tissues\"(https://doi.org/10.1101/2021.02.17.431699) \n",
    "\n",
    "**Biological question, that we want to answer with this notebook:**\n",
    "\n",
    "1. Find Transcriptionfactor-co-occurences, which only occure in one (or more) \"cluster/celltypes of a tissue\".\n",
    "    Maybe we can identified a cluster through this co-occurences.\n",
    "\n",
    "** How to use this notbook: **\n",
    "1. Please adapted the paths in Constants, Path and interface defintions for your approach\n",
    "2. Please make sure you have installed the kernel as it is described in the ReadMe\n",
    "3. Check if the WP2 Data structure is correctly provided (ReadMe).\n",
    "    - Has all tissues as a folder in it: ../OUPTPUT_FOLDER/\n",
    "    - You find the data (open cromatin regions per cluster .bed-files) e.g ../OUPTPUT_FOLDER/<tissue>/WP6/*cluster_x.bed\n",
    "4. Execute each notebook window from top to bottom one after another."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa1aab90",
   "metadata": {},
   "source": [
    "## 1. Constants, Path and Interface Definitions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "487586df",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tfcomb import CombObj, DiffCombObj, utils\n",
    "import os\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "'''\n",
    "Constants for this script.\n",
    "\n",
    "This window contains all paths and constants, which are later used for this juypter notbook.\n",
    "\n",
    "Please adapt paths or constants, if use other files. \n",
    "For example adapted the genome path, if you use another genome. \n",
    "'''\n",
    "\n",
    "# Path to genome fasta file. Is used for the market basket analysis of tfcomb.\n",
    "genome_path=\"/mnt/workspace_stud/allstud/homo_sapiens.104.mainChr.fa\"\n",
    "\n",
    "# Path to the jaspar file (contains transcription factors (TF) binding profiles\n",
    "# as position frequency matrices (PFMs)). Is used for the market basket analysis of tfcomb\n",
    "main_jaspar_file=\"../testdaten/JASPAR2020_CORE_vertebrates.meme\" \n",
    "\n",
    "# Path where results of this notebook will be written to (eg. TF_COMB objects, .pkl).\n",
    "result_path=\"./results/wp2/\"\n",
    "\n",
    "# Paths in the result folder:\n",
    "# Path to folder, where the resulting market basket analysis for a cluster/celltype is put \n",
    "main_analysis_path=f\"{result_path}main/\"\n",
    "\n",
    "# Path to folder, where the differential analysis for a tissue is put \n",
    "differential_analysis_path=f\"{result_path}diff_analysis/\"\n",
    "# Path to folder, where answers of our results are put to\n",
    "answers_path=f\"{result_path}answers/\"\n",
    "\n",
    "# Path to folder (Interface folder) of wp2, where the clusters of each tissue can be found.\n",
    "path_to_tissues=\"/mnt/workspace_stud/stud3/WP2_OUTPUT/FINISHED/\"\n",
    "\n",
    "# Tag for WP6 data in WP2 Interface\n",
    "cluster_folder_tag=\"WP6/\"\n",
    "\n",
    "# Path to Folder with celltype annotation tables of wp2\n",
    "celltype_annotation_path = \"/mnt/workspace_stud/stud4/celltype_assignment_tables/\"\n",
    "\n",
    "# The following lines, initally check if all file/paths are available. \n",
    "#If a result folder does not exist it is created automatically\n",
    "if not os.path.exists(result_path):\n",
    "     pathlib.Path(result_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(main_analysis_path):\n",
    "     pathlib.Path(main_analysis_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(differential_analysis_path):\n",
    "     pathlib.Path(differential_analysis_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(answers_path):\n",
    "     pathlib.Path(answers_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(genome_path):\n",
    "    print(f\"ERROR: path {genome_path} does not exist\")\n",
    "\n",
    "if not os.path.exists(main_jaspar_file):\n",
    "    print(f\"ERROR: path {main_jaspar_file} does not exist\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11adea88",
   "metadata": {},
   "source": [
    "### Helper functions for reading-in folders/files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "92e608ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_folder_names_in_folder(rel_folder_path:str):\n",
    "    ''' \n",
    "        Read in the names of the folders in a folder.(rel_folder_path)\n",
    "        ---\n",
    "        Parameters:\n",
    "        \n",
    "        rel_folder_path: String\n",
    "            relative Path to the folder location that is read in.\n",
    "        ---\n",
    "        Return: a List of Strings (folder names)\n",
    "    '''\n",
    "\n",
    "    dirlist = [ item for item in os.listdir(rel_folder_path) if os.path.isdir(os.path.join(rel_folder_path, item))]\n",
    "    folder_names = []\n",
    "    for folder in dirlist:\n",
    "        folder_names.append(folder)\n",
    "    return folder_names\n",
    "\n",
    "def read_in_file_names_of_folder(rel_path:str):\n",
    "    ''' \n",
    "        Read in the file names in a folder (rel_path).\n",
    "        ---\n",
    "        Parameters:\n",
    "        \n",
    "        rel_path: String\n",
    "            relative Path to the folder location.\n",
    "        ---\n",
    "        Return: a List of Strings (file names)\n",
    "    '''\n",
    "    return [f for f in os.listdir(rel_path) if os.path.isfile(os.path.join(rel_path, f))]\n",
    "\n",
    "\n",
    "#cluster_file_names = read_in_file_names_of_folder(rel_path=path_to_clusters)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0dba2e",
   "metadata": {},
   "source": [
    "## 2. Market basket analysis with tf-comb\n",
    "We do a market basket analysis with tfcomb for each cluster/celltype, which has been culstered by WP2 and comes from the raw single-cell ATAC-Data of the cell atlas project. As a result, we get the transcription-factor-co-occurences for each cluster. The Trancriptionsfactor motif´s come as a postion-frequency-matrix from https://jaspar.genereg.net/search?q=&collection=CORE&tax_group=vertebrates. The corresponding genome, which is used is **homo_sapiens.104.mainChr.fa** .\n",
    "\n",
    "Approach:\n",
    "1. Read-in tissue folders of WP2\n",
    "2. For each tissue: read-in single .bed (Content= open-chromatin regions) files for each cluster/celltype\n",
    "3. Do market basket analysis for each cluster/celltype\n",
    "4. Result .pkl CombObj files can be found under **/{result_path}/{main_analysis_path}/{tissue_name}/{cluster_name}.pkl **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "2f22f857",
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_market_basket_analyses_for_cell_cluster(mb_file_name: str, cell_cluster_path:str, tissue:str):\n",
    "    '''\n",
    "        Does market basket analysis with tfcomb. Saves the tfcomb-Object as .pkl file to main_analysis_path.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        mb_file_name: string\n",
    "            Name for the result file of the market basket analyses. \n",
    "            e.g \"<tissue>_<cluster_number>_<celltype>\".\n",
    "        \n",
    "        cell_cluster_path: string\n",
    "            Path to the .bed-File with genome regions to check for tf-co-occurences\n",
    "         \n",
    "        tissue: string\n",
    "            Tissue name from where the cluster corresponds to.\n",
    "    '''    \n",
    "    # Save path initalization, if folder for tissue does not exists, new folder is created.\n",
    "    save_path = f'{main_analysis_path}{tissue}/'\n",
    "    if not os.path.exists(save_path):\n",
    "         pathlib.Path(save_path).mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # TF-comb market basket analysis\n",
    "    comb = CombObj()\n",
    "    comb.TFBS_from_motifs(regions= cell_cluster_path,\n",
    "                   motifs=main_jaspar_file,\n",
    "                   genome=genome_path,\n",
    "                   threads=4)\n",
    "    \n",
    "    print(f'Start market basket analyses for cell-cluster/type: {mb_file_name}')\n",
    "    comb.market_basket(threads=10)\n",
    "    \n",
    "    # if rules are empty nothing is saved \n",
    "    if len(comb.rules) <= 0:\n",
    "        print(f'Could not find TF-cooccurences for cell-cluster/type: {mb_file_name}')\n",
    "        return\n",
    "    print(f'Finished market basket analyses for cell-cluster/type: {mb_file_name}')\n",
    "    print(f'Found rules: {len(comb.rules)}')\n",
    "    \n",
    "    # save tf-comb obj to .pkl\n",
    "    comb.to_pickle(f'{save_path}{mb_file_name}.pkl')\n",
    "    print(f'Saved: {save_path}{mb_file_name}.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "c0dceb84",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C1PX3_thoracic_aorta', 'IQYCP_lower_leg_skin', 'CSSD4_omental_fat_pad', 'JF1O9_gastrocnemius_medialis', 'JF1O6_body_of_pancreas', 'A8CPH_esophagus_muscularis_mucosa', 'ACCPU_upper_lobe_of_left_lung', 'JF1O8_colon_sigmoid', 'JF1NP_stomach']\n"
     ]
    }
   ],
   "source": [
    "# Load tissue folder names for wp2 data\n",
    "tissues=get_folder_names_in_folder(rel_folder_path=path_to_tissues)\n",
    "print(tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f2f4a27",
   "metadata": {},
   "source": [
    "### Functions for adding the celltype annotations of wp2 to the corresponding clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "92e9bb1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSSD4_omental_fat_pad\n",
      "['celltype_assignment_table_CSSD4_omental_fat_pad', 'celltype_assignment_table_JF1O9_gastrocnemius_medialis', 'celltype_assignment_table_JF1O6_body_of_pancreas', 'celltype_assignment_table_IQYCP_lower_leg_skin']\n"
     ]
    }
   ],
   "source": [
    "def get_celltype_annotation_table(celltype_annotation_path:str) -> pd.DataFrame:\n",
    "    '''\n",
    "    Read in the the given annotation file(celltype_annotation_path) with pandas read_table function\n",
    "    ---\n",
    "    Parameters:\n",
    "    celltype_annotation_path : string\n",
    "        Is the path to a file which has the celltype annotations for the cluster numbers.\n",
    "        e.g. file content:\n",
    "        6\tStromal cells\n",
    "        3\tAdipocyte progenitor cells\n",
    "        4\tStromal cells\n",
    "        7\tFibroblasts\n",
    "    ---\n",
    "    Returns a celltype annotation table as a pandas Dataframe.\n",
    "    '''\n",
    "    df = pd.read_table(celltype_annotation_path, header=None, index_col=None)\n",
    "    return df\n",
    "\n",
    "def get_celltype_for_cluster(cluster_number: int, celltype_annotation_table: pd.DataFrame):\n",
    "    '''\n",
    "    Find celltype annotation for cluster number in the corresponding celltype_annotation_table\n",
    "    ---\n",
    "    Parameters:\n",
    "    cluster_number: integer\n",
    "        number of the cluster\n",
    "    \n",
    "    celltype_annotation_table: pd.DataFrame\n",
    "        table with the celltype annotations, e.g. Dataframe\n",
    "         0\t1 \n",
    "       1 6\tStromal cells\n",
    "       2 3\tAdipocyte progenitor cells\n",
    "       3 4\tStromal cells\n",
    "       4 7\tFibroblasts\n",
    "    ---\n",
    "    Returns the celltype for a cluster_number. (if no celltype kann be found \"na\" is returned) \n",
    "    '''\n",
    "    # initalize celltype name with na (not available), if no celltype can be found na is returned\n",
    "    celltype_name = \"na\"\n",
    "    try:\n",
    "        # Try to find the cell-type annotation in the table, if no annotation can be found key error is raised\n",
    "        celltype_name = df[df[0] == int(cluster_number)].at[0,1]\n",
    "        # replace whitespaces in celltype names, e.g. \"Stromal cell\" -> \"Stromal_cell\"\n",
    "        celltype_name = celltype_name.replace(\" \", \"_\")\n",
    "    except KeyError:\n",
    "        print(f\"Could not find celltype annotation for cluster: {cluster_number}\")\n",
    "   \n",
    "    return celltype_name\n",
    "\n",
    "##remove\n",
    "celltype_annotations=read_in_file_names_of_folder(rel_path=celltype_annotation_path)\n",
    "print(celltype_annotations[0].split('_table_')[1])\n",
    "df = pd.read_table(f\"{celltype_annotation_path}{celltype_annotations[0]}\", header=None, index_col=None)\n",
    "print(celltype_annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "d1ac05c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stromal cells\n",
      "Stromal_cells\n"
     ]
    }
   ],
   "source": [
    "### remove \n",
    "try:\n",
    "    name = df[df[0] == 6].at[0,1]\n",
    "    name1 = name.replace(\" \", \"_\")\n",
    "    print(name)\n",
    "    print(name1)\n",
    "except KeyError:\n",
    "    print(f\"Error\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "392e4bd1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def make_mb_for_clusters(path_to_clusters:str, tissue:str):\n",
    "    '''\n",
    "        Wrapper function, that does the market basket analysis for all clusters/celltypes in a tissue.\n",
    "        Also annotates the cluster with a celltype.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        path_to_clusters: string\n",
    "            Path to the .bed files(cluster) of a tissue.\n",
    "         \n",
    "        tissue: string\n",
    "            Tissue name from where the cluster corresponds to.\n",
    "        --- \n",
    "        Catch Exceptions with a message, if any error occures in the market basket analyses a message is printed\n",
    "        and the programm continues with the next tissue.\n",
    "    '''\n",
    "    # Read in the .bed files for each cluster of the specific tissue\n",
    "    cluster_file_names = read_in_file_names_of_folder(rel_path=path_to_clusters)\n",
    "    print(cluster_file_names)\n",
    "    \n",
    "    # Get the celltype annotation table. To add celltype for a cluster. e.g. cluster6 = fibroblast\n",
    "    celltype_table = pd.DataFrame()\n",
    "    # TODO:\n",
    "    #celltype_table = get_celltype_annotation_table(\n",
    "     #   celltype_annotation_path=f'{path_to_tissues}annotaion/annotation.txt') # pandas dataframe is returned\n",
    "    \n",
    "    # Do a market basket analysis for each cluster of a tissue\n",
    "    for file_name in cluster_file_names:\n",
    "        # e.g JF1O6_body_of_pancreas.10_peaks.bed -> [JF1O6_body_of_pancreas.10_peaks] = clustername = JF1O6_body_of_pancreas.10_peaks\n",
    "        cluster_name = file_name.split('.bed')[0]\n",
    "        #e.g JF1O6_body_of_pancreas.10_peaks -> [JF1O6_body_of_pancreas, 10_peaks], cluster_number_with_tag = 10_peaks\n",
    "        cluster_number_with_tag = cluster_name.split('.')[1]\n",
    "        # e.g 10_peaks -> [10 , peaks], cluster_number = 10\n",
    "        cluster_number = int(cluster_number_with_tag.split('_')[0])\n",
    "        \n",
    "        celltype_name = get_celltype_for_cluster(cluster_number=cluster_number,\n",
    "                                                 celltype_annotation_table=celltype_table)\n",
    "        \n",
    "        try:\n",
    "            print(cluster_name)\n",
    "            print(cluster_number)\n",
    "            print(celltype_name)\n",
    "            print(file_name)\n",
    "            # Prepare names and paths\n",
    "            cluster_path=f\"{path_to_clusters}{file_name}\"\n",
    "            # e.g. JF1O6_body_of_pancreas_c3_fibroblast\n",
    "            mb_file_name = f\"{tissue}_c{str(cluster_number)}_{celltype_name}\"\n",
    "            \n",
    "            do_market_basket_analyses_for_cell_cluster(mb_file_name=mb_file_name, cell_cluster_path=cluster_path, tissue=tissue)\n",
    "        except Exception:\n",
    "            print(f\"ERROR: Market basket for cluster:{cluster_name} in tissue {tissue}, did not work\")\n",
    "            continue\n",
    "            \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "e14e7350",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C1PX3_thoracic_aorta.1_peaks.bed', 'C1PX3_thoracic_aorta.6_peaks.bed', 'C1PX3_thoracic_aorta.7_peaks.bed', 'C1PX3_thoracic_aorta.2_peaks.bed', 'C1PX3_thoracic_aorta.12_peaks.bed', 'C1PX3_thoracic_aorta.4_peaks.bed', 'C1PX3_thoracic_aorta.3_peaks.bed', 'C1PX3_thoracic_aorta.5_peaks.bed', 'C1PX3_thoracic_aorta.8_peaks.bed', 'C1PX3_thoracic_aorta.10_peaks.bed', 'C1PX3_thoracic_aorta.9_peaks.bed', 'C1PX3_thoracic_aorta.11_peaks.bed']\n",
      "Could not find celltype annotation for cluster: 1\n",
      "C1PX3_thoracic_aorta.1_peaks\n",
      "1\n",
      "na\n",
      "C1PX3_thoracic_aorta.1_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 127872 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c1_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 26%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 56%\n",
      "INFO: Progress: 66%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 96%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c1_na\n",
      "Found rules: 177934\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c1_na.pkl\n",
      "C1PX3_thoracic_aorta.6_peaks\n",
      "6\n",
      "Stromal cells\n",
      "C1PX3_thoracic_aorta.6_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 41%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 256421 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c6_Stromal cells\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 22%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 44%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c6_Stromal cells\n",
      "Found rules: 271212\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c6_Stromal cells.pkl\n",
      "Could not find celltype annotation for cluster: 7\n",
      "C1PX3_thoracic_aorta.7_peaks\n",
      "7\n",
      "na\n",
      "C1PX3_thoracic_aorta.7_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 31%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 102970 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c7_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 16%\n",
      "INFO: Progress: 26%\n",
      "INFO: Progress: 36%\n",
      "INFO: Progress: 42%\n",
      "INFO: Progress: 54%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 88%\n",
      "INFO: Progress: 98%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c7_na\n",
      "Found rules: 140720\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c7_na.pkl\n",
      "Could not find celltype annotation for cluster: 2\n",
      "C1PX3_thoracic_aorta.2_peaks\n",
      "2\n",
      "na\n",
      "C1PX3_thoracic_aorta.2_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 62%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 285967 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c2_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c2_na\n",
      "Found rules: 288276\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c2_na.pkl\n",
      "Could not find celltype annotation for cluster: 12\n",
      "C1PX3_thoracic_aorta.12_peaks\n",
      "12\n",
      "na\n",
      "C1PX3_thoracic_aorta.12_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 41%\n",
      "INFO: Progress: 51%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 378133 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c12_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 64%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c12_na\n",
      "Found rules: 334961\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c12_na.pkl\n",
      "Could not find celltype annotation for cluster: 4\n",
      "C1PX3_thoracic_aorta.4_peaks\n",
      "4\n",
      "na\n",
      "C1PX3_thoracic_aorta.4_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 871122 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c4_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c4_na\n",
      "Found rules: 457051\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c4_na.pkl\n",
      "Could not find celltype annotation for cluster: 3\n",
      "C1PX3_thoracic_aorta.3_peaks\n",
      "3\n",
      "na\n",
      "C1PX3_thoracic_aorta.3_peaks.bed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 61%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 801952 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c3_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c3_na\n",
      "Found rules: 441830\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c3_na.pkl\n",
      "Could not find celltype annotation for cluster: 5\n",
      "C1PX3_thoracic_aorta.5_peaks\n",
      "5\n",
      "na\n",
      "C1PX3_thoracic_aorta.5_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 41%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 1043665 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c5_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c5_na\n",
      "Found rules: 471365\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c5_na.pkl\n",
      "Could not find celltype annotation for cluster: 8\n",
      "C1PX3_thoracic_aorta.8_peaks\n",
      "8\n",
      "na\n",
      "C1PX3_thoracic_aorta.8_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 388825 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c8_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 56%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 94%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c8_na\n",
      "Found rules: 325970\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c8_na.pkl\n",
      "Could not find celltype annotation for cluster: 10\n",
      "C1PX3_thoracic_aorta.10_peaks\n",
      "10\n",
      "na\n",
      "C1PX3_thoracic_aorta.10_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 383890 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c10_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 54%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c10_na\n",
      "Found rules: 343822\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c10_na.pkl\n",
      "Could not find celltype annotation for cluster: 9\n",
      "C1PX3_thoracic_aorta.9_peaks\n",
      "9\n",
      "na\n",
      "C1PX3_thoracic_aorta.9_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 71%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 91%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 407328 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c9_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 10%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 72%\n",
      "INFO: Progress: 82%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c9_na\n",
      "Found rules: 346711\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c9_na.pkl\n",
      "Could not find celltype annotation for cluster: 11\n",
      "C1PX3_thoracic_aorta.11_peaks\n",
      "11\n",
      "na\n",
      "C1PX3_thoracic_aorta.11_peaks.bed\n",
      "INFO: Scanning for TFBS with 4 thread(s)...\n",
      "INFO: Progress: 11%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 30%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 50%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 81%\n",
      "INFO: Progress: 90%\n",
      "INFO: Finished!\n",
      "INFO: Processing scanned TFBS\n",
      "INFO: Identified 430332 TFBS (746 unique names) within given regions\n",
      "Start market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c11_na\n",
      "WARNING: No counts found in <CombObj>. Running <CombObj>.count_within() with standard parameters.\n",
      "INFO: Setting up binding sites for counting\n",
      "INFO: Counting co-occurrences within sites\n",
      "INFO: Counting co-occurrence within background\n",
      "INFO: Progress: 12%\n",
      "INFO: Progress: 20%\n",
      "INFO: Progress: 32%\n",
      "INFO: Progress: 40%\n",
      "INFO: Progress: 52%\n",
      "INFO: Progress: 60%\n",
      "INFO: Progress: 70%\n",
      "INFO: Progress: 80%\n",
      "INFO: Progress: 92%\n",
      "INFO: Finished!\n",
      "INFO: Done finding co-occurrences! Run .market_basket() to estimate significant pairs\n",
      "INFO: Market basket analysis is done! Results are found in <CombObj>.rules\n",
      "Finished market basket analyses for cell-cluster/type: C1PX3_thoracic_aorta_c11_na\n",
      "Found rules: 343079\n",
      "Saved: ./results/wp2/main/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta_c11_na.pkl\n",
      "DONE: Created market basket analysis for each cluster\n"
     ]
    }
   ],
   "source": [
    "# Create a market basket analysis for each cluster of tissue\n",
    "for tissue in tissues:\n",
    "    path_to_clusters = f\"{path_to_tissues}{tissue}/{cluster_folder_tag}\"\n",
    "    make_mb_for_clusters(path_to_clusters=path_to_clusters, tissue=tissue)\n",
    "    break\n",
    "\n",
    "print(\"DONE: Created market basket analysis for each cluster\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce9872a4",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c16c5d",
   "metadata": {},
   "source": [
    "### Differential Analysis\n",
    "We use the differential analysis of tfcomb to identifiy the differences of tf-co-occurences between all cluster/celltypes of a tissue. For this we load all market basket analysis (tfcomb-objects) of a tissue (see point 2) into a DiffCombObj. After the differential analysis, we filter the object, so that we hopefully find tf-co-occurences that only occure in a single cluster of that tissue or only in a special celltype.  \n",
    "\n",
    "- mb = market basket analysis (CombObj of tf comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2f2b5cf7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['JF1O6_body_of_pancreas', 'C1PX3_thoracic_aorta', 'IQYCP_lower_leg_skin', 'CSSD4_omental_fat_pad', 'JF1O9_gastrocnemius_medialis']\n"
     ]
    }
   ],
   "source": [
    "# get the tissue names by folder names of the market basket analysis \n",
    "mb_tissues = get_folder_names_in_folder(rel_folder_path=main_analysis_path)\n",
    "print(mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "929be533",
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_differential_analysis_for_tissues(tissues=[]):\n",
    "    '''\n",
    "        Differential analysis between all clusters/celltypes of a tissue.\n",
    "        ---\n",
    "        Paramater:\n",
    "        \n",
    "        tissues: array\n",
    "            tissue names by market basket analysis.\n",
    "        ---\n",
    "        DiffCombObj are saved as .pkl to differential_analysis_path\n",
    "    '''\n",
    "    for tissue_folder in tissues:\n",
    "        \n",
    "        diff_save_path = f\"{differential_analysis_path}{tissue_folder}/\"\n",
    "        # Check if  folder for differential_analysis already exists for tissue, if not create new one\n",
    "        if not os.path.exists(diff_save_path):\n",
    "             pathlib.Path(diff_save_path).mkdir(parents=True, exist_ok=True)\n",
    "        \n",
    "        # get file names of the market basket analysis\n",
    "        tissue_mb_files = read_in_file_names_of_folder(rel_path=f\"{main_analysis_path}{tissue_folder}/\")\n",
    "        \n",
    "        # holds the combobj´s\n",
    "        tissue_mbs_to_compare = []\n",
    "        for file in tissue_mb_files:\n",
    "            print(file)\n",
    "            file_name = file.split('.pkl')[0]\n",
    "            # Load the CombObj (market basket analysis) for each cluster of a tissue\n",
    "            obj = CombObj().from_pickle(f\"{main_analysis_path}{tissue_folder}/{file}\")\n",
    "            obj.set_prefix(file_name)\n",
    "            tissue_mbs_to_compare.append(obj)\n",
    "        \n",
    "        # Create DiffCombObj with all Combobj (market basket analysis) of the clusters in a tissue\n",
    "        compare_obj = DiffCombObj(tissue_mbs_to_compare, measure=\"cosine\", join=\"outer\", fillna=True)\n",
    "        # save diffcombj\n",
    "        compare_obj.to_pickle(f'{diff_save_path}{tissue_folder}.pkl')\n",
    "        # Normalize the DiffCombObj\n",
    "        compare_obj.normalize()\n",
    "        compare_obj.calculate_foldchanges()\n",
    "        \n",
    "        # Remove rules which are doubled, e.g. A-B, B-A; (B-A) is removed\n",
    "        compare_obj.simplify_rules()\n",
    "        # Save the normalized, foldchange calculated and simplified diff_comb_obj to .pkl\n",
    "        compare_obj.to_pickle(f'{diff_save_path}{tissue_folder}_normalized.pkl')\n",
    "        print(f\"Done: Diff analysis for tissue {tissue_folder}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "42f520dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JF1O6_body_of_pancreas.6_peaks.pkl\n",
      "JF1O6_body_of_pancreas.1_peaks.pkl\n",
      "JF1O6_body_of_pancreas.2_peaks.pkl\n",
      "JF1O6_body_of_pancreas.5_peaks.pkl\n",
      "JF1O6_body_of_pancreas.3_peaks.pkl\n",
      "JF1O6_body_of_pancreas.4_peaks.pkl\n",
      "JF1O6_body_of_pancreas.10_peaks.pkl\n",
      "JF1O6_body_of_pancreas.8_peaks.pkl\n",
      "JF1O6_body_of_pancreas.11_peaks.pkl\n",
      "JF1O6_body_of_pancreas.7_peaks.pkl\n",
      "JF1O6_body_of_pancreas.9_peaks.pkl\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.5_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.6_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.5_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.1_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.5_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.2_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.5_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.3_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.4_peaks / JF1O6_body_of_pancreas.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.4_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.4_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.4_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.4_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.10_peaks / JF1O6_body_of_pancreas.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.10_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.10_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.10_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.8_peaks / JF1O6_body_of_pancreas.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.8_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.8_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.11_peaks / JF1O6_body_of_pancreas.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.11_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O6_body_of_pancreas.7_peaks / JF1O6_body_of_pancreas.9_peaks\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue JF1O6_body_of_pancreas\n",
      "C1PX3_thoracic_aorta.1_peaks.pkl\n",
      "C1PX3_thoracic_aorta.6_peaks.pkl\n",
      "C1PX3_thoracic_aorta.7_peaks.pkl\n",
      "C1PX3_thoracic_aorta.2_peaks.pkl\n",
      "C1PX3_thoracic_aorta.12_peaks.pkl\n",
      "C1PX3_thoracic_aorta.4_peaks.pkl\n",
      "C1PX3_thoracic_aorta.3_peaks.pkl\n",
      "C1PX3_thoracic_aorta.5_peaks.pkl\n",
      "C1PX3_thoracic_aorta.8_peaks.pkl\n",
      "C1PX3_thoracic_aorta.10_peaks.pkl\n",
      "C1PX3_thoracic_aorta.9_peaks.pkl\n",
      "C1PX3_thoracic_aorta.11_peaks.pkl\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.6_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.7_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.2_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.12_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.4_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.1_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.7_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.2_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.12_peaks\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.4_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.6_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.2_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.12_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.4_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.7_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.12_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.4_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.2_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.4_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.12_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.3_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.4_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.3_peaks / C1PX3_thoracic_aorta.5_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.3_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.3_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.3_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.3_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.5_peaks / C1PX3_thoracic_aorta.8_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.5_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.5_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.5_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.8_peaks / C1PX3_thoracic_aorta.10_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.8_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.8_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.10_peaks / C1PX3_thoracic_aorta.9_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.10_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: Calculating foldchange for contrast: C1PX3_thoracic_aorta.9_peaks / C1PX3_thoracic_aorta.11_peaks\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue C1PX3_thoracic_aorta\n",
      "IQYCP_lower_leg_skin.4_peaks.pkl\n",
      "IQYCP_lower_leg_skin.5_peaks.pkl\n",
      "IQYCP_lower_leg_skin.3_peaks.pkl\n",
      "IQYCP_lower_leg_skin.1_peaks.pkl\n",
      "IQYCP_lower_leg_skin.6_peaks.pkl\n",
      "IQYCP_lower_leg_skin.11_peaks.pkl\n",
      "IQYCP_lower_leg_skin.2_peaks.pkl\n",
      "IQYCP_lower_leg_skin.10_peaks.pkl\n",
      "IQYCP_lower_leg_skin.7_peaks.pkl\n",
      "IQYCP_lower_leg_skin.8_peaks.pkl\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.5_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.3_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.1_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.6_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.11_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.4_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.3_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.1_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.6_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.11_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.5_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.1_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.6_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.11_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.7_peaks\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.3_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.6_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.11_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.1_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.6_peaks / IQYCP_lower_leg_skin.11_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.6_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.6_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.6_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.6_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.11_peaks / IQYCP_lower_leg_skin.2_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.11_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.11_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.11_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.2_peaks / IQYCP_lower_leg_skin.10_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.2_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.2_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.10_peaks / IQYCP_lower_leg_skin.7_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.10_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: Calculating foldchange for contrast: IQYCP_lower_leg_skin.7_peaks / IQYCP_lower_leg_skin.8_peaks\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue IQYCP_lower_leg_skin\n",
      "CSSD4_omental_fat_pad.1_peaks.pkl\n",
      "CSSD4_omental_fat_pad.2_peaks.pkl\n",
      "CSSD4_omental_fat_pad.3_peaks.pkl\n",
      "CSSD4_omental_fat_pad.6_peaks.pkl\n",
      "CSSD4_omental_fat_pad.4_peaks.pkl\n",
      "CSSD4_omental_fat_pad.5_peaks.pkl\n",
      "CSSD4_omental_fat_pad.7_peaks.pkl\n",
      "CSSD4_omental_fat_pad.8_peaks.pkl\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.2_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.3_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.6_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.4_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.5_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.1_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.3_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.6_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.4_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.5_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.2_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.3_peaks / CSSD4_omental_fat_pad.6_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.3_peaks / CSSD4_omental_fat_pad.4_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.3_peaks / CSSD4_omental_fat_pad.5_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.3_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.3_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.6_peaks / CSSD4_omental_fat_pad.4_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.6_peaks / CSSD4_omental_fat_pad.5_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.6_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.6_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.4_peaks / CSSD4_omental_fat_pad.5_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.4_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.4_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.5_peaks / CSSD4_omental_fat_pad.7_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.5_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: Calculating foldchange for contrast: CSSD4_omental_fat_pad.7_peaks / CSSD4_omental_fat_pad.8_peaks\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue CSSD4_omental_fat_pad\n",
      "JF1O9_gastrocnemius_medialis.5_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.11_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.3_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.6_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.2_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.12_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.8_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.9_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.1_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.4_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.7_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.13_peaks.pkl\n",
      "JF1O9_gastrocnemius_medialis.10_peaks.pkl\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.11_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.6_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.12_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.5_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.3_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.6_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.12_peaks\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.11_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.6_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.12_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.3_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.2_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.12_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.6_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.12_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.2_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.8_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.12_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.9_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.8_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.9_peaks / JF1O9_gastrocnemius_medialis.1_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.9_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.9_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.9_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.9_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.1_peaks / JF1O9_gastrocnemius_medialis.4_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.1_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.1_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.1_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.4_peaks / JF1O9_gastrocnemius_medialis.7_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.4_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.4_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.7_peaks / JF1O9_gastrocnemius_medialis.13_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.7_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: Calculating foldchange for contrast: JF1O9_gastrocnemius_medialis.13_peaks / JF1O9_gastrocnemius_medialis.10_peaks\n",
      "INFO: The calculated log2fc's are found in the rules table (<DiffCombObj>.rules)\n",
      "Done: Diff analysis for tissue JF1O9_gastrocnemius_medialis\n"
     ]
    }
   ],
   "source": [
    "# Do the differential analysis for the all clusters in a tissue \n",
    "do_differential_analysis_for_tissues(tissues=mb_tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42d4d9b5",
   "metadata": {},
   "source": [
    "## Specific analysis:\n",
    "### Question:\n",
    "1. Find Transcriptionfactor-co-occurences, which only occure in one (or more) \"cluster/celltype of a tissue\".\n",
    "    - Solution with a hard threshold (a, all cluster) or a week threshold (b, celltypes in a tissue):\n",
    "    - a) The found TF-pair occures only in a single cluster of a tissue:\n",
    "        e.g. Cluster1 has TF1-TF1 and the other cluster in the tissue do not have TF1-TF1.\n",
    "    - b) A TF-pair occures significantly only in some clusters:\n",
    "      e.g. We have 3 endothelcells cluster in a tissue and this three clusters share a TF-pair which does not occure in the other cluster of the tissue. (need celltype annotation)\n",
    "####  Steps:\n",
    "   \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "505cb1db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['JF1O6_body_of_pancreas', 'C1PX3_thoracic_aorta', 'IQYCP_lower_leg_skin', 'CSSD4_omental_fat_pad', 'JF1O9_gastrocnemius_medialis']\n"
     ]
    }
   ],
   "source": [
    "# Get the folder names of the tissues that have a differential analysis\n",
    "diff_mb_tissues = get_folder_names_in_folder(rel_folder_path=differential_analysis_path)\n",
    "print(diff_mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "8d56ce5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_significant_rules(df:pd.DataFrame, cosine_col:str, cosine_threshold=0.001, log2fc_threshold_percent=0.05):\n",
    "    reduced_df = df.copy(deep=True)\n",
    "    for col in df.columns:\n",
    "        # calculate thresholds for log2fc columns or cosine_column\n",
    "        \n",
    "        if col == cosine_col:\n",
    "            # for the cosine column we only want to keep rows with a high value (no negative values)\n",
    "            measure_threshold = utils.get_threshold(df[col], \"both\", percent=cosine_threshold)\n",
    "            upper_threshold = measure_threshold[1]\n",
    "            \n",
    "            reduced_df = reduced_df[(reduced_df[col] > upper_threshold)]\n",
    "            print(upper_threshold)\n",
    "        else:\n",
    "            measure_threshold = utils.get_threshold(df[col], \"both\", percent=log2fc_threshold_percent)\n",
    "            upper_threshold = measure_threshold[1]\n",
    "            lower_threshold = measure_threshold[0]\n",
    "            reduced_df = reduced_df[(reduced_df[col] > upper_threshold) | (reduced_df[col] < lower_threshold)]\n",
    "    return reduced_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "540bc39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_specific_tf_cos_for_cluster(df:pd.DataFrame, cluster_name:str) -> pd.DataFrame:\n",
    "    \n",
    "# We try now to get only the columns of the cluster of the diff analysis, we would like to investigate.\n",
    "    # reduce to relevant columns of cluster\n",
    "    cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "    #print(cluster_cols)\n",
    "    #JF1O6_body_of_pancreas.7_peaks/JF1O6_body_of_pancreas.9_peaks_cosine_log2fc\n",
    "    relevant_cluster_cols = []\n",
    "    # add cosien value for of analysed cluster\n",
    "    cluster_cosine_col_name = f\"{cluster_name}_cosine\"\n",
    "    relevant_cluster_cols.append(cluster_cosine_col_name)\n",
    "    for entry in cluster_cols:\n",
    "        if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "            relevant_cluster_cols.append(entry)\n",
    "   # print(logfc_cluster_cols)\n",
    "    print(len(relevant_cluster_cols))\n",
    "\n",
    "    #print(logfc_cluster_cols)\n",
    "    \n",
    "# Now we get only the values (cosine of cluster + log2fc with each contrast of to the cluster) for the cluster to investigate .\n",
    "    reduced_df = df[relevant_cluster_cols]\n",
    "    print(f'Initial Count: {reduced_df.shape}')\n",
    "\n",
    "# Now we filter out 0.00 rows and than filter for signficant values    \n",
    "    # Count all entries in a row , which do not have a zero(0.00) in it.\n",
    "    # e.g. 15 cols have 0.00 => val_counts = 0, 10 cols not have a 0.00 => val_counts = 10 \n",
    "    val_counts = reduced_df[~reduced_df.isin([0])].count(axis=1).sort_values()\n",
    "    \n",
    "    # Set threshold \n",
    "    selection_threshold = len(relevant_cluster_cols) # z.B 15 could be varyied\n",
    "    # Keep all entries,which do have more values different higher than the threshold\n",
    "    tfs_occ = val_counts[val_counts >= selection_threshold].index\n",
    "    result = reduced_df.loc[tfs_occ]\n",
    "    print(f'Zero filtered: {result.shape}')\n",
    "\n",
    "    significants = get_significant_rules(df=result, cosine_col=cluster_cosine_col_name, cosine_threshold=0.001, log2fc_threshold_percent=0.05)\n",
    "    \n",
    "    print(f'Cluster: {cluster_name}: {significants.shape} ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue')\n",
    "    print(f'Done: Find specific tf co-occurences for cluster{cluster_name}.')\n",
    "    \n",
    "    return significants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "a5ce5809",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyse_differential_analysis_per_tissue(tissues=[]):\n",
    "    for tissue in tissues:\n",
    "        print(f\"Start analyse cluster for {tissue}:\")\n",
    "        answer_save_path = f\"{answers_path}{tissue}/\"\n",
    "        # Check if  folder for path already exists for tissue, if not create new one\n",
    "        if not os.path.exists(answer_save_path):\n",
    "             pathlib.Path(answer_save_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_path}{tissue}/{tissue}_normalized.pkl\")\n",
    "        diff_rules = diff_obj.rules     \n",
    "        files_main_mb = read_in_file_names_of_folder(rel_path=f\"{main_analysis_path}{tissue}\")\n",
    "   \n",
    "        for i, file in enumerate(files_main_mb):\n",
    "            # print(file)\n",
    "            cluster_name = file.split('.pkl')[0]\n",
    "            print(f\"Start Find specific tf-cos for: {cluster_name}\")\n",
    "            res = find_specific_tf_cos_for_cluster(df=diff_rules, cluster_name=cluster_name)\n",
    "            res.to_pickle(f\"{answer_save_path}{cluster_name}.pkl\")\n",
    "            \n",
    "\n",
    "           \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "378e96c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start analyse cluster for JF1O6_body_of_pancreas:\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.6_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (149093, 11)\n",
      "0.15500145633635898\n",
      "Cluster: JF1O6_body_of_pancreas.6_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.6_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.1_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (179415, 11)\n",
      "0.17582255058102025\n",
      "Cluster: JF1O6_body_of_pancreas.1_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.1_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.2_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (182953, 11)\n",
      "0.15360308089800387\n",
      "Cluster: JF1O6_body_of_pancreas.2_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.2_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.5_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (187394, 11)\n",
      "0.1357100237567804\n",
      "Cluster: JF1O6_body_of_pancreas.5_peaks: (4, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.5_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.3_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (200961, 11)\n",
      "0.06373685639747682\n",
      "Cluster: JF1O6_body_of_pancreas.3_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.3_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.4_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (148072, 11)\n",
      "0.1695326013487633\n",
      "Cluster: JF1O6_body_of_pancreas.4_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.4_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.10_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (58342, 11)\n",
      "0.16736775420337166\n",
      "Cluster: JF1O6_body_of_pancreas.10_peaks: (3, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.10_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.8_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (102360, 11)\n",
      "0.19901228616324362\n",
      "Cluster: JF1O6_body_of_pancreas.8_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.8_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.11_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (66730, 11)\n",
      "0.15826250253379523\n",
      "Cluster: JF1O6_body_of_pancreas.11_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.11_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.7_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (218044, 11)\n",
      "0.06619688043211004\n",
      "Cluster: JF1O6_body_of_pancreas.7_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.7_peaks.\n",
      "Start Find specific tf-cos for: JF1O6_body_of_pancreas.9_peaks\n",
      "11\n",
      "Initial Count: (251581, 11)\n",
      "Zero filtered: (211216, 11)\n",
      "0.06523147733145261\n",
      "Cluster: JF1O6_body_of_pancreas.9_peaks: (0, 11) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O6_body_of_pancreas.9_peaks.\n",
      "Start analyse cluster for C1PX3_thoracic_aorta:\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.1_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (89083, 12)\n",
      "0.15044667266976378\n",
      "Cluster: C1PX3_thoracic_aorta.1_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.1_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.6_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (135717, 12)\n",
      "0.14930733423220705\n",
      "Cluster: C1PX3_thoracic_aorta.6_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.6_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.7_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (70464, 12)\n",
      "0.15604317420989847\n",
      "Cluster: C1PX3_thoracic_aorta.7_peaks: (5, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.7_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.2_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (144270, 12)\n",
      "0.13709678556978908\n",
      "Cluster: C1PX3_thoracic_aorta.2_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.2_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.12_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (167611, 12)\n",
      "0.12701472140728012\n",
      "Cluster: C1PX3_thoracic_aorta.12_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.12_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.4_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (228708, 12)\n",
      "0.0707199668967861\n",
      "Cluster: C1PX3_thoracic_aorta.4_peaks: (2, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.4_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.3_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (221110, 12)\n",
      "0.06569532386862517\n",
      "Cluster: C1PX3_thoracic_aorta.3_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.3_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.5_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (235896, 12)\n",
      "0.07034053419553232\n",
      "Cluster: C1PX3_thoracic_aorta.5_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.5_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.8_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (163100, 12)\n",
      "0.13606867524959365\n",
      "Cluster: C1PX3_thoracic_aorta.8_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.8_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.10_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (172058, 12)\n",
      "0.10422535427075025\n",
      "Cluster: C1PX3_thoracic_aorta.10_peaks: (1, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.10_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.9_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (173489, 12)\n",
      "0.10408004102097303\n",
      "Cluster: C1PX3_thoracic_aorta.9_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.9_peaks.\n",
      "Start Find specific tf-cos for: C1PX3_thoracic_aorta.11_peaks\n",
      "12\n",
      "Initial Count: (250407, 12)\n",
      "Zero filtered: (171689, 12)\n",
      "0.10871770232653676\n",
      "Cluster: C1PX3_thoracic_aorta.11_peaks: (0, 12) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterC1PX3_thoracic_aorta.11_peaks.\n",
      "Start analyse cluster for IQYCP_lower_leg_skin:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.4_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (7749, 10)\n",
      "0.7883300905832706\n",
      "Cluster: IQYCP_lower_leg_skin.4_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.4_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.5_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (51311, 10)\n",
      "0.3542950618360161\n",
      "Cluster: IQYCP_lower_leg_skin.5_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.5_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.3_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (86066, 10)\n",
      "0.3199389483393252\n",
      "Cluster: IQYCP_lower_leg_skin.3_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.3_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.1_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (91158, 10)\n",
      "0.26779930254969603\n",
      "Cluster: IQYCP_lower_leg_skin.1_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.1_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.6_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (137538, 10)\n",
      "0.31033654866129734\n",
      "Cluster: IQYCP_lower_leg_skin.6_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.6_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.11_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (65695, 10)\n",
      "0.33804386162328687\n",
      "Cluster: IQYCP_lower_leg_skin.11_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.11_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.2_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (144795, 10)\n",
      "0.2755092475268185\n",
      "Cluster: IQYCP_lower_leg_skin.2_peaks: (1, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.2_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.10_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (176853, 10)\n",
      "0.31816256464531867\n",
      "Cluster: IQYCP_lower_leg_skin.10_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.10_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.7_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (163487, 10)\n",
      "0.3184549945531403\n",
      "Cluster: IQYCP_lower_leg_skin.7_peaks: (0, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.7_peaks.\n",
      "Start Find specific tf-cos for: IQYCP_lower_leg_skin.8_peaks\n",
      "10\n",
      "Initial Count: (215420, 10)\n",
      "Zero filtered: (31985, 10)\n",
      "0.39298490827168253\n",
      "Cluster: IQYCP_lower_leg_skin.8_peaks: (3, 10) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterIQYCP_lower_leg_skin.8_peaks.\n",
      "Start analyse cluster for CSSD4_omental_fat_pad:\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.1_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (42757, 8)\n",
      "0.21903474924553276\n",
      "Cluster: CSSD4_omental_fat_pad.1_peaks: (17, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.1_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.2_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (187121, 8)\n",
      "0.2006510898866939\n",
      "Cluster: CSSD4_omental_fat_pad.2_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.2_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.3_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (104725, 8)\n",
      "0.1762499558354273\n",
      "Cluster: CSSD4_omental_fat_pad.3_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.3_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.6_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (220224, 8)\n",
      "0.1854337620456882\n",
      "Cluster: CSSD4_omental_fat_pad.6_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.6_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.4_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (157073, 8)\n",
      "0.2114360944006682\n",
      "Cluster: CSSD4_omental_fat_pad.4_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.4_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.5_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (106120, 8)\n",
      "0.1433624105089424\n",
      "Cluster: CSSD4_omental_fat_pad.5_peaks: (1, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.5_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.7_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (202632, 8)\n",
      "0.18039910799491707\n",
      "Cluster: CSSD4_omental_fat_pad.7_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.7_peaks.\n",
      "Start Find specific tf-cos for: CSSD4_omental_fat_pad.8_peaks\n",
      "8\n",
      "Initial Count: (250935, 8)\n",
      "Zero filtered: (231287, 8)\n",
      "0.07194696338344887\n",
      "Cluster: CSSD4_omental_fat_pad.8_peaks: (0, 8) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterCSSD4_omental_fat_pad.8_peaks.\n",
      "Start analyse cluster for JF1O9_gastrocnemius_medialis:\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.5_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (97790, 13)\n",
      "0.20162807135236702\n",
      "Cluster: JF1O9_gastrocnemius_medialis.5_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.5_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.11_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (46653, 13)\n",
      "0.21268879624334247\n",
      "Cluster: JF1O9_gastrocnemius_medialis.11_peaks: (3, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.11_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.3_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (60974, 13)\n",
      "0.20947474107112996\n",
      "Cluster: JF1O9_gastrocnemius_medialis.3_peaks: (9, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.3_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.6_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (211420, 13)\n",
      "0.08069462354668158\n",
      "Cluster: JF1O9_gastrocnemius_medialis.6_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.6_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.2_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (129317, 13)\n",
      "0.16914717608001217\n",
      "Cluster: JF1O9_gastrocnemius_medialis.2_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.2_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.12_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (164317, 13)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1945669085241149\n",
      "Cluster: JF1O9_gastrocnemius_medialis.12_peaks: (1, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.12_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.8_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (155647, 13)\n",
      "0.2158563620313507\n",
      "Cluster: JF1O9_gastrocnemius_medialis.8_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.8_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.9_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (170548, 13)\n",
      "0.1789962046112875\n",
      "Cluster: JF1O9_gastrocnemius_medialis.9_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.9_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.1_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (238899, 13)\n",
      "0.08209617886873954\n",
      "Cluster: JF1O9_gastrocnemius_medialis.1_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.1_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.4_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (222937, 13)\n",
      "0.0820357452366778\n",
      "Cluster: JF1O9_gastrocnemius_medialis.4_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.4_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.7_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (119536, 13)\n",
      "0.17453984825019733\n",
      "Cluster: JF1O9_gastrocnemius_medialis.7_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.7_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.13_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (107990, 13)\n",
      "0.18292060263298288\n",
      "Cluster: JF1O9_gastrocnemius_medialis.13_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.13_peaks.\n",
      "Start Find specific tf-cos for: JF1O9_gastrocnemius_medialis.10_peaks\n",
      "13\n",
      "Initial Count: (247335, 13)\n",
      "Zero filtered: (185734, 13)\n",
      "0.0781061703781069\n",
      "Cluster: JF1O9_gastrocnemius_medialis.10_peaks: (0, 13) ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue\n",
      "Done: Find specific tf co-occurences for clusterJF1O9_gastrocnemius_medialis.10_peaks.\n"
     ]
    }
   ],
   "source": [
    "analyse_differential_analysis_per_tissue(tissues=diff_mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d4dc4c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rows_with_same_value(df:pd.DataFrame) -> pd.DataFrame:\n",
    "    return df[df.apply(lambda x: min(abs(x)) == max(abs(x)), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "dcde71dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.ipynb_checkpoints', 'JF1O6_body_of_pancreas', 'C1PX3_thoracic_aorta', 'IQYCP_lower_leg_skin', 'CSSD4_omental_fat_pad', 'JF1O9_gastrocnemius_medialis']\n"
     ]
    }
   ],
   "source": [
    "answers_mb_tissues = get_folder_names_in_folder(rel_folder_path=answers_path)\n",
    "print(answers_mb_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "85d4c2df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start analyse .ipynb_checkpoints:\n",
      "Start analyse JF1O6_body_of_pancreas:\n",
      "Start analyse C1PX3_thoracic_aorta:\n",
      "Found cluster with only occurence\n",
      "Start analyse IQYCP_lower_leg_skin:\n",
      "Start analyse CSSD4_omental_fat_pad:\n",
      "Start analyse JF1O9_gastrocnemius_medialis:\n"
     ]
    }
   ],
   "source": [
    "cluster_with_specials= []\n",
    "for tissue in answers_mb_tissues:\n",
    "    \n",
    "    print(f\"Start analysis {tissue}:\")\n",
    "    answer_specials_save_path = f\"{answers_path}{tissue}/specials/\"\n",
    "    # Check if  folder for path already exists for tissue, if not create new one\n",
    "    if not os.path.exists(answer_specials_save_path):\n",
    "         pathlib.Path(answer_specials_save_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    files = read_in_file_names_of_folder(rel_path=f\"{answers_path}{tissue}\")\n",
    "    for i, file in enumerate(files):\n",
    "        # print(file)\n",
    "        cluster_name = file.split('.pkl')[0]\n",
    "        df = pd.read_pickle(f\"{answers_path}{tissue}/{file}\")\n",
    "\n",
    "        # remove cosine from dataframe, to get all rows of the datafram, that have the same log2fc\n",
    "        df_copy = df.copy()\n",
    "        cosine_col_name = df_copy.columns[0]\n",
    "        cosine_col = df_copy[cosine_col_name]\n",
    "        df_only_log2fc = df_copy.drop(columns=[cosine_col_name])\n",
    "        #\n",
    "        df_res = find_rows_with_same_value(df_only_log2fc)\n",
    "        # check if we found a tf-coocurence which is significant and only occured in one cluster of a tissue\n",
    "        if df_res.shape[0] > 0: \n",
    "            print(f\"Found cluster: {cluster_name} with only-occurences in tissue: {df_res.shape[0]}   \")\n",
    "            #add cosine value back to dataframe\n",
    "            df_res[cosine_col_name]= cosine_col\n",
    "            cluster_with_specials.append(df_res)\n",
    "            df_res.to_csv(f\"{answer_specials_save_path}{cluster_name}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "30c3c481",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TBXT-TBXT</th>\n",
       "      <th>TBXT-GLI3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.1_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc</th>\n",
       "      <td>-6.488312</td>\n",
       "      <td>-4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.6_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc</th>\n",
       "      <td>-6.488312</td>\n",
       "      <td>-4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.2_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.12_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.4_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.3_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.5_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.8_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.10_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.9_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.11_peaks_cosine_log2fc</th>\n",
       "      <td>6.488312</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks_cosine</th>\n",
       "      <td>0.564035</td>\n",
       "      <td>0.186516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    TBXT-TBXT  TBXT-GLI3\n",
       "C1PX3_thoracic_aorta.1_peaks/C1PX3_thoracic_aor...  -6.488312  -4.923988\n",
       "C1PX3_thoracic_aorta.6_peaks/C1PX3_thoracic_aor...  -6.488312  -4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aor...   6.488312   4.923988\n",
       "C1PX3_thoracic_aorta.7_peaks_cosine                  0.564035   0.186516"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cluster_with_specials)\n",
    "cluster_with_specials[0].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "8a3d1f74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <th>TF1_count</th>\n",
       "      <th>TF2_count</th>\n",
       "      <th>cosine</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>140720.000000</td>\n",
       "      <td>140720.000000</td>\n",
       "      <td>140720.000000</td>\n",
       "      <td>140720.000000</td>\n",
       "      <td>1.407200e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>5.986640</td>\n",
       "      <td>323.169187</td>\n",
       "      <td>323.169187</td>\n",
       "      <td>0.020943</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>45.669842</td>\n",
       "      <td>619.554824</td>\n",
       "      <td>619.554824</td>\n",
       "      <td>0.024083</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002089</td>\n",
       "      <td>-1.136139e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>0.010341</td>\n",
       "      <td>-4.359000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>108.000000</td>\n",
       "      <td>108.000000</td>\n",
       "      <td>0.015873</td>\n",
       "      <td>4.833641e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>275.000000</td>\n",
       "      <td>275.000000</td>\n",
       "      <td>0.024510</td>\n",
       "      <td>1.611559e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3451.000000</td>\n",
       "      <td>4841.000000</td>\n",
       "      <td>4841.000000</td>\n",
       "      <td>0.828964</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF1_TF2_count      TF1_count      TF2_count         cosine  \\\n",
       "count  140720.000000  140720.000000  140720.000000  140720.000000   \n",
       "mean        5.986640     323.169187     323.169187       0.020943   \n",
       "std        45.669842     619.554824     619.554824       0.024083   \n",
       "min         1.000000       1.000000       1.000000       0.002089   \n",
       "25%         1.000000      46.000000      46.000000       0.010341   \n",
       "50%         1.000000     108.000000     108.000000       0.015873   \n",
       "75%         3.000000     275.000000     275.000000       0.024510   \n",
       "max      3451.000000    4841.000000    4841.000000       0.828964   \n",
       "\n",
       "             zscore  \n",
       "count  1.407200e+05  \n",
       "mean            inf  \n",
       "std             NaN  \n",
       "min   -1.136139e+01  \n",
       "25%   -4.359000e-01  \n",
       "50%    4.833641e-01  \n",
       "75%    1.611559e+00  \n",
       "max             inf  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb = CombObj().from_pickle(f\"{main_analysis_path}/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta.7_peaks.pkl\")\n",
    "comb.rules.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "2493c4c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TBXT-GLI3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TF1</th>\n",
       "      <td>TBXT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF2</th>\n",
       "      <td>GLI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF1_count</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF2_count</th>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cosine</th>\n",
       "      <td>0.19355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zscore</th>\n",
       "      <td>35.517601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               TBXT-GLI3\n",
       "TF1                 TBXT\n",
       "TF2                 GLI3\n",
       "TF1_TF2_count          7\n",
       "TF1_count             12\n",
       "TF2_count            109\n",
       "cosine           0.19355\n",
       "zscore         35.517601"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(comb.rules.loc['TBXT-GLI3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "3d951a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle(f\"{answers_path}/C1PX3_thoracic_aorta/C1PX3_thoracic_aorta.7_peaks.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "2ef9d5d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks_cosine</th>\n",
       "      <th>C1PX3_thoracic_aorta.1_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.6_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.2_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.12_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.4_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.3_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.5_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.8_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.10_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.9_peaks_cosine_log2fc</th>\n",
       "      <th>C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.11_peaks_cosine_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Arid3b-PHOX2A</th>\n",
       "      <td>0.162719</td>\n",
       "      <td>-4.734005</td>\n",
       "      <td>-4.734005</td>\n",
       "      <td>2.119932</td>\n",
       "      <td>1.665355</td>\n",
       "      <td>4.469920</td>\n",
       "      <td>3.551583</td>\n",
       "      <td>2.794985</td>\n",
       "      <td>3.697820</td>\n",
       "      <td>3.257512</td>\n",
       "      <td>3.340073</td>\n",
       "      <td>4.734005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBXT-TBXT</th>\n",
       "      <td>0.564035</td>\n",
       "      <td>-6.488312</td>\n",
       "      <td>-6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "      <td>6.488312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBXT-JUNvar.2</th>\n",
       "      <td>0.471684</td>\n",
       "      <td>-6.233492</td>\n",
       "      <td>-6.233492</td>\n",
       "      <td>5.136438</td>\n",
       "      <td>6.233492</td>\n",
       "      <td>4.800954</td>\n",
       "      <td>5.062541</td>\n",
       "      <td>4.819318</td>\n",
       "      <td>4.625762</td>\n",
       "      <td>6.233492</td>\n",
       "      <td>6.233492</td>\n",
       "      <td>4.472095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBXT-GLI3</th>\n",
       "      <td>0.186516</td>\n",
       "      <td>-4.923988</td>\n",
       "      <td>-4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "      <td>4.923988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBXT-GLI2</th>\n",
       "      <td>0.180938</td>\n",
       "      <td>-4.881649</td>\n",
       "      <td>-4.881649</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.739540</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.768487</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.881649</td>\n",
       "      <td>4.881649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               C1PX3_thoracic_aorta.7_peaks_cosine  \\\n",
       "Arid3b-PHOX2A                             0.162719   \n",
       "TBXT-TBXT                                 0.564035   \n",
       "TBXT-JUNvar.2                             0.471684   \n",
       "TBXT-GLI3                                 0.186516   \n",
       "TBXT-GLI2                                 0.180938   \n",
       "\n",
       "               C1PX3_thoracic_aorta.1_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                          -4.734005                         \n",
       "TBXT-TBXT                                              -6.488312                         \n",
       "TBXT-JUNvar.2                                          -6.233492                         \n",
       "TBXT-GLI3                                              -4.923988                         \n",
       "TBXT-GLI2                                              -4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.6_peaks/C1PX3_thoracic_aorta.7_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                          -4.734005                         \n",
       "TBXT-TBXT                                              -6.488312                         \n",
       "TBXT-JUNvar.2                                          -6.233492                         \n",
       "TBXT-GLI3                                              -4.923988                         \n",
       "TBXT-GLI2                                              -4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.2_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           2.119932                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           5.136438                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.12_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           1.665355                          \n",
       "TBXT-TBXT                                               6.488312                          \n",
       "TBXT-JUNvar.2                                           6.233492                          \n",
       "TBXT-GLI3                                               4.923988                          \n",
       "TBXT-GLI2                                               4.881649                          \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.4_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           4.469920                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           4.800954                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.739540                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.3_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           3.551583                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           5.062541                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.5_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           2.794985                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           4.819318                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.768487                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.8_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           3.697820                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           4.625762                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.10_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           3.257512                          \n",
       "TBXT-TBXT                                               6.488312                          \n",
       "TBXT-JUNvar.2                                           6.233492                          \n",
       "TBXT-GLI3                                               4.923988                          \n",
       "TBXT-GLI2                                               4.881649                          \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.9_peaks_cosine_log2fc  \\\n",
       "Arid3b-PHOX2A                                           3.340073                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           6.233492                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.881649                         \n",
       "\n",
       "               C1PX3_thoracic_aorta.7_peaks/C1PX3_thoracic_aorta.11_peaks_cosine_log2fc  \n",
       "Arid3b-PHOX2A                                           4.734005                         \n",
       "TBXT-TBXT                                               6.488312                         \n",
       "TBXT-JUNvar.2                                           4.472095                         \n",
       "TBXT-GLI3                                               4.923988                         \n",
       "TBXT-GLI2                                               4.881649                         "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d5e034e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>JF1O6_body_of_pancreas.6_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.1_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.2_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.3_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.4_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.10_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.8_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.11_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.7_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.9_peaks_cosine_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RARARXRA-Rarbvar.2</th>\n",
       "      <td>-2.125769</td>\n",
       "      <td>-4.785241</td>\n",
       "      <td>-2.895764</td>\n",
       "      <td>3.944596</td>\n",
       "      <td>4.232885</td>\n",
       "      <td>5.360727</td>\n",
       "      <td>3.461222</td>\n",
       "      <td>5.360727</td>\n",
       "      <td>3.920365</td>\n",
       "      <td>3.353811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rarbvar.2-RARARXRG</th>\n",
       "      <td>-2.397617</td>\n",
       "      <td>-4.223388</td>\n",
       "      <td>-2.601964</td>\n",
       "      <td>3.896619</td>\n",
       "      <td>5.302650</td>\n",
       "      <td>5.302650</td>\n",
       "      <td>5.302650</td>\n",
       "      <td>5.302650</td>\n",
       "      <td>3.539295</td>\n",
       "      <td>3.047005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGIF2LX-TGIF1</th>\n",
       "      <td>-5.324715</td>\n",
       "      <td>-1.535567</td>\n",
       "      <td>-1.552164</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "      <td>5.324715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGIF2-TGIF2LX</th>\n",
       "      <td>-5.291761</td>\n",
       "      <td>-1.531016</td>\n",
       "      <td>-1.553928</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>5.291761</td>\n",
       "      <td>4.749091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    JF1O6_body_of_pancreas.6_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                          -2.125769                             \n",
       "Rarbvar.2-RARARXRG                                          -2.397617                             \n",
       "TGIF2LX-TGIF1                                               -5.324715                             \n",
       "TGIF2-TGIF2LX                                               -5.291761                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.1_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                          -4.785241                             \n",
       "Rarbvar.2-RARARXRG                                          -4.223388                             \n",
       "TGIF2LX-TGIF1                                               -1.535567                             \n",
       "TGIF2-TGIF2LX                                               -1.531016                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.2_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                          -2.895764                             \n",
       "Rarbvar.2-RARARXRG                                          -2.601964                             \n",
       "TGIF2LX-TGIF1                                               -1.552164                             \n",
       "TGIF2-TGIF2LX                                               -1.553928                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.3_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           3.944596                             \n",
       "Rarbvar.2-RARARXRG                                           3.896619                             \n",
       "TGIF2LX-TGIF1                                                5.324715                             \n",
       "TGIF2-TGIF2LX                                                5.291761                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.4_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           4.232885                             \n",
       "Rarbvar.2-RARARXRG                                           5.302650                             \n",
       "TGIF2LX-TGIF1                                                5.324715                             \n",
       "TGIF2-TGIF2LX                                                5.291761                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.10_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           5.360727                              \n",
       "Rarbvar.2-RARARXRG                                           5.302650                              \n",
       "TGIF2LX-TGIF1                                                5.324715                              \n",
       "TGIF2-TGIF2LX                                                5.291761                              \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.8_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           3.461222                             \n",
       "Rarbvar.2-RARARXRG                                           5.302650                             \n",
       "TGIF2LX-TGIF1                                                5.324715                             \n",
       "TGIF2-TGIF2LX                                                5.291761                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.11_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           5.360727                              \n",
       "Rarbvar.2-RARARXRG                                           5.302650                              \n",
       "TGIF2LX-TGIF1                                                5.324715                              \n",
       "TGIF2-TGIF2LX                                                5.291761                              \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.7_peaks_cosine_log2fc  \\\n",
       "RARARXRA-Rarbvar.2                                           3.920365                             \n",
       "Rarbvar.2-RARARXRG                                           3.539295                             \n",
       "TGIF2LX-TGIF1                                                5.324715                             \n",
       "TGIF2-TGIF2LX                                                5.291761                             \n",
       "\n",
       "                    JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.9_peaks_cosine_log2fc  \n",
       "RARARXRA-Rarbvar.2                                           3.353811                            \n",
       "Rarbvar.2-RARARXRG                                           3.047005                            \n",
       "TGIF2LX-TGIF1                                                5.324715                            \n",
       "TGIF2-TGIF2LX                                                4.749091                            "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_copied = df.copy()\n",
    "\n",
    "cosine_col_name = df_copied.columns[0]\n",
    "cosine_col = df_copied[cosine_col_name]\n",
    "dropped = df_copied.drop(columns=[cosine_col_name])\n",
    "dropped\n",
    "#df_copied.drop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "ebf5d5de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>JF1O6_body_of_pancreas.6_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.1_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.2_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.3_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.4_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.10_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.8_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.11_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.7_peaks_cosine_log2fc</th>\n",
       "      <th>JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.9_peaks_cosine_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [JF1O6_body_of_pancreas.6_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc, JF1O6_body_of_pancreas.1_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc, JF1O6_body_of_pancreas.2_peaks/JF1O6_body_of_pancreas.5_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.3_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.4_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.10_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.8_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.11_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.7_peaks_cosine_log2fc, JF1O6_body_of_pancreas.5_peaks/JF1O6_body_of_pancreas.9_peaks_cosine_log2fc]\n",
       "Index: []"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df1 = find_rows_with_same_value(dropped)\n",
    "df1[cosine_col_name]= cosine_col\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "caae383c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35, 11)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "6d6bb600",
   "metadata": {},
   "outputs": [],
   "source": [
    "combObj = CombObj().from_pickle(f\"{main_analysis_path}/JF1O6_body_of_pancreas/JF1O6_body_of_pancreas.3_peaks.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "2edfbfd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "combObj.rules\n",
    "\n",
    "c = combObj.select_top_rules(n=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "a12f2cb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0009596639805392853, 0.06808819506421941)"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measure_threshold = utils.get_threshold(combObj.rules['cosine'], \"both\", percent=0.001)\n",
    "measure_threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ba5cf7f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <th>TF1_count</th>\n",
       "      <th>TF2_count</th>\n",
       "      <th>cosine</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>401559.000000</td>\n",
       "      <td>401559.000000</td>\n",
       "      <td>401559.000000</td>\n",
       "      <td>401559.000000</td>\n",
       "      <td>4.015590e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>9.106958</td>\n",
       "      <td>879.412308</td>\n",
       "      <td>879.412308</td>\n",
       "      <td>0.008374</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>69.216902</td>\n",
       "      <td>1322.606931</td>\n",
       "      <td>1322.606931</td>\n",
       "      <td>0.013691</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>26.000000</td>\n",
       "      <td>26.000000</td>\n",
       "      <td>0.000475</td>\n",
       "      <td>-2.756006e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>297.000000</td>\n",
       "      <td>297.000000</td>\n",
       "      <td>0.003639</td>\n",
       "      <td>-8.879052e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>492.000000</td>\n",
       "      <td>492.000000</td>\n",
       "      <td>0.005952</td>\n",
       "      <td>-8.671100e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>6.000000</td>\n",
       "      <td>793.000000</td>\n",
       "      <td>793.000000</td>\n",
       "      <td>0.009614</td>\n",
       "      <td>8.049845e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>8503.000000</td>\n",
       "      <td>12379.000000</td>\n",
       "      <td>12379.000000</td>\n",
       "      <td>0.791416</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF1_TF2_count      TF1_count      TF2_count         cosine  \\\n",
       "count  401559.000000  401559.000000  401559.000000  401559.000000   \n",
       "mean        9.106958     879.412308     879.412308       0.008374   \n",
       "std        69.216902    1322.606931    1322.606931       0.013691   \n",
       "min         1.000000      26.000000      26.000000       0.000475   \n",
       "25%         1.000000     297.000000     297.000000       0.003639   \n",
       "50%         3.000000     492.000000     492.000000       0.005952   \n",
       "75%         6.000000     793.000000     793.000000       0.009614   \n",
       "max      8503.000000   12379.000000   12379.000000       0.791416   \n",
       "\n",
       "             zscore  \n",
       "count  4.015590e+05  \n",
       "mean            inf  \n",
       "std             NaN  \n",
       "min   -2.756006e+01  \n",
       "25%   -8.879052e-01  \n",
       "50%   -8.671100e-02  \n",
       "75%    8.049845e-01  \n",
       "max             inf  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combObj.rules.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "927e2377",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1</th>\n",
       "      <th>TF2</th>\n",
       "      <th>TF1_TF2_count</th>\n",
       "      <th>TF1_count</th>\n",
       "      <th>TF2_count</th>\n",
       "      <th>cosine</th>\n",
       "      <th>zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KLF15-KLF5</th>\n",
       "      <td>KLF15</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>8503</td>\n",
       "      <td>12379</td>\n",
       "      <td>9325</td>\n",
       "      <td>0.791416</td>\n",
       "      <td>169.958347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF5-KLF15</th>\n",
       "      <td>KLF5</td>\n",
       "      <td>KLF15</td>\n",
       "      <td>8503</td>\n",
       "      <td>9325</td>\n",
       "      <td>12379</td>\n",
       "      <td>0.791416</td>\n",
       "      <td>169.958347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF5-SP3</th>\n",
       "      <td>KLF5</td>\n",
       "      <td>SP3</td>\n",
       "      <td>5053</td>\n",
       "      <td>9325</td>\n",
       "      <td>5960</td>\n",
       "      <td>0.677800</td>\n",
       "      <td>153.997648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP3-KLF5</th>\n",
       "      <td>SP3</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>5053</td>\n",
       "      <td>5960</td>\n",
       "      <td>9325</td>\n",
       "      <td>0.677800</td>\n",
       "      <td>153.997648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP1-SP4</th>\n",
       "      <td>SP1</td>\n",
       "      <td>SP4</td>\n",
       "      <td>3391</td>\n",
       "      <td>5230</td>\n",
       "      <td>4946</td>\n",
       "      <td>0.666730</td>\n",
       "      <td>171.325065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zfx-MEF2A</th>\n",
       "      <td>Zfx</td>\n",
       "      <td>MEF2A</td>\n",
       "      <td>218</td>\n",
       "      <td>4575</td>\n",
       "      <td>2223</td>\n",
       "      <td>0.068358</td>\n",
       "      <td>8.578725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF5-NR2C2</th>\n",
       "      <td>KLF5</td>\n",
       "      <td>NR2C2</td>\n",
       "      <td>226</td>\n",
       "      <td>9325</td>\n",
       "      <td>1174</td>\n",
       "      <td>0.068305</td>\n",
       "      <td>6.957267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR2C2-KLF5</th>\n",
       "      <td>NR2C2</td>\n",
       "      <td>KLF5</td>\n",
       "      <td>226</td>\n",
       "      <td>1174</td>\n",
       "      <td>9325</td>\n",
       "      <td>0.068305</td>\n",
       "      <td>6.957267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ELK1-FLI1</th>\n",
       "      <td>ELK1</td>\n",
       "      <td>FLI1</td>\n",
       "      <td>42</td>\n",
       "      <td>828</td>\n",
       "      <td>459</td>\n",
       "      <td>0.068128</td>\n",
       "      <td>15.256272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLI1-ELK1</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>42</td>\n",
       "      <td>459</td>\n",
       "      <td>828</td>\n",
       "      <td>0.068128</td>\n",
       "      <td>15.256272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1975 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TF1    TF2  TF1_TF2_count  TF1_count  TF2_count    cosine  \\\n",
       "KLF15-KLF5  KLF15   KLF5           8503      12379       9325  0.791416   \n",
       "KLF5-KLF15   KLF5  KLF15           8503       9325      12379  0.791416   \n",
       "KLF5-SP3     KLF5    SP3           5053       9325       5960  0.677800   \n",
       "SP3-KLF5      SP3   KLF5           5053       5960       9325  0.677800   \n",
       "SP1-SP4       SP1    SP4           3391       5230       4946  0.666730   \n",
       "...           ...    ...            ...        ...        ...       ...   \n",
       "Zfx-MEF2A     Zfx  MEF2A            218       4575       2223  0.068358   \n",
       "KLF5-NR2C2   KLF5  NR2C2            226       9325       1174  0.068305   \n",
       "NR2C2-KLF5  NR2C2   KLF5            226       1174       9325  0.068305   \n",
       "ELK1-FLI1    ELK1   FLI1             42        828        459  0.068128   \n",
       "FLI1-ELK1    FLI1   ELK1             42        459        828  0.068128   \n",
       "\n",
       "                zscore  \n",
       "KLF15-KLF5  169.958347  \n",
       "KLF5-KLF15  169.958347  \n",
       "KLF5-SP3    153.997648  \n",
       "SP3-KLF5    153.997648  \n",
       "SP1-SP4     171.325065  \n",
       "...                ...  \n",
       "Zfx-MEF2A     8.578725  \n",
       "KLF5-NR2C2    6.957267  \n",
       "NR2C2-KLF5    6.957267  \n",
       "ELK1-FLI1    15.256272  \n",
       "FLI1-ELK1    15.256272  \n",
       "\n",
       "[1975 rows x 7 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#c.rules\n",
    "r = combObj.rules\n",
    "r[(r['cosine'] >= 0.06808819506421941)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "09e66116",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TF1                  OVOL1\n",
       "TF2                   OTX1\n",
       "TF1_TF2_count            2\n",
       "TF1_count               38\n",
       "TF2_count               87\n",
       "cosine            0.034784\n",
       "zscore           10.002083\n",
       "Name: OVOL1-OTX1, dtype: object"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r.loc['OVOL1-OTX1']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6642cd7",
   "metadata": {},
   "source": [
    "# OLD\n",
    "# -----------------------------------------------------------------------------------------\n",
    "# OLD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9597b944",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analyses with whole diffcombj\n",
    "diff_file_names=read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "\n",
    "normalized_diff_objects = []\n",
    "diff_objects = []\n",
    "\n",
    "for file in diff_file_names:\n",
    "    obj = DiffCombObj().from_pickle(f\"{differential_analysis_path}{file}\")\n",
    "    if \"normalized\" in file:\n",
    "        normalized_diff_objects.append(obj)\n",
    "    else:\n",
    "        diff_objects.append(obj)\n",
    "\n",
    "print(normalized_diff_objects)\n",
    "print(diff_objects) \n",
    "\n",
    "normalized_dfs = []\n",
    "for obj in normalized_diff_objects:\n",
    "    normalized_dfs.append(obj.rules)\n",
    "print(\"Done: Preparing rules of DiffObj\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a11d6ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj = normalized_diff_objects[0]\n",
    "t = normalized_diff_objects[0].rules\n",
    "t[t['right-lobe-of-liver.10_cosine'] > 0.7]\n",
    "obj.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3836030f",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df = []\n",
    "for df in normalized_dfs:\n",
    "\n",
    "    for i, file in enumerate(files_main_mb):\n",
    "        # print(file)\n",
    "        cluster_name = file.split('.pkl')[0]\n",
    "        print(cluster_name)\n",
    "        \n",
    "        # reduce to relevant columns of cluster\n",
    "        cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "        # NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "        # This is important: for right-lob-of-liver-1 ,     #right-lobe-of-liver.10_cosine\n",
    "        logfc_cluster_cols = []\n",
    "        for entry in cluster_cols:\n",
    "            if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "                logfc_cluster_cols.append(entry)\n",
    "        \n",
    "        #print(logfc_cluster_cols)\n",
    "        print(len(logfc_cluster_cols))\n",
    "        #print(logfc_cluster_cols)\n",
    "        \n",
    "        reduced_df = df[logfc_cluster_cols]\n",
    "        print(f'Initial Count: {reduced_df.shape}')\n",
    "        \n",
    "        # Count all entries in a row , which do not have a zero(0.00) in it.\n",
    "        # e.g. 15 cols have 0.00 => val_counts = 0, 10 cols not have a 0.00 => val_counts = 10 \n",
    "        val_counts = reduced_df[~reduced_df.isin([0])].count(axis=1).sort_values()\n",
    "        #print(tmp_val_counts)\n",
    "        # Set threshold \n",
    "        selection_threshold = len(logfc_cluster_cols) # z.B 15 could be varyied\n",
    "        \n",
    "        ## Keep all entries,which do have more values different higher than the threshold\n",
    "        tfs_occ = val_counts[val_counts >= selection_threshold].index\n",
    "        result = reduced_df.loc[tfs_occ]\n",
    "        print(f'Zero filtered: {result.shape}')\n",
    "        \n",
    "        significants = get_significant_log2fc_rules(result, threshold_percent=0.05)\n",
    "        results_df.append(significants)\n",
    "        print(f'Cluster: {cluster_name}: {significants.shape} ,tf-pairs with significant log2fc-changes in comparison to all the other clusters in tissue: {tissue_name} ')\n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af3a075",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df1 = results_df[0]\n",
    "print(df1.shape)\n",
    "df1\n",
    "# found tf-pairs, which have significant changes in comparison to all the other clusters in the tissue\n",
    "# this 10 tf-pairs of cluster 10, show significant changes in comparion to all the other clusters in the tissue right lob of liver. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42fd4e9f",
   "metadata": {},
   "source": [
    "#####  nur cosine werte vergleichbar, daher zum finden von spezifischen tf-cos nur cosines nutzen.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b900b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = compare_obj.rules\n",
    "df = pd.DataFrame(td.loc['EN2-MYBL2'])\n",
    "df.head(50)\n",
    "test\n",
    "\n",
    "# nur cosine werte vergleichbar, daher zum finden von spezifischen tf-cos nur cosines nutzen.  \n",
    "np.percentile(test['right-lobe-of-liver.11_cosine'], 75)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16a033b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c684a4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "test['right-lobe-of-liver.11_cosine'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a48bb61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster 6\n",
    "df1 = results_df[13]\n",
    "print(df1.shape)\n",
    "df1.head(50)\n",
    "df1.loc['Sox17-Dlx4':'TP63-FOSJUNB']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f4388d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filtering if, nan values occure\n",
    "filtered = df[logfc_cluster_cols]\n",
    "\n",
    "val_counts = filtered.count(axis=1).sort_values()\n",
    "#print(val_counts)\n",
    "tfs_occ = val_counts[val_counts >=16].index\n",
    "final = filtered.loc[tfs_occ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b10cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = normalized_dfs[0]\n",
    "cluster_name = \"right-lobe-of-liver.1\"\n",
    "cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "# NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "# This is important: for right-lob-of-liver-1 \n",
    "logfc_cluster_cols = []\n",
    "for entry in cluster_cols:\n",
    "    if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "        logfc_cluster_cols.append(entry)\n",
    "        \n",
    "tmp = df[logfc_cluster_cols]\n",
    "# wtf, tmp[~tmp.isin([0])]??\n",
    "tmp_val_counts = tmp[~tmp.isin([0])].count(axis=1).sort_values()\n",
    "#print(tmp_val_counts)\n",
    "tmp_tfs_occ = tmp_val_counts[tmp_val_counts == len(logfc_cluster_cols)].index\n",
    "result = tmp.loc[tmp_tfs_occ]\n",
    "print(result.shape)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed56edb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = normalized_dfs[0]\n",
    "cluster_name = \"right-lobe-of-liver.1\"\n",
    "cluster_cols = list(filter(lambda x: f'{cluster_name}' in x , df.columns))\n",
    "# NOT WORKING: logfc_cluster_cols = list(filter(lambda x: (f'{cluster_name}/' || f'/{cluster_name}') in x , cluster_cols)) \n",
    "# This is important: for right-lob-of-liver-1 \n",
    "logfc_cluster_cols = []\n",
    "for entry in cluster_cols:\n",
    "    if (f'{cluster_name}/' in entry) or (f'/{cluster_name}_cosine_log2fc' in entry):\n",
    "        logfc_cluster_cols.append(entry)\n",
    "        \n",
    "tmp = df[logfc_cluster_cols]\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89acdd63",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure_threshold_1 = utils.get_threshold(result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc']\n",
    "                                        , \"both\", percent=0.05)\n",
    "measure_threshold_2 = utils.get_threshold(result['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc']\n",
    "                                        , \"both\", percent=0.05)\n",
    "print(f'1: {measure_threshold_1}')\n",
    "print(f'1: {measure_threshold_2}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31975340",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure_threshold_1[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3dda76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#t[t['right-lobe-of-liver.10_cosine'] > 0.7]\n",
    "reduced_result_1 = result[(result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc'] > measure_threshold_1[1]) | (result['right-lobe-of-liver.10/right-lobe-of-liver.1_cosine_log2fc'] < measure_threshold_1[0])]\n",
    "reduced_result = reduced_result_1[(reduced_result_1['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc'] > measure_threshold_2[1]) | (reduced_result_1['right-lobe-of-liver.11/right-lobe-of-liver.1_cosine_log2fc'] < measure_threshold_2[0])]\n",
    "reduced_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70edc13d",
   "metadata": {},
   "outputs": [],
   "source": [
    "f.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed9893bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "m1 = pd.DataFrame(result.mean(axis=1), columns=['mean'])\n",
    "m1['sum'] = result.sum(axis=1)\n",
    "m1.plot.hist(by='mean', bins=100)\n",
    "m1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1782c3ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = filtered.fillna(0)\n",
    "tmp_val_counts = tmp[~tmp.isin([0])].count(axis=1).sort_values()\n",
    "\n",
    "print(tmp_val_counts)\n",
    "tmp_tfs_occ = tmp_val_counts[tmp_val_counts >=16].index\n",
    "result = filtered.loc[tmp_tfs_occ]\n",
    "\n",
    "m1 = pd.DataFrame(result.mean(axis=1), columns=['mean'])\n",
    "m1['sum'] = result.sum(axis=1)\n",
    "m1.plot.hist(by='mean', bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf6b31b0",
   "metadata": {},
   "source": [
    "## Old: Self implemented - Differential analysis - comparing each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f9bb030",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mb market basket analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "print(f\"Files: {files_main_mb}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172636a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diff analysis for all clusters of a tissue:\n",
    "# TODO: what should be compared ? All of a Tissue? or All? Naming?\n",
    "tissue_name = 'right-lobe-of-liver'\n",
    "combObj_to_compare = []\n",
    "for i, file in enumerate(files_main_mb):\n",
    "    print(file)\n",
    "    name_i = file.split('.pkl')[0]\n",
    "    tissue_name = file.split('.')[0]\n",
    "    obj = CombObj().from_pickle(f\"{main_analysis_path}{file}\")\n",
    "    obj.set_prefix(name_i)\n",
    "    #print(obj)\n",
    "    combObj_to_compare.append(obj)\n",
    "    \n",
    "compare_obj = DiffCombObj(combObj_to_compare, measure=\"cosine\", join=\"outer\", fillna=True)\n",
    "compare_obj.to_pickle(f'{differential_analysis_path}{tissue_name}.pkl')\n",
    "compare_obj.normalize()\n",
    "compare_obj.calculate_foldchanges()\n",
    "compare_obj.simplify_rules()\n",
    "compare_obj.to_pickle(f'{differential_analysis_path}{tissue_name}_normalized.pkl')\n",
    "#selection does not work?\n",
    "#selected_std = compare_obj.select_rules()\n",
    "#selected_std.to_pickle(f'{differential_analysis_selection_path}{tissue_name}.pkl')\n",
    "print(\"Done differential analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb1f7b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "td = compare_obj.rules\n",
    "\n",
    "tmp_val_counts = td[td.isin([0])].count(axis=1).sort_values()\n",
    "#tmp_val_counts = td[td.isna()].count(axis=1).sort_values()\n",
    "\n",
    "print(tmp_val_counts)\n",
    "#tmp_tfs_occ = tmp_val_counts[tmp_val_counts >=16].index\n",
    "#result = filtered.loc[tmp_tfs_occ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5266b634",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check original market basket analysis for count of tf-cooccurences\n",
    "print(compare_obj.rules.shape)\n",
    "count_of_all = 0\n",
    "count_of_all_significant = 0\n",
    "for obj in combObj_to_compare:\n",
    "    print(obj.prefix)\n",
    "    print(obj.rules.shape)\n",
    "    obj.simplify_rules()\n",
    "    print(obj.rules.shape)\n",
    "    count_of_all = count_of_all + obj.rules.shape[0]\n",
    "    sig = obj.select_significant_rules()\n",
    "    print(sig.rules.shape)\n",
    "    count_of_all_significant = count_of_all_significant + sig.rules.shape[0]\n",
    "\n",
    "print(\"all simplyfied for tissue\")\n",
    "print(count_of_all)    \n",
    "print(\"all simplyfied for tissue and significant\")\n",
    "print(count_of_all_significant) \n",
    "   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47f3d2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diff analysis between each cluster:\n",
    "for i, file in enumerate(files_main_mb):\n",
    "    print(file)\n",
    "    name_i = file.split('.pkl')[0]\n",
    "    \n",
    "    for j in range(i + 1, len(files_main_mb), 1):\n",
    "        file_j = files_main_mb[j]\n",
    "        name_j = file_j.split('.pkl')[0]\n",
    "        print(j)\n",
    "        print(name_j)\n",
    "        A = CombObj().from_pickle(f\"{main_analysis_path}{file}\")\n",
    "        print(A)\n",
    "        A.set_prefix(name_i)\n",
    "        B = CombObj().from_pickle(f\"{main_analysis_path}{file_j}\")\n",
    "        print(B)\n",
    "        B.set_prefix(name_j)\n",
    "        compare_obj = A.compare(B)\n",
    "        compare_obj.to_pickle(f'{differential_analysis_path}{name_i}__{name_j}.pkl')\n",
    "        \n",
    "        selected_std = compare_obj.select_rules()\n",
    "        \n",
    "        #TODO: Save autamatically generated thresholds\n",
    "        # utils.get_threshold(new.rules.iloc[:,4], 'both', percent=0.05)\n",
    "        # logfc threshold (-xxx , +xxx)\n",
    "        #  utils.get_threshold(new.rules.iloc[:,2:4].mean(axis=1), 'upper', percent=0.05)\n",
    "        # cosine threshold\n",
    "        selected_std.to_pickle(f'{differential_analysis_selection_path}{name_i}__{name_j}.pkl')\n",
    "        \n",
    "        break;\n",
    "        \n",
    "        \n",
    "print(\"Done differential analysis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "777b6701",
   "metadata": {},
   "source": [
    "## OLD implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f3afe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_diff_obj_dataframe(diff_obj: DiffCombObj) -> pd.DataFrame:\n",
    "    \n",
    "    # possible prefix names ['right-lobe-of-liver.10', 'right-lobe-of-liver.16']\n",
    "    df = diff_obj.rules\n",
    "    tissue_name_c1 , cluster_nr_c1  = diff_obj.prefixes[0].split('.')\n",
    "    tissue_name_c2 , cluster_nr_c2  = diff_obj.prefixes[1].split('.')\n",
    "    suff = \"\"\n",
    "    if tissue_name_c1 == tissue_name_c2:\n",
    "        suff += f\"_{tissue_name_c1}\"\n",
    "    else:\n",
    "        suff += f\"_{tissue_name_c1}_{tissue_name_c2}\"\n",
    "\n",
    "    if cluster_nr_c1 == cluster_nr_c2:\n",
    "        suff += f\"_{cluster_nr_c1}\"\n",
    "    else:\n",
    "        suff += f\"_{cluster_nr_c1}_{cluster_nr_c2}\"\n",
    "\n",
    "    df['log2fc_class'] = df.apply(lambda x: 'negativ' if x[4] < 0 else 'positiv', axis=1)\n",
    "    df.columns = [f'{x}{suff}' for x in df.columns]\n",
    "    \n",
    "    return df.copy(deep=True)\n",
    "\n",
    "\n",
    "# Find the specific tf_cooccurences of a tissue that unique for the specific cluster in the tissue.\n",
    "# 1. Diff analyse , \n",
    "# 2. Read in Diffanalyse for the specific cluster\n",
    "# 3. Find tf-cooccurence of the diffob , which are occuring in each cluster\n",
    "# Read in file Names of all analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "#print(f\"Files: {files_main_mb}\")\n",
    "\n",
    "files_diff= read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "print(f\"Count of Files: {len(files_diff)}\")\n",
    "\n",
    "test = \"\"\n",
    "for file_mb in files_main_mb:\n",
    "    cluster_name = file_mb.split('.pkl')[0]\n",
    "    print(cluster_name)\n",
    "    diffs = list(filter(lambda x: cluster_name in x, files_diff))\n",
    "    print(len(diffs))\n",
    "    print(diffs)\n",
    "    \n",
    "    # Keeps the read in DiffCombObj diff_objects:\n",
    "    diff_objects = []\n",
    "    \n",
    "    for diff in diffs:\n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}{diff}\")\n",
    "        diff_objects.append(diff_obj)\n",
    "    \n",
    "    erg = None\n",
    "    for i in range(len(diff_objects)-1):\n",
    "        \n",
    "        if erg is None:\n",
    "            obj_1= diff_objects[i]\n",
    "            obj_2 = diff_objects[i + 1]\n",
    "            df1 = prepare_diff_obj_dataframe(diff_obj = obj_1)\n",
    "            df2 = prepare_diff_obj_dataframe(diff_obj = obj_2)\n",
    "            \n",
    "            erg = df1.merge(df2, how='outer', left_index=True, right_index=True)\n",
    "        else:\n",
    "            obj_2 = diff_objects[i + 1] \n",
    "            df2 = prepare_diff_obj_dataframe(diff_obj = obj_2)\n",
    "            erg = erg.merge(df2, how='outer', left_index=True, right_index=True)\n",
    "       \n",
    "    test = erg\n",
    "    erg.to_pickle(path=f\"{answers_path}{cluster_name}.pkl\")\n",
    "    \n",
    "print(\"Done\")    \n",
    "test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bde5900",
   "metadata": {},
   "outputs": [],
   "source": [
    "answer_file_names=read_in_file_names_of_folder(rel_path=answers_path)\n",
    "print(answer_file_names)\n",
    "cluster_dfs = []\n",
    "df = None\n",
    "for name in answer_file_names:\n",
    "    df = pd.read_pickle(f\"{answers_path}{name}\")\n",
    "    cluster_dfs.append(name)\n",
    "    df = df\n",
    "\n",
    "    #df.groupby(['class', 'value']).count()\n",
    "    break;\n",
    "filter_columns = list(filter(lambda x: 'log2fc_class' in x , df.columns))\n",
    "#len(filter_columns)\n",
    "filtered_df = df[df[filter_columns].notna().all(1)] #\n",
    "filtered_df\n",
    "df\n",
    "#df3.iloc[:, 2:3]\n",
    "#df = pd.read_pickle(f\"{answers_path}right-lobe-of-liver.6.pkl\")\n",
    "#df = pd.read_pickle(f\"{differential_analysis_selection_path}{right-lobe-of-liver.6.pkl}\")\n",
    "\n",
    "\n",
    "#original = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.6.pkl\")\n",
    "#original.rules.loc[df.index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee683c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle(f\"{answers_path}right-lobe-of-liver.6.pkl\")\n",
    "selection = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.10__right-lobe-of-liver.16.pkl\")\n",
    "selection_orig = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.10__right-lobe-of-liver.16.pkl\")\n",
    "selection_orig\n",
    "selection\n",
    "\n",
    "original = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.6.pkl\")\n",
    "original.rules.loc[df.index]\n",
    "selection.prefixes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "245c7d57",
   "metadata": {},
   "source": [
    "### Try and Error section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c1e564",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#mb_obj = CombObj().from_pickle(f\"{main_analysis_path}right-lobe-of-liver.10.pkl\")\n",
    "obj_1 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.1.pkl\")\n",
    "obj_2 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.2.pkl\")\n",
    "obj_3 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.3.pkl\")\n",
    "obj_4 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.4.pkl\")\n",
    "obj_5 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.5.pkl\")\n",
    "obj_6 = DiffCombObj().from_pickle(f\"{differential_analysis_path}right-lobe-of-liver.16__right-lobe-of-liver.6.pkl\")\n",
    "#mb_obj.rules\n",
    "#type(diff_obj.rules)\n",
    "#diff_obj2.rules\n",
    "df_diff = pd.concat([obj_1.rules, obj_2.rules, obj_3.rules, obj_4.rules, obj_5.rules, obj_6.rules], join=\"inner\")\n",
    "\n",
    "\n",
    "\n",
    "#df_diff2 = pd.concat([diff_obj_1_1.rules,diff_obj2.rules, diff_obj1_2.rules])\n",
    "\n",
    "unified_duplicates = df_diff[df_diff.duplicated(subset=['TF1', 'TF2'], keep='first')]\n",
    "\n",
    "df_diff2 = pd.concat([unified_duplicates, diff_obj1_2.rules])\n",
    "\n",
    "unified_duplicates2 = df_diff2[df_diff2.duplicated(subset=['TF1', 'TF2'], keep='first')]\n",
    "\n",
    "#df_diff = df_diff.drop_duplicates(subset=['TF1', 'TF2'])\n",
    "#unified_duplicates\n",
    "#unified_duplicates\n",
    "#unified_duplicates2\n",
    "#diff_obj_1_1.rules\n",
    "#unified_duplicates\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3318406",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_1 = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.16__right-lobe-of-liver.1.pkl\")\n",
    "obj_2 = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}right-lobe-of-liver.16__right-lobe-of-liver.2.pkl\")\n",
    "\n",
    "obj_1.simplify_rules()\n",
    "obj_2.simplify_rules()\n",
    "obj3 = obj_1.rules.merge(obj_2.rules, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "obj3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85338848",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['log2fc_class'] = df.apply(lambda x: 'negativ' if x[4] < 0 else 'positiv', axis=1)\n",
    "\n",
    "#removedNAN = df[df.notna().all(1)]\n",
    "\n",
    "#df2 = removedNAN[(removedNAN[filter_columns] > 0.0) | (removedNAN[filter_columns] < 0.0)]\n",
    "#df2[df2.notna().all(1)]\n",
    "#filtered_df = df[df[filter_columns].notna().all(1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d35c51b",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_obj_1_1.rules.loc['Foxd3-ONECUT2']\n",
    "diff_obj2.rules.loc['Foxd3-ONECUT2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "732674cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C = selectedC.select_top_rules(n=30)\n",
    "top30C.rules.head(31)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcd725aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame({\n",
    "                    'value1': [-1, -3.33, 3, 6,8,7],\n",
    "                    'value2': [-1, 3.33, 3.4, 9,2,7],\n",
    "                    'value3': [1, -3.33, 3, 2,4,7],\n",
    "                    'value4': [1, 3.33, 3, 1,9,7]},\n",
    "                   index=['my1', 'my2', 'my3', 'my4', 'my5', 'my6'])\n",
    "df2 = pd.DataFrame({'rkey': ['foo', 'bar', 'baz', 'foo','test'],\n",
    "                    'value': [5, 6, 7, 8, 9]},\n",
    "                  index=['my1', 'not2', 'my3', 'not4', 'my5'])\n",
    "df3 = pd.DataFrame({'rkey': ['new', 'lol'],\n",
    "                    'value': [5, 6]},\n",
    "                  index=['my1', 'not2'])\n",
    "df1\n",
    "\n",
    "#df1.merge(df2, left_on='lkey', right_on='rkey')\n",
    "#erg = df1.merge(df2, left_index=True, right_index=True, suffixes=(\"_test\", \"_test2\"))\n",
    "#erg = df1.merge(df2, how='outer', left_index=True, right_index=True, suffixes=(\"_x\", \"_y\"))\n",
    "#erg = erg.merge(df3, how='outer', left_index=True, right_index=True, suffixes=(\"_x\", \"_\"))\n",
    "#erg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5c3de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.loc[(df1['value1'] == 1) & df1['value2']== 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607bc563",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1[df1.apply(lambda x: min(abs(x)) == max(abs(x)), 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c4db0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.columns = [f'{x}_df2' for x in df2.columns]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8024c034",
   "metadata": {},
   "outputs": [],
   "source": [
    "[x+ 1 for x in df2.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "615d5305",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1['class'] = df1.apply(lambda x: 'niedrig' if x[1] < 5 else 'hoch', axis=1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d1c3d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1[df1['class'] == 'hoch']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa2c8394",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.groupby(['class', 'value']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa0f86e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4cd1175",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3.iloc[:, 2:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91b30ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C.plot_bubble()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa096c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "top30C.plot_network()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2788d143",
   "metadata": {},
   "outputs": [],
   "source": [
    "selectedC.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1edc217",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34cfa1b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### OLD \n",
    "# Find the specific tf_cooccurences of a tissue that unique for the specific cluster in the tissue.\n",
    "# 1. Diff analyse , \n",
    "# 2. Read in Diffanalyse for the specific cluster\n",
    "# 3. Find tf-cooccurence of the diffob , which are occuring in each cluster\n",
    "# Read in file Names of all analysis\n",
    "files_main_mb= read_in_file_names_of_folder(rel_path=main_analysis_path)\n",
    "print(f\"Count of Files: {len(files_main_mb)}\")\n",
    "#print(f\"Files: {files_main_mb}\")\n",
    "\n",
    "files_diff= read_in_file_names_of_folder(rel_path=differential_analysis_path)\n",
    "print(f\"Count of Files: {len(files_diff)}\")\n",
    "#print(f\"Files: {files_diff}\")\n",
    "test = \"\"\n",
    "for file_mb in files_main_mb:\n",
    "    cluster_name = file_mb.split('.pkl')[0]\n",
    "    print(cluster_name)\n",
    "    diffs = list(filter(lambda x: cluster_name in x, files_diff))\n",
    "    print(len(diffs))\n",
    "    print(diffs)\n",
    "    \n",
    "    # Keeps the read in DiffCombObj diff_objects:\n",
    "    diff_objects = []\n",
    "    \n",
    "    for diff in diffs:\n",
    "        diff_obj = DiffCombObj().from_pickle(f\"{differential_analysis_selection_path}{diff}\")\n",
    "        diff_objects.append(diff_obj)\n",
    "    \n",
    "    # erste DiffObj dataframe \n",
    "    initial_df = diff_objects[0].rules\n",
    "    \n",
    "    #has neg and pos foldchange\n",
    "    cross_product_merged = initial_df\n",
    "    \n",
    "    # only pos foldchange\n",
    "    pos_merged = initial_df[initial_df.iloc[:,4] > 0.00]\n",
    "    \n",
    "    # only neg foldchange\n",
    "    neg_merged = initial_df[initial_df.iloc[:,4] < 0.00]\n",
    "    for i in range(len(diff_objects)-1):\n",
    "        obj_1= diff_objects[i]\n",
    "        obj_2 = diff_objects[i + 1]\n",
    "            \n",
    "        # cross_product merge rules-dataframe by index (TF´s)\n",
    "        cross_product = cross_product_merged.merge(obj_2.rules, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        cross_product_merged = cross_product.copy(deep=True)\n",
    "        \n",
    "        # pos merge rules-dataframe by index (TF´s)\n",
    "        obj2_df_pos = obj_2.rules[obj_2.rules.iloc[:,4] > 0.00]\n",
    "        df_pos_merged = pos_merged.merge(obj2_df_pos, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        pos_merged = df_pos_merged.copy(deep=True)\n",
    "        \n",
    "        # neg merge rules-dataframe by index (TF´s)\n",
    "        obj2_df_neg = obj_2.rules[obj_2.rules.iloc[:,4] < 0.00]\n",
    "        df_neg_merged = neg_merged.merge(obj2_df_neg, left_index=True, right_index=True, suffixes=(f\"_{obj_1.prefixes[0]}_{obj_1.prefixes[1]}\", f\"_{obj_2.prefixes[0]}_{obj_2.prefixes[1]}\"))\n",
    "        neg_merged = df_neg_merged.copy(deep=True)\n",
    "        \n",
    "        \n",
    "    cross_product_merged.to_pickle(path=f\"{answers_path}{cluster_name}_cross.pkl\")\n",
    "    pos_merged.to_pickle(path=f\"{answers_path}{cluster_name}_pos.pkl\")\n",
    "    neg_merged.to_pickle(path=f\"{answers_path}{cluster_name}_neg.pkl\")\n",
    "    \n",
    "print(\"Done\")    \n",
    "test\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tfcomb_env",
   "language": "python",
   "name": "tfcomb_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
