# Datenanalyse-2021

Students Project done at [Loosolab](https://github.molgen.mpg.de/pages/loosolab/www/index.html) [(Max Planck Institute for Heart and Lung Research Bad Nauheim)](https://www.mpi-hlr.de/2778/en) as part of the Masters Degree "Bioinformatics and Systems Biology" at [Justus-Liebig-University Gie√üen](https://www.uni-giessen.de/welcome) and [University of Applied Sciences Mittelhessen (THM)](https://www.thm.de/site/en/). 

The goal of the project is an investigation of the landscape of chromatin accessibility in human cells, computation of transcription factor activity and derivation of transcription factor co-binding and regulatory circuits.

This is done with single cell ATAC-Seq data provided by the [CATlas](http://catlas.org/humanenhancer/#!/) database that is processed in multiple steps by various work packages. 

[Work package 1](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp1/):

WP1 uses [scanpy](https://github.com/theislab/scanpy) and [episcanpy](https://github.com/colomemaria/epiScanpy) for processing of anndata files, 
which includes filtering and clustering of the provided data.

After clustering, peaks are assigned to genes and non-assigned/intergenics are removed.
The [SCSA](https://github.com/bioinfo-ibms-pumc/SCSA) script is used to assign cell types to each cluster, based on the most relevant marker genes.

[Work package 2](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp2/):

The ***preprocessing of snATAC-seq .fastq files using SnapATAC*** is done by WP2. First, a .snap file is created using [SnapTools](https://github.com/r3fang/SnapTools) and the respective .fastq files. This .snap file is read into R as a snap object, which is essential for further processing of the data. Now the following steps will be carried out using the R package [SnapATAC](https://github.com/r3fang/SnapATAC): Barcode filtering, bin filtering, dimensionality reduction (clustering) and finally peak calling of each cluster.
Apart from the processing steps, we provide a Python script to assign cell types to the respective clusters using [Uropa](https://github.com/loosolab/UROPA), [Panglao DB](https://panglaodb.se) and the created peak files.

[Work package 3](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp3):

WP3 regards ***cell-type-specific transcription factor activity***. 
The data provided by WP2 is used to analyze the transcription factor activity of different tissues and their clusters.
Comparisons are drawn between different cell types and tissues to gain insight into the transcription factors responsible for their identity. 

[Work package 4](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp4/):

WP 4 takes a closer look at ***chromatin peak co-accessibilities***. Peaks describe areas of open chromatin. The Cicero algorithm calculates if there are distal peaks to a peak that show the same pattern of open or closed chromatin. Open promoter peaks and corresponding distal peaks are displayed, indicating whether there are open chromatin sites in the distal genome for transcriptional regulation by enhancers or transcription factors.


[Work package 5](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp5/):

The goal of WP5 is to ***discover new motifs*** based on the data generated by WP3.
The WP provides a pipeline to prepare and automatically run the motif discovery pipeline. 
It also offers some scripts to further analyse the newly found motifs and give first insights in their potential biologial meaning.

[Work package 6](https://github.com/loosolab/Datenanalyse-2021/tree/main/wp6/):

WP6 is searching for and ***comparing transcription factor co-occurrences*** based on the data of WP1, WP2, WP3 and WP5 using the python package TF-COMB. WP6 offers jupyter notebooks for finding transcription factor co-occurrences, a notebook for looking into the correlation of the binding orientation and the binding distance between transcription factors and a notebook for a closer look at the difference in the binding distance of same transcription factor co-occurrences in different clusters.

![workflow](https://user-images.githubusercontent.com/81377794/160153768-db0006b2-85a2-41dc-baca-fce3c1250334.png)

More information about the individual work packages can be gathered by following the links provided above or taking a look at the [wiki](https://github.com/loosolab/Datenanalyse-2021/wiki).
