# Work Package 5 - Motif Discovery

## Introduction

The goal of WP5 is to ***discover new motifs*** based on the data generated by WP3.
The WP provides a pipeline to prepare and automatically run the motif discovery pipeline by Hendrik Schultheis. 
Furthermore the workpackage offers some scripts to analyse the newly found motifs and give first insights in their potential biologial meaning.

## Setup and Installation
To use this work package it is important that you have two further tools installed:
* [TOBIAS](https://github.com/loosolab/TOBIAS)
* [motif discovery pipeline](https://github.com/loosolab/motif-discovery-pipeline)

Please follow the installation instructions of those projects and make sure you have the according conda environments installed. It is neccecary that the TOBIAS environment is called "TOBIAS_ENV". 
Furthermore, you need an environment for snakemake wich should also be called "snakemake".

Additionally 
To install the WP5-Package please ... # TODO
--> TODO: environment with Plotly

To use the annotation analysis you have to use two online tools:
* [GO analysis](http://www.pantherdb.org)
* [pathway analysis](https://reactome.org)

For the gene ontology (GO) analysis you have to insert the gene sets from the files you want to analyze. From the analysis results you have to export the "Table". --> TODO: how to rename and store the files
For the pathway analysis you have to choose the [Analysis Tools](https://reactome.org/PathwayBrowser/#TOOL=AT). There you have to insert the gene sets from the files you want to analyze (example: "Gene name list"). From the analysis results you have to export the "Pathway analysis results" and the "Not found identifiers". --> TODO: how to rename and store the files

## Usage
This work package consists of several scripts (found in the *scripts* directory) which can be run individually or as a concatenated pipeline. For more detailed information please refer to the [WP5-Wiki](https://github.com/loosolab/Datenanalyse-2021/wiki/WP5).

The single steps of the pipeline consist of the following scripts:
1. [generate_configs.sh](#1-generate-configs)    
2. [run_pipeline.sh](#2-run-pipeline)
3. [check_logs.sh](#3-check-logs) (optional)
4. [renameAllMotifs.sh](#4-rename-motifs)
5. [cluster_all.sh](#5-cluster-all)
6. [Eval_Motif_similarity.py](#6-evaluate-motif-similarity)
7. [generate_gene_sets.sh](#7-generate-gene-sets)
8. [generate_gene_sets_TFs.sh](#8-generate-gene-sets-for-transcription-factors)
9. [compare_gene_sets.py](#9-compare-the-gene-sets)
10. [analyze_GO_and_pathway.py](#10-analyze-the-gene-sets)

Steps 1-4 are contain the actual motif-discovery-pipeline run, aswell as the needed preparation and post-processing.
Steps 5 and 6 are for the analysis of similarities between the found motifs in different pipeline runs.
Steps 7-10 are for the analysis of the found motifs with the aim to hypothesize the important functions of the motifs.
In case of the given ATAC-Seq data the motif discovery pipeline is ran for each celltype of each tissue given.

Before starting the pipeline (or the single scripts) it is *important* to adjust the config file or the global variables. 
For this please open **global_vars.cnf** and adjust the 4 given variables.
The variables have the following meaning:

* PROJECT_DIR : Path to where all the outputs should be written
* TBSDIR : Path to output of WP3 (should contain all Tissue folders)
* MDP_PIPELINE : path to your installation of the motif discovery pipeline
* ANN_CHECKER=yes : if 'yes' then the motif discovery will be ran with annotation, else annotation will be skipped.

After adjusting the config file for the global variables you can start the pipeline by running the following command:

```
./run_all.sh
```

Please make sure to run this command in the folder where the *run_all.sh* script is stored.
> **Attention**: If you are working via an ssh connection make sure to start this script in a screen, as it might take a while.

### 1. Generate configs
The script *generate_configs.sh* is called as follows:

```
./generate_configs.sh
```

This script will create config files from WP3 output. For each run of WP3 one config will be created. Those configs will be used to start the motif discovery pipeline for the different conditions of the different runs.
Furthermore this script will create all needed directories for further analysis.

### 2. Run pipeline
The script *run_pipeline.sh* is called as follows:

```
./run_pipeline.sh
```

The script will start the motif discovery pipeline for all configs found in the *configs* directory insinde the project directory specified in the config. 
> **Attention**: If you are working via an ssh connection make sure to start this script in a screen, as it might take a while.

### 3. Check logs
This script is an optional script to get a fast overview if all pipeline runs worked as suspected.
The script *check_logs.sh* is called as follows:

```
./check_logs.sh
```

The script will check the *log files* of the analyzed tissues and the belonging cell types to evaluate if the pipeline runs were successful (without annotation ---> TODO). 

### 4. Rename motifs
This script is neccesary, as the motif discovery pipeline will give the motif dummy names. 
As those might appear in several runs, it is important for further analysis to rename the motifs in a unique manner.
This will be done automatically by this script for the MEME file generated by the motif discovery pipeline.

The script can be called as follows:

```
./renameAllMotifs.sh
```
### 5. Cluster all
To establish if any of the found motifs appear in different tissues or celltypes in a similar way, the scripts *cluster_all.sh* and 
*eval_Motif_similarity.py* are provided. 
The idea is to cluster all found motifs with a relatively low threshold (=0.3) using TOBIAS. All clusters found will then be treated as the same motif.
For each cluster a consensus sequence is calculated, and may be used for further analysis. Though, manual inspection is advised before using the consensus sequence.

To perform the clustering, the script is called follows:

```
./cluster_all.sh <Clustering_Name> 
```

The parameter `<Clustering_Name>` will be the prefix the output folder of the clustering will have. Thus it should be somewhat descriptive of the output. In the pipeline *run_all.sh* the given prefix is "overall".

### 6. Evaluate motif similarity
Here the results of the clustering can be evaluated by clustering them. 
To do so you can call the script as follows:

```
./eval_Motif_similarity.py --runs_dir <PATH/TO/RUNS> --motifs <MOTIF_CLUSTER.yml> --out <FILENAME_prefix>
```

For further information on the parameters please refer to the wiki or the help. 
The script will provide several plots to visualize the appearance of similar motifs in different tissues and celltypes. It will also save the names of the found motifs together with their cluster and celltype and tissue in a csv table.

### 7. Generate gene sets
The script *generate_gene_sets.sh* is called as follows:

```
./generate_gene_sets.sh
```

The script will...TODO!!!
> **Attention**: If you are working via an ssh connection make sure to start this script in a screen, as it might take a while.

### 8. Generate gene sets for transcription factors
The script *generate_gene_sets_TFs.sh* is called as follows:

```
./generate_gene_sets_TFs.sh
```

The script will...TODO!!!
> **Attention**: If you are working via an ssh connection make sure to start this script in a screen, as it might take a while.

### 9. Compare the gene sets
The script *compare_gene_sets.py* is called as follows:

```
./compare_gene_sets.py --params TODO!!!
```

The script will... TODO!!!

### 10. Analyze the gene sets
The script *analyze_GO_and_pathway.py* is called as follows:

```
./analyze_GO_and_pathway.py --params TODO!!!
```

The script will... TODO!!!

## Example
TODO
